{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __init__ import *\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import snmcseq_utils\n",
    "import CEMBA_mc_gene_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RS1 mapping summary example\n",
    "\n",
    "# f2 = os.path.join(PATH_DATASETS, 'CEMBA_3C_171206', 'mapping_summary_CEMBA_3C_171206.tsv')\n",
    "# df2 = pd.read_table(f2)\n",
    "# print(df2.shape)\n",
    "# df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add header to RS2 mapping summary\n",
    "# dataset = 'CEMBA_RS2_17Q4'\n",
    "# f = os.path.join(PATH_DATASETS, dataset, '17Q4_meta.txt')\n",
    "# df = pd.read_table(f, header=None, names=df2.columns.tolist()[:15] + ['RS2'])\n",
    "# output = os.path.join(PATH_DATASETS, dataset, 'mapping_summary_{}.tsv'.format(dataset))\n",
    "# df.to_csv(output, sep='\\t', na_rep='NA', header=True, index=False)\n",
    "# print(df.shape)\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "File b'/cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_RS2_17Q4/mapping_summary_CEMBA_RS2_17Q4.tsv' does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-c2ec8f05cbc5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mdataset\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'CEMBA_RS2_17Q4'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mPATH_DATASETS\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdataset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'mapping_summary_{}.tsv'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/cndd/fangming/venvs/venv_banjo_snmcseq/lib/python3.4/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    653\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    654\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 655\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    656\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    657\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/cndd/fangming/venvs/venv_banjo_snmcseq/lib/python3.4/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    403\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    404\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 405\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    406\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    407\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/cndd/fangming/venvs/venv_banjo_snmcseq/lib/python3.4/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    762\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    763\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 764\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    765\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    766\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/cndd/fangming/venvs/venv_banjo_snmcseq/lib/python3.4/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m    983\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    984\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'c'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 985\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    986\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    987\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'python'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/cndd/fangming/venvs/venv_banjo_snmcseq/lib/python3.4/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1603\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'allow_leading_cols'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_col\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1604\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1605\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1607\u001b[0m         \u001b[0;31m# XXX\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__ (pandas/_libs/parsers.c:4209)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source (pandas/_libs/parsers.c:8873)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: File b'/cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_RS2_17Q4/mapping_summary_CEMBA_RS2_17Q4.tsv' does not exist"
     ]
    }
   ],
   "source": [
    "dataset = 'CEMBA_RS2_17Q4'\n",
    "f = os.path.join(PATH_DATASETS, dataset, 'mapping_summary_{}.tsv'.format(dataset))\n",
    "df = pd.read_table(f)\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Total reads</th>\n",
       "      <th>Mapped reads</th>\n",
       "      <th>Mapping rate</th>\n",
       "      <th>Nonclonal reads</th>\n",
       "      <th>% Nonclonal rate</th>\n",
       "      <th>Filtered reads</th>\n",
       "      <th>Filtered rate</th>\n",
       "      <th>Lambda mC/C</th>\n",
       "      <th>mCCC/CCC</th>\n",
       "      <th>mCG/CG</th>\n",
       "      <th>mCH/CH</th>\n",
       "      <th>Estimated mCG/CG</th>\n",
       "      <th>Estimated mCH/CH</th>\n",
       "      <th>% Genome covered</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Bm3C_rep1 MOP 3C SSp male</th>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>54</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bm3C_rep2 MOP 3C SSp male</th>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>31</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bm4B_rep1 MOP 4B SSp male</th>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>41</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bm4B_rep2 MOP 4B SSp male</th>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>43</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pf3C MOP 3C STR female</th>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>1</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pf4B MOP 4B STR female</th>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>48</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pm3C MOP 3C STR male</th>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>56</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pm4B MOP 4B STR male</th>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>44</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tf3C MOP 3C SC female</th>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>41</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tf4B MOP 4B SC female</th>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>49</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tf6B SSp 6B SC female</th>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>53</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tf7B SSp 7B SC female</th>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>49</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tm3C MOP 3C SC male</th>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>65</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tm4B MOP 4B SC male</th>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>45</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tm6B SSp 6B SC male</th>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>56</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tm7B SSp 7B SC male</th>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>47</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Sample  Total reads  Mapped reads  Mapping rate  \\\n",
       "RS2                                                                          \n",
       "Bm3C_rep1 MOP 3C SSp male      91           91            91            91   \n",
       "Bm3C_rep2 MOP 3C SSp male      50           50            50            50   \n",
       "Bm4B_rep1 MOP 4B SSp male      89           89            89            89   \n",
       "Bm4B_rep2 MOP 4B SSp male      83           83            83            83   \n",
       "Pf3C MOP 3C STR female         96           96            96            96   \n",
       "Pf4B MOP 4B STR female         95           95            95            95   \n",
       "Pm3C MOP 3C STR male           82           82            82            82   \n",
       "Pm4B MOP 4B STR male           92           92            92            92   \n",
       "Tf3C MOP 3C SC female          55           55            55            55   \n",
       "Tf4B MOP 4B SC female          92           92            92            92   \n",
       "Tf6B SSp 6B SC female          85           85            85            85   \n",
       "Tf7B SSp 7B SC female          82           82            82            82   \n",
       "Tm3C MOP 3C SC male            82           82            82            82   \n",
       "Tm4B MOP 4B SC male            86           86            86            86   \n",
       "Tm6B SSp 6B SC male            89           89            89            89   \n",
       "Tm7B SSp 7B SC male            81           81            81            81   \n",
       "\n",
       "                           Nonclonal reads  % Nonclonal rate  Filtered reads  \\\n",
       "RS2                                                                            \n",
       "Bm3C_rep1 MOP 3C SSp male               91                91              91   \n",
       "Bm3C_rep2 MOP 3C SSp male               50                50              50   \n",
       "Bm4B_rep1 MOP 4B SSp male               89                89              89   \n",
       "Bm4B_rep2 MOP 4B SSp male               83                83              83   \n",
       "Pf3C MOP 3C STR female                  96                96              96   \n",
       "Pf4B MOP 4B STR female                  95                95              95   \n",
       "Pm3C MOP 3C STR male                    82                82              82   \n",
       "Pm4B MOP 4B STR male                    92                92              92   \n",
       "Tf3C MOP 3C SC female                   55                55              55   \n",
       "Tf4B MOP 4B SC female                   92                92              92   \n",
       "Tf6B SSp 6B SC female                   85                85              85   \n",
       "Tf7B SSp 7B SC female                   82                82              82   \n",
       "Tm3C MOP 3C SC male                     82                82              82   \n",
       "Tm4B MOP 4B SC male                     86                86              86   \n",
       "Tm6B SSp 6B SC male                     89                89              89   \n",
       "Tm7B SSp 7B SC male                     81                81              81   \n",
       "\n",
       "                           Filtered rate  Lambda mC/C  mCCC/CCC  mCG/CG  \\\n",
       "RS2                                                                       \n",
       "Bm3C_rep1 MOP 3C SSp male             91           54        91      91   \n",
       "Bm3C_rep2 MOP 3C SSp male             50           31        50      50   \n",
       "Bm4B_rep1 MOP 4B SSp male             89           41        89      89   \n",
       "Bm4B_rep2 MOP 4B SSp male             83           43        83      83   \n",
       "Pf3C MOP 3C STR female                96            1        96      96   \n",
       "Pf4B MOP 4B STR female                95           48        95      95   \n",
       "Pm3C MOP 3C STR male                  82           56        82      82   \n",
       "Pm4B MOP 4B STR male                  92           44        92      92   \n",
       "Tf3C MOP 3C SC female                 55           41        55      55   \n",
       "Tf4B MOP 4B SC female                 92           49        92      92   \n",
       "Tf6B SSp 6B SC female                 85           53        85      85   \n",
       "Tf7B SSp 7B SC female                 82           49        82      82   \n",
       "Tm3C MOP 3C SC male                   82           65        82      82   \n",
       "Tm4B MOP 4B SC male                   86           45        86      86   \n",
       "Tm6B SSp 6B SC male                   89           56        89      89   \n",
       "Tm7B SSp 7B SC male                   81           47        81      81   \n",
       "\n",
       "                           mCH/CH  Estimated mCG/CG  Estimated mCH/CH  \\\n",
       "RS2                                                                     \n",
       "Bm3C_rep1 MOP 3C SSp male      91                91                91   \n",
       "Bm3C_rep2 MOP 3C SSp male      50                50                50   \n",
       "Bm4B_rep1 MOP 4B SSp male      89                89                89   \n",
       "Bm4B_rep2 MOP 4B SSp male      83                83                83   \n",
       "Pf3C MOP 3C STR female         96                96                96   \n",
       "Pf4B MOP 4B STR female         95                95                95   \n",
       "Pm3C MOP 3C STR male           82                82                82   \n",
       "Pm4B MOP 4B STR male           92                92                92   \n",
       "Tf3C MOP 3C SC female          55                55                55   \n",
       "Tf4B MOP 4B SC female          92                92                92   \n",
       "Tf6B SSp 6B SC female          85                85                85   \n",
       "Tf7B SSp 7B SC female          82                82                82   \n",
       "Tm3C MOP 3C SC male            82                82                82   \n",
       "Tm4B MOP 4B SC male            86                86                86   \n",
       "Tm6B SSp 6B SC male            89                89                89   \n",
       "Tm7B SSp 7B SC male            81                81                81   \n",
       "\n",
       "                           % Genome covered  \n",
       "RS2                                          \n",
       "Bm3C_rep1 MOP 3C SSp male                91  \n",
       "Bm3C_rep2 MOP 3C SSp male                50  \n",
       "Bm4B_rep1 MOP 4B SSp male                89  \n",
       "Bm4B_rep2 MOP 4B SSp male                83  \n",
       "Pf3C MOP 3C STR female                   96  \n",
       "Pf4B MOP 4B STR female                   95  \n",
       "Pm3C MOP 3C STR male                     82  \n",
       "Pm4B MOP 4B STR male                     92  \n",
       "Tf3C MOP 3C SC female                    55  \n",
       "Tf4B MOP 4B SC female                    92  \n",
       "Tf6B SSp 6B SC female                    85  \n",
       "Tf7B SSp 7B SC female                    82  \n",
       "Tm3C MOP 3C SC male                      82  \n",
       "Tm4B MOP 4B SC male                      86  \n",
       "Tm6B SSp 6B SC male                      89  \n",
       "Tm7B SSp 7B SC male                      81  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('RS2').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(91, 15)\n",
      "(50, 15)\n",
      "(89, 15)\n",
      "(83, 15)\n",
      "(96, 15)\n",
      "(95, 15)\n",
      "(82, 15)\n",
      "(92, 15)\n",
      "(55, 15)\n",
      "(92, 15)\n",
      "(85, 15)\n",
      "(82, 15)\n",
      "(82, 15)\n",
      "(86, 15)\n",
      "(89, 15)\n",
      "(81, 15)\n"
     ]
    }
   ],
   "source": [
    "# create dataset dir and split mapping summary\n",
    "for rs2_code, df_sub in df.groupby('RS2'):\n",
    "    rs2_code = rs2_code.split(' ')\n",
    "    dataset = 'CEMBA_RS2_{}'.format(rs2_code[0])\n",
    "    # create dataset dir\n",
    "    os.makedirs(os.path.join(PATH_DATASETS, dataset))\n",
    "    # copy metadata over\n",
    "    df_sub = df_sub.drop('RS2', axis=1)\n",
    "    df_sub.to_csv(os.path.join(PATH_DATASETS, d taset, 'mapping_summary_{}.tsv'.format(dataset)), \n",
    "                  sep='\\t', na_rep='NA', header=True, index=False)\n",
    "    print(df_sub.shape)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# distribute files over \n",
    "dataset_origin = 'CEMBA_RS2_17Q4'\n",
    "for rs2_code, df_sub in df.groupby('RS2'):\n",
    "    rs2_code = rs2_code.split(' ')\n",
    "    # dataset\n",
    "    dataset = 'CEMBA_RS2_{}'.format(rs2_code[0])\n",
    "    \n",
    "    # cells in the dataset\n",
    "    cells = df_sub['Sample'].values\n",
    "    \n",
    "    allc_srcs = [os.path.join(PATH_DATASETS, dataset_origin, 'allc', 'allc_{}.tsv.bgz'.format(cell))\n",
    "                 for cell in cells]\n",
    "    allc_dsts = [os.path.join(PATH_DATASETS, dataset, 'allc', \n",
    "                              os.path.basename(allc_src)) for allc_src in allc_srcs] \n",
    "    \n",
    "    # create allc dir\n",
    "    if not os.path.isdir(os.path.join(PATH_DATASETS, dataset, 'allc')): \n",
    "        os.makedirs(os.path.join(PATH_DATASETS, dataset, 'allc'))\n",
    "        \n",
    "#     for src, dst in zip(allc_srcs, allc_dsts):\n",
    "#         os.rename(src, dst)\n",
    "\n",
    "        \n",
    "#     # create binc dir\n",
    "#     if not os.path.isdir(os.path.join(PATH_DATASETS, dataset, 'binc')): \n",
    "#         os.makedirs(os.path.join(PATH_DATASETS, dataset, 'binc'))\n",
    "        \n",
    "#     # create gene_level dir\n",
    "#     if not os.path.isdir(os.path.join(PATH_DATASETS, dataset, 'gene_level')): \n",
    "#         os.makedirs(os.path.join(PATH_DATASETS, dataset, 'gene_level'))\n",
    "\n",
    "#     allc_tabix_srcs = [os.path.join(PATH_DATASETS, dataset_origin, 'allc', 'allc_{}.tsv.bgz.tbi'.format(cell))\n",
    "#                  for cell in cells]\n",
    "#     allc_tabix_dsts = [os.path.join(PATH_DATASETS, dataset, 'allc', \n",
    "#                               os.path.basename(allc_tabix_src)) for allc_tabix_src in allc_tabix_srcs] \n",
    "    \n",
    "#     binc_srcs = [os.path.join(PATH_DATASETS, dataset_origin, 'binc', 'binc_{}_10000.tsv.bgz'.format(cell))\n",
    "#                  for cell in cells]\n",
    "#     binc_dsts = [os.path.join(PATH_DATASETS, dataset, 'binc', \n",
    "#                               os.path.basename(binc_src)) for binc_src in binc_srcs] \n",
    "    \n",
    "#     genebody_srcs = [os.path.join(PATH_DATASETS, dataset_origin, 'gene_level', 'genebody_{}.tsv.bgz'.format(cell))\n",
    "#                  for cell in cells]\n",
    "#     genebody_dsts = [os.path.join(PATH_DATASETS, dataset, 'gene_level', \n",
    "#                               os.path.basename(genebody_src)) for genebody_src in genebody_srcs] \n",
    "        \n",
    "#     print(binc_srcs[:1]) \n",
    "#     print(binc_dsts[:1]) \n",
    "#     print(genebody_srcs[:1]) \n",
    "#     print(genebody_dsts[:1]) \n",
    "\n",
    "#     for src, dst in zip(allc_tabix_srcs, allc_tabix_dsts):\n",
    "#         os.rename(src, dst)\n",
    "\n",
    "#     for src, dst in zip(binc_srcs, binc_dsts):\n",
    "#         try:\n",
    "#             os.rename(src, dst)\n",
    "#         except:\n",
    "#             pass\n",
    "#     for src, dst in zip(genebody_srcs, genebody_dsts):\n",
    "#         try:\n",
    "#             os.rename(src, dst)\n",
    "#         except:\n",
    "#             pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4C Error in opening mapping summary\n",
      "CEMBA_1A_180226 1366 All good!\n",
      "CEMBA_1A_180227 1100 All good!\n",
      "CEMBA_1B_180119 1435 All good!\n",
      "CEMBA_1B_180213 1404 All good!\n",
      "CEMBA_1C_180208 1383 All good!\n",
      "CEMBA_1C_180212 1266 All good!\n",
      "CEMBA_2B_180305 1040 All good!\n",
      "CEMBA_2B_180306 1418 All good!\n",
      "CEMBA_2D_180313 544 All good!\n",
      "CEMBA_2D_180319 314 All good!\n",
      "CEMBA_2D_180403 1608 All good!\n",
      "CEMBA_2E_180220 1445 All good!\n",
      "CEMBA_2E_180222 1253 All good!\n",
      "CEMBA_3A_180129 1384 All good!\n",
      "CEMBA_3A_180130 1291 All good!\n",
      "CEMBA_3C_171206 1202 All good!\n",
      "CEMBA_3C_171207 1270 All good!\n",
      "CEMBA_3E_180320 1339 All good!\n",
      "CEMBA_3E_180326 1190 All good!\n",
      "CEMBA_3F_180105 628 All good!\n",
      "CEMBA_3F_180109 1059 All good!\n",
      "CEMBA_3F_180109_v2 691 All good!\n",
      "CEMBA_4A_180205 1190 All good!\n",
      "CEMBA_4A_180206 1302 All good!\n",
      "CEMBA_4B_171212 296 All good!\n",
      "CEMBA_4B_171213 1261 All good!\n",
      "CEMBA_4B_180104 1323 All good!\n",
      "CEMBA_4C_180417 1049 Error genebody! (1042)\n",
      "CEMBA_4C_180419 1005 All good!\n",
      "CEMBA_4D_171214 1106 All good!\n",
      "CEMBA_4D_171219 1433 All good!\n",
      "CEMBA_4E_180110 1224 All good!\n",
      "CEMBA_4E_180111 540 All good!\n",
      "CEMBA_4F_180329 1336 All good!\n",
      "CEMBA_4F_180402 1258 All good!\n",
      "CEMBA_RS2_Bm3C_rep1 91 All good!\n",
      "CEMBA_RS2_Bm3C_rep2 50 All good!\n",
      "CEMBA_RS2_Bm4B_rep1 89 All good!\n",
      "CEMBA_RS2_Bm4B_rep2 83 All good!\n",
      "CEMBA_RS2_Pf3C 96 All good!\n",
      "CEMBA_RS2_Pf4B 95 All good!\n",
      "CEMBA_RS2_Pm3C 82 All good!\n",
      "CEMBA_RS2_Pm4B 92 All good!\n",
      "CEMBA_RS2_Tf3C 55 All good!\n",
      "CEMBA_RS2_Tf4B 92 All good!\n",
      "CEMBA_RS2_Tf6B 85 All good!\n",
      "CEMBA_RS2_Tf7B 82 All good!\n",
      "CEMBA_RS2_Tm3C 82 All good!\n",
      "CEMBA_RS2_Tm4B 86 All good!\n",
      "CEMBA_RS2_Tm6B 89 All good!\n",
      "CEMBA_RS2_Tm7B 81 All good!\n",
      "CEMBA_SCI_2017 Error in opening mapping summary\n",
      "MOp_downsample_05 Error in opening mapping summary\n",
      "archives Error in opening mapping summary\n"
     ]
    }
   ],
   "source": [
    "# check if all datasets are good dataset\n",
    "with snmcseq_utils.cd(PATH_DATASETS):\n",
    "    datasets = sorted(glob.glob('*'))\n",
    "for dataset in datasets:\n",
    "    with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset)):\n",
    "        try:\n",
    "            df_meta = pd.read_table('mapping_summary_{}.tsv'.format(dataset))\n",
    "            n_cells = df_meta.shape[0]\n",
    "        except:\n",
    "            print(dataset, \"Error in opening mapping summary\")\n",
    "            continue\n",
    "        \n",
    "        n_allc = len(glob.glob('./allc/allc_*.tsv.bgz'))\n",
    "        n_allc_tbi = len(glob.glob('./allc/allc_*.tsv.bgz.tbi'))\n",
    "        \n",
    "        n_binc = len(glob.glob('./binc/binc_*.tsv.bgz'))\n",
    "        n_genebody = len(glob.glob('./gene_level/genebody_*.tsv.bgz'))\n",
    "        \n",
    "        flag = 0\n",
    "        if n_cells != n_allc:\n",
    "            print(dataset, n_cells, 'Error allc! ({})'.format(n_allc))\n",
    "            flag += 1\n",
    "        if n_cells != n_allc_tbi:\n",
    "            print(dataset, n_cells, 'Error allc tabix! ({})'.format(n_allc_tbi))\n",
    "            flag += 1\n",
    "        if n_cells != n_binc:\n",
    "            print(dataset, n_cells, 'Error binc! ({})'.format(n_binc))\n",
    "            flag += 1\n",
    "        if n_cells != n_genebody:\n",
    "            print(dataset, n_cells, 'Error genebody! ({})'.format(n_genebody))\n",
    "            flag += 1\n",
    "        \n",
    "        if flag == 0:\n",
    "            pass\n",
    "            print(dataset, n_cells, \"All good!\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**\n",
      "**\n",
      "**\n"
     ]
    }
   ],
   "source": [
    "dataset = 'CEMBA_4C_180417'\n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'allc')):\n",
    "    allc_files = glob.glob('*.bgz')\n",
    "    cells_allc = [allc_file[len('allc_'):-len('.tsv.bgz')] for allc_file in allc_files]\n",
    "    \n",
    "f = '/cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/{0}/mapping_summary_{0}.tsv'.format(dataset)\n",
    "cells = pd.read_table(f)['Sample'].values\n",
    "\n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'binc')):\n",
    "    files = glob.glob('*.bgz')\n",
    "    cells_binc = [file[len('binc_'):-len('_10000.tsv.bgz')] for file in files]\n",
    "    \n",
    "for cell_allc in cells_allc:\n",
    "    if cell_allc not in cells:\n",
    "        print(cell_allc)\n",
    "\n",
    "print('**')\n",
    "\n",
    "for cell in cells:\n",
    "    if cell not in cells_allc:\n",
    "        print(cell)\n",
    "        \n",
    "print('**')\n",
    "\n",
    "for cell_binc in cells_binc:\n",
    "    if cell_binc not in cells:\n",
    "        print(cell_binc)\n",
    "\n",
    "print('**')\n",
    "\n",
    "for cell in cells:\n",
    "    if cell not in cells_binc:\n",
    "        print(cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed',\n",
       " '180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed',\n",
       " '180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed',\n",
       " '180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed',\n",
       " '180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed',\n",
       " '180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed',\n",
       " '180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'CEMBA_4C_180417'\n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'allc')):\n",
    "    allc_files = glob.glob('*.bgz')\n",
    "    cells = [allc_file[len('allc_'):-len('.tsv.bgz')] for allc_file in allc_files]\n",
    "    \n",
    "    \n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'gene_level')):\n",
    "    files = glob.glob('*')\n",
    "    cells_redo = []\n",
    "    for cell in cells:\n",
    "        if 'genebody_'+cell+'.tsv.bgz' not in files:\n",
    "            cells_redo.append(cell)\n",
    "cells_redo "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:44:40 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      "08/15/2018 11:45:41 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed.tsv\n",
      "08/15/2018 11:45:41 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:47:18 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed.tsv\n",
      "08/15/2018 11:47:18 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:48:22 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed.tsv\n",
      "08/15/2018 11:48:22 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:49:31 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed.tsv\n",
      "08/15/2018 11:49:32 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:50:41 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed.tsv\n",
      "08/15/2018 11:50:41 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:51:40 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed.tsv\n",
      "08/15/2018 11:51:41 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:52:59 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed.tsv\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed\n"
     ]
    }
   ],
   "source": [
    "log = snmcseq_utils.create_logger()\n",
    "\n",
    "for cell in cells_redo:\n",
    "    allc_file = os.path.join(PATH_DATASETS, dataset, 'allc', 'allc_' + cell + '.tsv.bgz')\n",
    "\n",
    "    try:\n",
    "        CEMBA_mc_gene_level.mc_gene_level(allc_file,\n",
    "            contexts=CONTEXTS, \n",
    "            genebody=GENEBODY,\n",
    "            convention='CEMBA', \n",
    "            overwrite=True)\n",
    "        print(cell)\n",
    "    except:\n",
    "        pass\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mataset = 'CEMBA_4F_180329'\n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'allc')):\n",
    "    allc_files = glob.glob('*.bgz')\n",
    "    mells = [allc_file[len('allc_'):-len('.tsv.bgz')] for allc_file in allc_files]\n",
    "    \n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'binc')):\n",
    "    files = glob.glob('*')\n",
    "    cells_redo = []\n",
    "    for cell in cells:\n",
    "        if 'binc_'+cell+'_10000.tsv.bgz' not in files:\n",
    "            cells_redo.append(cell)\n",
    "cells_redo "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/06/2018 09:37:18 AM binc processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4F_180329/allc/allc_180405_CEMBA_mm_P56_P63_4F_CEMBA180329_4F_1_CEMBA180329_4F_2_H1_AD012_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      "08/06/2018 09:40:08 AM Done with binc processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4F_180329/allc/allc_180405_CEMBA_mm_P56_P63_4F_CEMBA180329_4F_1_CEMBA180329_4F_2_H1_AD012_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      "Saving results to: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4F_180329/binc/binc_180405_CEMBA_mm_P56_P63_4F_CEMBA180329_4F_1_CEMBA180329_4F_2_H1_AD012_indexed_10000.tsv\n"
     ]
    }
   ],
   "source": [
    "import CEMBA_bin_allc_files\n",
    "\n",
    "log = snmcseq_utils.create_logger()\n",
    "for cell in cells_redo:\n",
    "    allc_file = os.path.join(PATH_DATASETS, dataset, 'allc', 'allc_' + cell + '.tsv.bgz')\n",
    "    \n",
    "    CEMBA_bin_allc_files.bin_allc(allc_file, \n",
    "        convention='CEMBA',\n",
    "        bin_size=BIN_SIZE, \n",
    "        contexts=CONTEXTS,\n",
    "        chromosomes=None, \n",
    "        species='mouse',\n",
    "        compression='gzip',\n",
    "        overwrite=False\n",
    "        )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
