{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __init__ import *\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "\n",
    "import snmcseq_utils\n",
    "import CEMBA_mc_gene_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1202, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Total reads</th>\n",
       "      <th>Mapped reads</th>\n",
       "      <th>Mapping rate</th>\n",
       "      <th>Nonclonal reads</th>\n",
       "      <th>% Nonclonal rate</th>\n",
       "      <th>Filtered reads</th>\n",
       "      <th>Filtered rate</th>\n",
       "      <th>Lambda mC/C</th>\n",
       "      <th>mCCC/CCC</th>\n",
       "      <th>mCG/CG</th>\n",
       "      <th>mCH/CH</th>\n",
       "      <th>Estimated mCG/CG</th>\n",
       "      <th>Estimated mCH/CH</th>\n",
       "      <th>% Genome covered</th>\n",
       "      <th>mCA/CA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...</td>\n",
       "      <td>2279422</td>\n",
       "      <td>1443850</td>\n",
       "      <td>0.633</td>\n",
       "      <td>941772</td>\n",
       "      <td>0.652</td>\n",
       "      <td>819822</td>\n",
       "      <td>0.871</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00789</td>\n",
       "      <td>0.76195</td>\n",
       "      <td>0.02756</td>\n",
       "      <td>0.76006</td>\n",
       "      <td>0.01983</td>\n",
       "      <td>3.36</td>\n",
       "      <td>0.050792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...</td>\n",
       "      <td>4252158</td>\n",
       "      <td>2826725</td>\n",
       "      <td>0.665</td>\n",
       "      <td>2184223</td>\n",
       "      <td>0.773</td>\n",
       "      <td>1747325</td>\n",
       "      <td>0.800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00773</td>\n",
       "      <td>0.76572</td>\n",
       "      <td>0.02803</td>\n",
       "      <td>0.76389</td>\n",
       "      <td>0.02046</td>\n",
       "      <td>7.17</td>\n",
       "      <td>0.052219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...</td>\n",
       "      <td>3649262</td>\n",
       "      <td>2447983</td>\n",
       "      <td>0.671</td>\n",
       "      <td>1887885</td>\n",
       "      <td>0.771</td>\n",
       "      <td>1629253</td>\n",
       "      <td>0.863</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.00856</td>\n",
       "      <td>0.77589</td>\n",
       "      <td>0.03182</td>\n",
       "      <td>0.77396</td>\n",
       "      <td>0.02346</td>\n",
       "      <td>6.77</td>\n",
       "      <td>0.059366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...</td>\n",
       "      <td>4386850</td>\n",
       "      <td>2943685</td>\n",
       "      <td>0.671</td>\n",
       "      <td>2301085</td>\n",
       "      <td>0.782</td>\n",
       "      <td>1959547</td>\n",
       "      <td>0.852</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00818</td>\n",
       "      <td>0.77278</td>\n",
       "      <td>0.02875</td>\n",
       "      <td>0.77091</td>\n",
       "      <td>0.02074</td>\n",
       "      <td>8.04</td>\n",
       "      <td>0.053225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...</td>\n",
       "      <td>3861170</td>\n",
       "      <td>2568841</td>\n",
       "      <td>0.665</td>\n",
       "      <td>1870749</td>\n",
       "      <td>0.728</td>\n",
       "      <td>1594420</td>\n",
       "      <td>0.852</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00759</td>\n",
       "      <td>0.76714</td>\n",
       "      <td>0.02690</td>\n",
       "      <td>0.76536</td>\n",
       "      <td>0.01946</td>\n",
       "      <td>6.59</td>\n",
       "      <td>0.049690</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Sample  Total reads  \\\n",
       "0  171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...      2279422   \n",
       "1  171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...      4252158   \n",
       "2  171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...      3649262   \n",
       "3  171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...      4386850   \n",
       "4  171213_CEMBA_mm_P56_P63_3C_MOp_CEMBA171206_3C_...      3861170   \n",
       "\n",
       "   Mapped reads  Mapping rate  Nonclonal reads  % Nonclonal rate  \\\n",
       "0       1443850         0.633           941772             0.652   \n",
       "1       2826725         0.665          2184223             0.773   \n",
       "2       2447983         0.671          1887885             0.771   \n",
       "3       2943685         0.671          2301085             0.782   \n",
       "4       2568841         0.665          1870749             0.728   \n",
       "\n",
       "   Filtered reads  Filtered rate  Lambda mC/C  mCCC/CCC   mCG/CG   mCH/CH  \\\n",
       "0          819822          0.871         0.00   0.00789  0.76195  0.02756   \n",
       "1         1747325          0.800          NaN   0.00773  0.76572  0.02803   \n",
       "2         1629253          0.863         0.75   0.00856  0.77589  0.03182   \n",
       "3         1959547          0.852          NaN   0.00818  0.77278  0.02875   \n",
       "4         1594420          0.852          NaN   0.00759  0.76714  0.02690   \n",
       "\n",
       "   Estimated mCG/CG  Estimated mCH/CH  % Genome covered    mCA/CA  \n",
       "0           0.76006           0.01983              3.36  0.050792  \n",
       "1           0.76389           0.02046              7.17  0.052219  \n",
       "2           0.77396           0.02346              6.77  0.059366  \n",
       "3           0.77091           0.02074              8.04  0.053225  \n",
       "4           0.76536           0.01946              6.59  0.049690  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RS1 mapping summary example\n",
    "\n",
    "f2 = os.path.join(PATH_DATASETS, 'CEMBA_3C_171206', 'mapping_summary_CEMBA_3C_171206.tsv')\n",
    "df2 = pd.read_table(f2)\n",
    "print(df2.shape)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add header to RS2 mapping summary\n",
    "# dataset = 'CEMBA_RS2_17Q4'\n",
    "# f = os.path.join(PATH_DATASETS, dataset, '17Q4_meta.txt')\n",
    "# df = pd.read_table(f, header=None, names=df2.columns.tolist()[:15] + ['RS2'])\n",
    "# output = os.path.join(PATH_DATASETS, dataset, 'mapping_summary_{}.tsv'.format(dataset))\n",
    "# df.to_csv(output, sep='\\t', na_rep='NA', header=True, index=False)\n",
    "# print(df.shape)\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1535, 26)\n",
      "(1535, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Total reads</th>\n",
       "      <th>Mapped reads</th>\n",
       "      <th>Mapping rate</th>\n",
       "      <th>Nonclonal reads</th>\n",
       "      <th>% Nonclonal rate</th>\n",
       "      <th>Filtered reads</th>\n",
       "      <th>Filtered rate</th>\n",
       "      <th>Lambda mC/C</th>\n",
       "      <th>mCCC/CCC</th>\n",
       "      <th>...</th>\n",
       "      <th>CCC</th>\n",
       "      <th>mCH</th>\n",
       "      <th>CH</th>\n",
       "      <th>mCG</th>\n",
       "      <th>CG</th>\n",
       "      <th>projection</th>\n",
       "      <th>source region</th>\n",
       "      <th>source slides</th>\n",
       "      <th>target region</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1069475</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005129</td>\n",
       "      <td>...</td>\n",
       "      <td>1385553</td>\n",
       "      <td>478280</td>\n",
       "      <td>20427236</td>\n",
       "      <td>648736</td>\n",
       "      <td>868511</td>\n",
       "      <td>Pm6B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>6B</td>\n",
       "      <td>STR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1127711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.005936</td>\n",
       "      <td>...</td>\n",
       "      <td>1439820</td>\n",
       "      <td>573022</td>\n",
       "      <td>21370107</td>\n",
       "      <td>703418</td>\n",
       "      <td>898142</td>\n",
       "      <td>Pm7B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>7B</td>\n",
       "      <td>STR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1149770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006778</td>\n",
       "      <td>...</td>\n",
       "      <td>1463330</td>\n",
       "      <td>699355</td>\n",
       "      <td>21736617</td>\n",
       "      <td>700924</td>\n",
       "      <td>908103</td>\n",
       "      <td>Pf6B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>6B</td>\n",
       "      <td>STR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>965382</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.004047</td>\n",
       "      <td>...</td>\n",
       "      <td>1246244</td>\n",
       "      <td>301371</td>\n",
       "      <td>18138018</td>\n",
       "      <td>574039</td>\n",
       "      <td>751098</td>\n",
       "      <td>Pf7B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>6B</td>\n",
       "      <td>STR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>714865</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005094</td>\n",
       "      <td>...</td>\n",
       "      <td>928585</td>\n",
       "      <td>290528</td>\n",
       "      <td>13583684</td>\n",
       "      <td>443955</td>\n",
       "      <td>577062</td>\n",
       "      <td>Pm6B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>6B</td>\n",
       "      <td>STR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Sample  Total reads  \\\n",
       "0  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "1  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "2  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "3  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "4  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "\n",
       "   Mapped reads  Mapping rate  Nonclonal reads  % Nonclonal rate  \\\n",
       "0           NaN           NaN          1069475               NaN   \n",
       "1           NaN           NaN          1127711               NaN   \n",
       "2           NaN           NaN          1149770               NaN   \n",
       "3           NaN           NaN           965382               NaN   \n",
       "4           NaN           NaN           714865               NaN   \n",
       "\n",
       "   Filtered reads  Filtered rate  Lambda mC/C  mCCC/CCC   ...        CCC  \\\n",
       "0             NaN            NaN          0.0  0.005129   ...    1385553   \n",
       "1             NaN            NaN          NaN  0.005936   ...    1439820   \n",
       "2             NaN            NaN          0.0  0.006778   ...    1463330   \n",
       "3             NaN            NaN          NaN  0.004047   ...    1246244   \n",
       "4             NaN            NaN          0.0  0.005094   ...     928585   \n",
       "\n",
       "      mCH        CH     mCG      CG  projection  source region  source slides  \\\n",
       "0  478280  20427236  648736  868511        Pm6B            SSp             6B   \n",
       "1  573022  21370107  703418  898142        Pm7B            SSp             7B   \n",
       "2  699355  21736617  700924  908103        Pf6B            SSp             6B   \n",
       "3  301371  18138018  574039  751098        Pf7B            SSp             6B   \n",
       "4  290528  13583684  443955  577062        Pm6B            SSp             6B   \n",
       "\n",
       "   target region  gender  \n",
       "0            STR    male  \n",
       "1            STR    male  \n",
       "2            STR  female  \n",
       "3            STR  female  \n",
       "4            STR    male  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dataset = 'CEMBA_RS2_18Q1'\n",
    "# # f = os.path.join(PATH_DATASETS, dataset, 'mapping_summary_{}.tsv'.format(dataset))\n",
    "# f = '/cndd/Public_Datasets/CEMBA/snmCSeq/References/Mapping_summary/mapping_summary_CEMBA_RS2_18Q1.tsv'\n",
    "# df = pd.read_table(f)\n",
    "# df['Total reads'] = np.nan\n",
    "# df['Mapped reads'] = np.nan\n",
    "# df['Mapping rate'] = np.nan\n",
    "# df['Filtered reads'] = np.nan\n",
    "# df['Filtered rate'] = np.nan\n",
    "# df['% Nonclonal rate'] = np.nan\n",
    "# df['Estimated mCG/CG'] = np.nan\n",
    "# df['Estimated mCH/CH'] = np.nan\n",
    "\n",
    "# print(df.shape)\n",
    "\n",
    "# x = df[df2.columns.tolist()[:-1] \n",
    "#         + ['mCCC', 'CCC', 'mCH', 'CH', 'mCG', 'CG', \n",
    "#            'projection', 'source region', 'source slides', 'target region', 'gender']]\n",
    "# output = '/cndd/Public_Datasets/CEMBA/snmCSeq/References/Mapping_summary/mapping_summary_CEMBA_RS2_18Q1.tsv'\n",
    "# x.to_csv(output, sep='\\t', na_rep='NA', header=True, index=False)\n",
    "\n",
    "# print(x.shape)\n",
    "# x.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1535, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Total reads</th>\n",
       "      <th>Mapped reads</th>\n",
       "      <th>Mapping rate</th>\n",
       "      <th>Nonclonal reads</th>\n",
       "      <th>% Nonclonal rate</th>\n",
       "      <th>Filtered reads</th>\n",
       "      <th>Filtered rate</th>\n",
       "      <th>Lambda mC/C</th>\n",
       "      <th>mCCC/CCC</th>\n",
       "      <th>...</th>\n",
       "      <th>CCC</th>\n",
       "      <th>mCH</th>\n",
       "      <th>CH</th>\n",
       "      <th>mCG</th>\n",
       "      <th>CG</th>\n",
       "      <th>projection</th>\n",
       "      <th>source region</th>\n",
       "      <th>source slides</th>\n",
       "      <th>target region</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1069475</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005129</td>\n",
       "      <td>...</td>\n",
       "      <td>1385553</td>\n",
       "      <td>478280</td>\n",
       "      <td>20427236</td>\n",
       "      <td>648736</td>\n",
       "      <td>868511</td>\n",
       "      <td>Pm6B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>6B</td>\n",
       "      <td>STR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1127711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.005936</td>\n",
       "      <td>...</td>\n",
       "      <td>1439820</td>\n",
       "      <td>573022</td>\n",
       "      <td>21370107</td>\n",
       "      <td>703418</td>\n",
       "      <td>898142</td>\n",
       "      <td>Pm7B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>7B</td>\n",
       "      <td>STR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1149770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006778</td>\n",
       "      <td>...</td>\n",
       "      <td>1463330</td>\n",
       "      <td>699355</td>\n",
       "      <td>21736617</td>\n",
       "      <td>700924</td>\n",
       "      <td>908103</td>\n",
       "      <td>Pf6B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>6B</td>\n",
       "      <td>STR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>965382</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.004047</td>\n",
       "      <td>...</td>\n",
       "      <td>1246244</td>\n",
       "      <td>301371</td>\n",
       "      <td>18138018</td>\n",
       "      <td>574039</td>\n",
       "      <td>751098</td>\n",
       "      <td>Pf7B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>6B</td>\n",
       "      <td>STR</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>714865</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005094</td>\n",
       "      <td>...</td>\n",
       "      <td>928585</td>\n",
       "      <td>290528</td>\n",
       "      <td>13583684</td>\n",
       "      <td>443955</td>\n",
       "      <td>577062</td>\n",
       "      <td>Pm6B</td>\n",
       "      <td>SSp</td>\n",
       "      <td>6B</td>\n",
       "      <td>STR</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Sample  Total reads  \\\n",
       "0  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "1  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "2  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "3  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "4  180221_CEMBA_mm_P56_P63_RS2_180221_P1_180221_P...          NaN   \n",
       "\n",
       "   Mapped reads  Mapping rate  Nonclonal reads  % Nonclonal rate  \\\n",
       "0           NaN           NaN          1069475               NaN   \n",
       "1           NaN           NaN          1127711               NaN   \n",
       "2           NaN           NaN          1149770               NaN   \n",
       "3           NaN           NaN           965382               NaN   \n",
       "4           NaN           NaN           714865               NaN   \n",
       "\n",
       "   Filtered reads  Filtered rate  Lambda mC/C  mCCC/CCC   ...        CCC  \\\n",
       "0             NaN            NaN          0.0  0.005129   ...    1385553   \n",
       "1             NaN            NaN          NaN  0.005936   ...    1439820   \n",
       "2             NaN            NaN          0.0  0.006778   ...    1463330   \n",
       "3             NaN            NaN          NaN  0.004047   ...    1246244   \n",
       "4             NaN            NaN          0.0  0.005094   ...     928585   \n",
       "\n",
       "      mCH        CH     mCG      CG  projection  source region  source slides  \\\n",
       "0  478280  20427236  648736  868511        Pm6B            SSp             6B   \n",
       "1  573022  21370107  703418  898142        Pm7B            SSp             7B   \n",
       "2  699355  21736617  700924  908103        Pf6B            SSp             6B   \n",
       "3  301371  18138018  574039  751098        Pf7B            SSp             6B   \n",
       "4  290528  13583684  443955  577062        Pm6B            SSp             6B   \n",
       "\n",
       "   target region  gender  \n",
       "0            STR    male  \n",
       "1            STR    male  \n",
       "2            STR  female  \n",
       "3            STR  female  \n",
       "4            STR    male  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/cndd/Public_Datasets/CEMBA/snmCSeq/References/Mapping_summary/mapping_summary_CEMBA_RS2_18Q1.tsv'\n",
    "df = pd.read_table(f)\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50\n",
      "58\n",
      "59\n",
      "57\n",
      "85\n",
      "64\n",
      "54\n",
      "59\n",
      "58\n",
      "61\n",
      "125\n",
      "61\n",
      "51\n",
      "47\n",
      "63\n",
      "39\n",
      "89\n",
      "58\n",
      "57\n",
      "55\n",
      "52\n",
      "62\n",
      "121\n",
      "50\n"
     ]
    }
   ],
   "source": [
    "# create dataset dir and split mapping summary\n",
    "for rs2_code, df_sub in df.groupby('projection'):\n",
    "    dataset = 'CEMBA_RS2_{}'.format(rs2_code)\n",
    "#     # create dataset dir\n",
    "#     os.makedirs(os.path.join(PATH_DATASETS, dataset))\n",
    "#     # copy metadata over\n",
    "#     df_sub = df_sub[df2.columns.values[:-1]]\n",
    "#     df_sub.to_csv(os.path.join(PATH_DATASETS, dataset, 'mapping_summary_{}.tsv'.format(dataset)), \n",
    "#                   sep='\\t', na_rep='NA', header=True, index=False)\n",
    "#     print(dataset)\n",
    "    print(df_sub.shape[0])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# distribute files over \n",
    "dataset_origin = 'CEMBA_RS2_repo18q1'\n",
    "for rs2_code, df_sub in df.groupby('projection'):\n",
    "    # dataset\n",
    "    dataset = 'CEMBA_RS2_{}'.format(rs2_code)\n",
    "    \n",
    "    # create allc dir\n",
    "    if not os.path.isdir(os.path.join(PATH_DATASETS, dataset, 'allc')): \n",
    "        os.makedirs(os.path.join(PATH_DATASETS, dataset, 'allc'))\n",
    "        \n",
    "    # cells in the dataset\n",
    "    cells = df_sub['Sample'].values\n",
    "    \n",
    "    # allc\n",
    "    allc_srcs = [os.path.join(PATH_DATASETS, dataset_origin, 'allc', 'allc_{}.tsv.bgz'.format(cell))\n",
    "                 for cell in cells]\n",
    "    allc_dsts = [os.path.join(PATH_DATASETS, dataset, 'allc', \n",
    "                              os.path.basename(allc_src)) for allc_src in allc_srcs] \n",
    "        \n",
    "    for src, dst in zip(allc_srcs, allc_dsts):\n",
    "        os.rename(src, dst)\n",
    "        \n",
    "    # allc tabix\n",
    "    allc_tabix_srcs = [os.path.join(PATH_DATASETS, dataset_origin, 'allc', 'allc_{}.tsv.bgz.tbi'.format(cell))\n",
    "                 for cell in cells]\n",
    "    allc_tabix_dsts = [os.path.join(PATH_DATASETS, dataset, 'allc', \n",
    "                              os.path.basename(allc_tabix_src)) for allc_tabix_src in allc_tabix_srcs] \n",
    "    for src, dst in zip(allc_tabix_srcs, allc_tabix_dsts):\n",
    "        os.rename(src, dst)\n",
    "\n",
    "        \n",
    "#     # create binc dir\n",
    "#     if not os.path.isdir(os.path.join(PATH_DATASETS, dataset, 'binc')): \n",
    "#         os.makedirs(os.path.join(PATH_DATASETS, dataset, 'binc'))\n",
    "        \n",
    "#     # create gene_level dir\n",
    "#     if not os.path.isdir(os.path.join(PATH_DATASETS, dataset, 'gene_level')): \n",
    "#         os.makedirs(os.path.join(PATH_DATASETS, dataset, 'gene_level'))\n",
    "\n",
    "    \n",
    "#     binc_srcs = [os.path.join(PATH_DATASETS, dataset_origin, 'binc', 'binc_{}_10000.tsv.bgz'.format(cell))\n",
    "#                  for cell in cells]\n",
    "#     binc_dsts = [os.path.join(PATH_DATASETS, dataset, 'binc', \n",
    "#                               os.path.basename(binc_src)) for binc_src in binc_srcs] \n",
    "    \n",
    "#     genebody_srcs = [os.path.join(PATH_DATASETS, dataset_origin, 'gene_level', 'genebody_{}.tsv.bgz'.format(cell))\n",
    "#                  for cell in cells]\n",
    "#     genebody_dsts = [os.path.join(PATH_DATASETS, dataset, 'gene_level', \n",
    "#                               os.path.basename(genebody_src)) for genebody_src in genebody_srcs] \n",
    "        \n",
    "#     print(binc_srcs[:1]) \n",
    "#     print(binc_dsts[:1]) \n",
    "#     print(genebody_srcs[:1]) \n",
    "#     print(genebody_dsts[:1]) \n",
    "\n",
    "\n",
    "#     for src, dst in zip(binc_srcs, binc_dsts):\n",
    "#         try:\n",
    "#             os.rename(src, dst)\n",
    "#         except:\n",
    "#             pass\n",
    "#     for src, dst in zip(genebody_srcs, genebody_dsts):\n",
    "#         try:\n",
    "#             os.rename(src, dst)\n",
    "#         except:\n",
    "#             pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CEMBA_RS2_Bm3C_rep1 91 All good!\n",
      "CEMBA_RS2_Bm3C_rep2 50 All good!\n",
      "CEMBA_RS2_Bm4B_rep1 89 All good!\n",
      "CEMBA_RS2_Bm4B_rep2 83 All good!\n",
      "CEMBA_RS2_Pf10A 50 All good!\n",
      "CEMBA_RS2_Pf10C 58 All good!\n",
      "CEMBA_RS2_Pf11B 59 All good!\n",
      "CEMBA_RS2_Pf12B 57 All good!\n",
      "CEMBA_RS2_Pf3C 96 All good!\n",
      "CEMBA_RS2_Pf3D 85 All good!\n",
      "CEMBA_RS2_Pf4A 64 All good!\n",
      "CEMBA_RS2_Pf4B 95 All good!\n",
      "CEMBA_RS2_Pf5A 54 All good!\n",
      "CEMBA_RS2_Pf6B 59 All good!\n",
      "CEMBA_RS2_Pf7B 58 All good!\n",
      "CEMBA_RS2_Pf9A 61 All good!\n",
      "CEMBA_RS2_Pf9B 125 All good!\n",
      "CEMBA_RS2_Pf9D 61 All good!\n",
      "CEMBA_RS2_Pm10A 51 All good!\n",
      "CEMBA_RS2_Pm10C 47 All good!\n",
      "CEMBA_RS2_Pm11B 63 All good!\n",
      "CEMBA_RS2_Pm12B 39 All good!\n",
      "CEMBA_RS2_Pm3C 82 All good!\n",
      "CEMBA_RS2_Pm3D 89 All good!\n",
      "CEMBA_RS2_Pm4A 58 All good!\n",
      "CEMBA_RS2_Pm4B 92 All good!\n",
      "CEMBA_RS2_Pm5A 57 All good!\n",
      "CEMBA_RS2_Pm6B 55 All good!\n",
      "CEMBA_RS2_Pm7B 52 All good!\n",
      "CEMBA_RS2_Pm9A 62 All good!\n",
      "CEMBA_RS2_Pm9B 121 All good!\n",
      "CEMBA_RS2_Pm9D 50 All good!\n",
      "CEMBA_RS2_Tf3C 55 All good!\n",
      "CEMBA_RS2_Tf4B 92 All good!\n",
      "CEMBA_RS2_Tf6B 85 All good!\n",
      "CEMBA_RS2_Tf7B 82 All good!\n",
      "CEMBA_RS2_Tm3C 82 All good!\n",
      "CEMBA_RS2_Tm4B 86 All good!\n",
      "CEMBA_RS2_Tm6B 89 All good!\n",
      "CEMBA_RS2_Tm7B 81 All good!\n",
      "CEMBA_RS2_repo18q1 Error in opening mapping summary\n"
     ]
    }
   ],
   "source": [
    "# check if all datasets are good dataset\n",
    "with snmcseq_utils.cd(PATH_DATASETS):\n",
    "    datasets = sorted(glob.glob('*'))\n",
    "\n",
    "# regions = ['2A', '2C', '3B', '3D', '5B', '5D', '7B']\n",
    " \n",
    "for dataset in datasets:\n",
    "    if \"RS2\" in dataset:\n",
    "#     if len(dataset)==(5+2+6+2) and dataset.split('_')[1] in regions: \n",
    "        with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset)):\n",
    "            try:\n",
    "                df_meta = pd.read_table('mapping_summary_{}.tsv'.format(dataset))\n",
    "                n_cells = df_meta.shape[0]\n",
    "            except:\n",
    "                print(dataset, \"Error in opening mapping summary\")\n",
    "                continue\n",
    "\n",
    "            n_allc = len(glob.glob('./allc/allc_*.tsv.bgz'))\n",
    "            n_allc_tbi = len(glob.glob('./allc/allc_*.tsv.bgz.tbi'))\n",
    "\n",
    "            n_binc = len(glob.glob('./binc/binc_*.tsv.bgz'))\n",
    "            n_genebody = len(glob.glob('./gene_level/genebody_*.tsv.bgz'))\n",
    "\n",
    "            flag = 0\n",
    "            if n_cells != n_allc:\n",
    "                print(dataset, n_cells, 'Error allc! ({})'.format(n_allc))\n",
    "                flag += 1\n",
    "            if n_cells != n_allc_tbi:\n",
    "                print(dataset, n_cells, 'Error allc tabix! ({})'.format(n_allc_tbi))\n",
    "                flag += 1\n",
    "            if n_cells != n_binc:\n",
    "                print(dataset, n_cells, 'Error binc! ({})'.format(n_binc))\n",
    "                flag += 1\n",
    "            if n_cells != n_genebody:\n",
    "                print(dataset, n_cells, 'Error genebody! ({})'.format(n_genebody))\n",
    "                flag += 1\n",
    "\n",
    "            if flag == 0:\n",
    "                pass\n",
    "                print(dataset, n_cells, \"All good!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**\n",
      "**\n",
      "**\n",
      "180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_G11_AD004\n",
      "180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_H10_AD006\n"
     ]
    }
   ],
   "source": [
    "dataset = 'CEMBA_RS2_Pm9B'\n",
    "\n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'allc')):\n",
    "    allc_files = glob.glob('*.bgz')\n",
    "    cells_allc = [allc_file[len('allc_'):-len('.tsv.bgz')] for allc_file in allc_files]\n",
    "    \n",
    "f = '/cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/{0}/mapping_summary_{0}.tsv'.format(dataset)\n",
    "cells = pd.read_table(f)['Sample'].values\n",
    "\n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'binc')):\n",
    "    files = glob.glob('*.bgz')\n",
    "    cells_binc = [file[len('binc_'):-len('_10000.tsv.bgz')] for file in files]\n",
    "    \n",
    "for cell_allc in cells_allc:\n",
    "    if cell_allc not in cells:\n",
    "        print(cell_allc)\n",
    "\n",
    "print('**')\n",
    "\n",
    "for cell in cells:\n",
    "    if cell not in cells_allc:\n",
    "        print(cell)\n",
    "        \n",
    "print('**')\n",
    "\n",
    "for cell_binc in cells_binc:\n",
    "    if cell_binc not in cells:\n",
    "        print(cell_binc)\n",
    "\n",
    "print('**')\n",
    "\n",
    "for cell in cells:\n",
    "    if cell not in cells_binc:\n",
    "        print(cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dataset = 'CEMBA_4C_180417'\n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'allc')):\n",
    "    allc_files = glob.glob('*.bgz')\n",
    "    cells = [allc_file[len('allc_'):-len('.tsv.bgz')] for allc_file in allc_files]\n",
    "    \n",
    "    \n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'gene_level')):\n",
    "    files = glob.glob('*')\n",
    "    cells_redo = []\n",
    "    for cell in cells:\n",
    "        if 'genebody_'+cell+'.tsv.bgz' not in files:\n",
    "            cells_redo.append(cell)\n",
    "cells_redo "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:44:40 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      "08/15/2018 11:45:41 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed.tsv\n",
      "08/15/2018 11:45:41 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_A8_AD002_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:47:18 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed.tsv\n",
      "08/15/2018 11:47:18 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_B10_AD002_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:48:22 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed.tsv\n",
      "08/15/2018 11:48:22 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_F7_AD001_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:49:31 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed.tsv\n",
      "08/15/2018 11:49:32 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_C1_AD010_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:50:41 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed.tsv\n",
      "08/15/2018 11:50:41 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_C8_AD001_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:51:40 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed.tsv\n",
      "08/15/2018 11:51:41 PM mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_7_CEMBA180417_4C_8_B4_AD006_indexed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/15/2018 11:52:59 PM Done with mc_gene_level processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/allc/allc_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed.tsv.bgz ['CH', 'CG', 'CA']\n",
      " Saving results to /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_4C_180417/gene_level/genebody_180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed.tsv\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180605_CEMBA_mm_P56_P63_4C_CEMBA180417_4C_5_CEMBA180417_4C_6_D7_AD002_indexed\n"
     ]
    }
   ],
   "source": [
    "log = snmcseq_utils.create_logger()\n",
    "\n",
    "for cell in cells_redo:\n",
    "    allc_file = os.path.join(PATH_DATASETS, dataset, 'allc', 'allc_' + cell + '.tsv.bgz')\n",
    "\n",
    "    try:\n",
    "        CEMBA_mc_gene_level.mc_gene_level(allc_file,\n",
    "            contexts=CONTEXTS, \n",
    "            genebody=GENEBODY,\n",
    "            convention='CEMBA', \n",
    "            overwrite=True)\n",
    "        print(cell)\n",
    "    except:\n",
    "        pass\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_G11_AD004',\n",
       " '180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_H10_AD006']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dataset = 'CEMBA_4F_180329'\n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'allc')):\n",
    "    allc_files = glob.glob('*.bgz')\n",
    "    mells = [allc_file[len('allc_'):-len('.tsv.bgz')] for allc_file in allc_files]\n",
    "    \n",
    "with snmcseq_utils.cd(os.path.join(PATH_DATASETS, dataset, 'binc')):\n",
    "    files = glob.glob('*')\n",
    "    cells_redo = []\n",
    "    for cell in cells:\n",
    "        if 'binc_'+cell+'_10000.tsv.bgz' not in files:\n",
    "            cells_redo.append(cell)\n",
    "cells_redo "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "02/26/2019 05:20:16 PM binc processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_RS2_Pm9B/allc/allc_180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_G11_AD004.tsv.bgz ['CH', 'CG', 'CA']\n",
      "02/26/2019 05:24:44 PM Done with binc processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_RS2_Pm9B/allc/allc_180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_G11_AD004.tsv.bgz ['CH', 'CG', 'CA']\n",
      "Saving results to: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_RS2_Pm9B/binc/binc_180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_G11_AD004_10000.tsv\n",
      "02/26/2019 05:24:49 PM binc processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_RS2_Pm9B/allc/allc_180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_H10_AD006.tsv.bgz ['CH', 'CG', 'CA']\n",
      "02/26/2019 05:36:06 PM Done with binc processing: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_RS2_Pm9B/allc/allc_180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_H10_AD006.tsv.bgz ['CH', 'CG', 'CA']\n",
      "Saving results to: /cndd/Public_Datasets/CEMBA/snmCSeq/Datasets/CEMBA_RS2_Pm9B/binc/binc_180313_CEMBA_mm_P56_P63_RS2_180313_P5_180313_P6_H10_AD006_10000.tsv\n"
     ]
    }
   ],
   "source": [
    "import CEMBA_bin_allc_files\n",
    "\n",
    "log = snmcseq_utils.create_logger()\n",
    "for cell in cells_redo:\n",
    "    allc_file = os.path.join(PATH_DATASETS, dataset, 'allc', 'allc_' + cell + '.tsv.bgz')\n",
    "    \n",
    "    CEMBA_bin_allc_files.bin_allc(allc_file, \n",
    "        bin_size=BIN_SIZE, \n",
    "        contexts=CONTEXTS,\n",
    "        chromosomes=None, \n",
    "        species='mouse',\n",
    "        compression='gzip',\n",
    "        overwrite=False\n",
    "        )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
