{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Split allc by chromosomes\\n'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Split allc by chromosomes\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess as sp\n",
    "\n",
    "from __init__ import *\n",
    "import snmcseq_utils\n",
    "from snmcseq_utils import cd\n",
    "from snmcseq_utils import create_logger\n",
    "from CEMBA_update_mysql import connect_sql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "log = create_logger()\n",
    "time_stamp = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/07/2018 08:09:16 PM /scratch/tmp_allc_split_1520482155.2838366 created!\n",
      "03/07/2018 08:09:16 PM Splitting allc files to /scratch/tmp_allc_split_1520482155.2838366\n",
      "03/07/2018 08:09:16 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_10_Ens2.tsv...\n",
      "03/07/2018 08:12:38 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_11_Ens2.tsv...\n",
      "03/07/2018 08:16:05 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_12_Ens2.tsv...\n",
      "03/07/2018 08:19:21 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_13_Ens2.tsv...\n",
      "03/07/2018 08:22:32 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_14_Ens2.tsv...\n",
      "03/07/2018 08:25:36 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_15_Ens2.tsv...\n",
      "03/07/2018 08:28:41 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_16_Ens2.tsv...\n",
      "03/07/2018 08:31:42 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_17_Ens2.tsv...\n",
      "03/07/2018 08:34:26 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_18_Ens2.tsv...\n",
      "03/07/2018 08:37:29 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_1_Ens2.tsv...\n",
      "03/07/2018 08:41:20 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_2_Ens2.tsv...\n",
      "03/07/2018 08:45:08 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_3_Ens2.tsv...\n",
      "03/07/2018 08:48:56 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_4_Ens2.tsv...\n",
      "03/07/2018 08:52:36 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_5_Ens2.tsv...\n",
      "03/07/2018 08:56:17 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_6_Ens2.tsv...\n",
      "03/07/2018 08:59:46 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_7_Ens2.tsv...\n",
      "03/07/2018 09:03:20 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_8_Ens2.tsv...\n",
      "03/07/2018 09:06:54 PM Processing allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_9_Ens2.tsv...\n"
     ]
    }
   ],
   "source": [
    "# split allc tables of an ensemble to /scratch/tmp_allc_split\n",
    "# /scratch/tmp_allc_split needs to be removed\n",
    "\n",
    "ens = 'Ens2'\n",
    "\n",
    "output_folder = '/scratch/tmp_allc_split_{}'.format(time_stamp)\n",
    "try:\n",
    "    os.makedirs(output_folder)\n",
    "    logging.info('{} created!'.format(output_folder))\n",
    "except:\n",
    "    raise ValueError\n",
    "\n",
    "logging.info(\"Splitting allc files to {}\".format(output_folder))\n",
    "for allc_file in sorted(glob.glob(os.path.join(PATH_ENSEMBLES, ens, 'allc_merged/allc_merged_*.tsv'))):\n",
    "    # split allc files\n",
    "    logging.info('Processing {}...'.format(os.path.basename(allc_file)))\n",
    "    df = snmcseq_utils.read_allc_CEMBA(allc_file, pindex=False, compression='infer')\n",
    "    for chrom, df_chrom in df.groupby('chr'):\n",
    "        output = '{}/{}_{}.tsv'.format(output_folder, \n",
    "            os.path.basename(allc_file)[:-len('.tsv')], chrom) \n",
    "        df_chrom.to_csv(output, sep=\"\\t\", na_rep='NA', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename splitted allc for further processes (added clst removed merged)\n",
    "allc_split_files = sorted(glob.glob(os.path.join(output_folder, 'allc_merged_*.tsv')))\n",
    "\n",
    "for file in allc_split_files:\n",
    "    src = file\n",
    "    file = os.path.basename(file)\n",
    "    file_l = file.split('_')\n",
    "    file_l[-3] = 'clst' + file_l[-3]\n",
    "    file = '_'.join(file_l)\n",
    "    dst = os.path.join(output_folder, file)\n",
    "    os.rename(src, dst)\n",
    "    \n",
    "allc_split_files = sorted(glob.glob(os.path.join(output_folder, 'allc_merged_*.tsv')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/07/2018 09:10:24 PM Allc to mc files...\n",
      "03/07/2018 09:29:07 PM /scratch/tmp_mc_1520482155.2838366 created!\n"
     ]
    }
   ],
   "source": [
    "# allc to mc\n",
    "# mc files are in /scratch -> /scratch/tmp_mc_split_$time\n",
    "\n",
    "# # ***\n",
    "# logging.info(\"allc to mc...\")\n",
    "# output_folder = '/scratch/tmp_allc_split'\n",
    "# allc_split_files = glob.glob(os.path.join(output_folder, 'allc_*.tsv'))\n",
    "# # ***\n",
    "\n",
    "logging.info(\"Allc to mc files...\")\n",
    "allc_split_files = sorted(glob.glob(os.path.join(output_folder, 'allc_merged_*.tsv')))\n",
    "mc_split_files = ['mc_' + os.path.basename(file)[len('allc_'):] \n",
    "                  for file in allc_split_files]\n",
    "\n",
    "# allc to mc\n",
    "with cd(output_folder):\n",
    "    cmd = '/cndd/fangming/lab_scripts/allc2mc.pl ./allc_merged_*.tsv'\n",
    "    sp.call(cmd, shell=True)\n",
    "\n",
    "# move all relevent mc files to a tmp folder\n",
    "output_folder_mc = '/scratch/tmp_mc_{}'.format(time_stamp) \n",
    "try:\n",
    "    os.makedirs(output_folder_mc)\n",
    "    logging.info('{} created!'.format(output_folder_mc))\n",
    "except:\n",
    "    raise ValueError\n",
    "    \n",
    "for mc_split_file in mc_split_files:\n",
    "    src = os.path.join('/scratch', mc_split_file)\n",
    "    dst = os.path.join(output_folder_mc, mc_split_file)\n",
    "    os.rename(src, dst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/07/2018 09:29:07 PM Mc files to mySQL...\n"
     ]
    }
   ],
   "source": [
    "# mc to mysql \n",
    "\n",
    "# cd output_folder_mc\n",
    "logging.info('Mc files to mySQL...')\n",
    "with cd(output_folder_mc):\n",
    "    cmd = '/cndd/fangming/lab_scripts/load_mysql_mc CEMBA_annoj > /cndd/fangming/CEMBA/snmcseq_dev/logs/annoj_1.log'\n",
    "    sp.call(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # clean up: allc_split, mc\n",
    "\n",
    "# # ***\n",
    "# output_folder = '/scratch/tmp_allc_split'\n",
    "# # ***\n",
    "\n",
    "# cmd = 'rm -r {}'.format(output_folder)\n",
    "# sp.call(cmd, shell=True)\n",
    "# cmd = 'rm -r {}'.format(output_folder_mc)\n",
    "# sp.call(cmd, shell=True)\n",
    "# logging.info(\"Tmp folders removed: {} {}\".format(output_folder, output_folder_mc))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/07/2018 09:38:06 PM DMR to bed files...\n",
      "03/07/2018 09:38:09 PM Created directory: /scratch/tmp_dmr_1520482155.2838366\n",
      "03/07/2018 09:38:10 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst1.bed\n",
      "03/07/2018 09:38:10 PM (29057,)\n",
      "03/07/2018 09:38:11 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst2.bed\n",
      "03/07/2018 09:38:11 PM (30873,)\n",
      "03/07/2018 09:38:12 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst3.bed\n",
      "03/07/2018 09:38:12 PM (25359,)\n",
      "03/07/2018 09:38:13 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst4.bed\n",
      "03/07/2018 09:38:13 PM (13053,)\n",
      "03/07/2018 09:38:14 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst5.bed\n",
      "03/07/2018 09:38:14 PM (4382,)\n",
      "03/07/2018 09:38:14 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst6.bed\n",
      "03/07/2018 09:38:14 PM (6009,)\n",
      "03/07/2018 09:38:15 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst7.bed\n",
      "03/07/2018 09:38:15 PM (8099,)\n",
      "03/07/2018 09:38:16 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst8.bed\n",
      "03/07/2018 09:38:16 PM (2125,)\n",
      "03/07/2018 09:38:17 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst9.bed\n",
      "03/07/2018 09:38:17 PM (2592,)\n",
      "03/07/2018 09:38:17 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst10.bed\n",
      "03/07/2018 09:38:17 PM (1437,)\n",
      "03/07/2018 09:38:18 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst11.bed\n",
      "03/07/2018 09:38:18 PM (3371,)\n",
      "03/07/2018 09:38:19 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst12.bed\n",
      "03/07/2018 09:38:19 PM (1796,)\n",
      "03/07/2018 09:38:20 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst13.bed\n",
      "03/07/2018 09:38:20 PM (2819,)\n",
      "03/07/2018 09:38:21 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst14.bed\n",
      "03/07/2018 09:38:21 PM (1528,)\n",
      "03/07/2018 09:38:21 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst15.bed\n",
      "03/07/2018 09:38:21 PM (3374,)\n",
      "03/07/2018 09:38:22 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst16.bed\n",
      "03/07/2018 09:38:22 PM (2138,)\n",
      "03/07/2018 09:38:23 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst17.bed\n",
      "03/07/2018 09:38:23 PM (323,)\n",
      "03/07/2018 09:38:23 PM Saved to /scratch/tmp_dmr_1520482155.2838366/dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst18.bed\n",
      "03/07/2018 09:38:23 PM (2333,)\n"
     ]
    }
   ],
   "source": [
    "# dmr to bed\n",
    "\n",
    "logging.info(\"DMR to bed files...\")\n",
    "\n",
    "cluster_type = 'cluster_mCHmCG_lv_npc50_k30'\n",
    "annotation_type = 'annotation' + cluster_type[len('cluster'):]\n",
    "engine = connect_sql('CEMBA')\n",
    "output_folder_dmr = '/scratch/tmp_dmr_{}'.format(time_stamp)\n",
    "\n",
    "# cluster annotation\n",
    "sql = \"\"\"SELECT {}, {} FROM {}\"\"\".format(cluster_type, annotation_type, ens)\n",
    "df_info = pd.read_sql(sql, engine)\n",
    "df_info = df_info.sort_values(cluster_type).drop_duplicates().set_index(cluster_type)\n",
    "\n",
    "ens_path = os.path.join(PATH_ENSEMBLES, ens)\n",
    "n_clusters = len(glob.glob(os.path.join(ens_path, 'allc_merged/allc_merged_mCG_{}_*_{}.tsv'.format(cluster_type, ens))))\n",
    "\n",
    "# dmr results\n",
    "input_f = os.path.join(ens_path, 'dmr/dmr_allc_merged_mCG_{}_rms_results_collapsed.tsv'.format(cluster_type))\n",
    "df = pd.read_table(input_f, index_col=['#chr', 'start', 'end'], dtype={'#chr': object})\n",
    "\n",
    "# df_hypo\n",
    "df_hypo = df.loc[(df['number_of_dms']>=3) & (~df['hypomethylated_samples'].isnull()), 'hypomethylated_samples'].apply(\n",
    "        lambda x: x.split(','))\n",
    "# df_hyper = df.loc[(df['number_of_dms']>=3) & (~df['hypermethylated_samples'].isnull()), 'hypermethylated_samples'].apply(lambda x: x.split(','))\n",
    "\n",
    "try:\n",
    "    os.makedirs(output_folder_dmr)\n",
    "    logging.info('Created directory: {}'.format(output_folder_dmr))\n",
    "except:\n",
    "    raise ValueError\n",
    "    \n",
    "for i in range(n_clusters):\n",
    "    df_hypo_cluster = df_hypo.loc[\n",
    "        df_hypo.apply(lambda x: ('merged_mCG_{}_{}_{}'.format(cluster_type, i+1, ens) in x))]\n",
    "    output = os.path.join(output_folder_dmr, 'dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst{}.bed'.format(i+1))\n",
    "    df_hypo_cluster.to_csv(output, sep='\\t', header=False, index=True, na_rep='NA')\n",
    "    logging.info(\"Saved to {}\".format(output))\n",
    "    logging.info(df_hypo_cluster.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename bed\n",
    "with cd(output_folder_dmr):\n",
    "    for file in glob.glob('*'):\n",
    "        src = file \n",
    "        dst = '{}_{}.bed'.format(src[:-len('.bed')], ens).replace('allc_merged_', '') \n",
    "        os.rename(src, dst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/07/2018 09:38:23 PM Uploading bed files (DMR) to mySQL databases...\n"
     ]
    }
   ],
   "source": [
    "# bed to mysql\n",
    "\n",
    "logging.info(\"Uploading bed files (DMR) to mySQL databases...\")\n",
    "bed_files = sorted(glob.glob(os.path.join(output_folder_dmr, '*')))\n",
    "with cd(output_folder_dmr):\n",
    "    for bed_file in bed_files:\n",
    "        # fn, db, table_prefix\n",
    "        cmd = '/cndd/fangming/lab_scripts/bed_load_mysql.sh {} {} {}'.format(os.path.basename(bed_file), \n",
    "                                                                             'CEMBA_annoj', \n",
    "                                                                             os.path.basename(bed_file)[:-len('.bed')])\n",
    "#         print(cmd)\n",
    "        sp.call(cmd, shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # clean up: bed \n",
    "# cmd = 'rm -r {}'.format(output_folder_dmr)\n",
    "# sp.call(cmd, shell=True)\n",
    "# logging.info(\"Tmp folders removed: {}\".format(output_folder_dmr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create php file\n",
    "path = '/var/www/html/annoj_private/CEMBA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "TEMPLATE_MC_PHP = (\n",
    "\"\"\"<?php\n",
    "$append_assembly = true;\n",
    "$table = 'CEMBA_annoj.{}_';\n",
    "$title = '{}';\n",
    "$info = '{}';\n",
    "$link = mysql_connect(\"banjo.ucsd.edu\", \"cndd_annoj\", \"jonna_ddnc\") or die(\"failed\");\n",
    "require_once '../includes/common_wgta.php';\n",
    "?>\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "TEMPLATE_BED_PHP = (\n",
    "\"\"\"<?php\n",
    "$append_assembly = true;\n",
    "$table = 'CEMBA_annoj.{}_';\n",
    "$title = '{}';\n",
    "$info = '{}';\n",
    "$link = mysql_connect(\"banjo.ucsd.edu\", \"cndd_annoj\", \"jonna_ddnc\") or die(\"failed\");\n",
    "require_once '../includes/common_masks.php';\n",
    "?>\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "TEMPLATE_DMR_PHP = TEMPLATE_BED_PHP \n",
    "\n",
    "TEMPLATE_ATAC_PHP = TEMPLATE_ATAC_PHP \n",
    "\n",
    "def gen_php_template(track_name, template_string):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    return template_string.format(track_name, track_name, track_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # check what's in mysql\n",
    "# track_names = ['dmr_hypo_allc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst{}'.format(i+1) for i in range(17)] \n",
    "# for track_name in track_names:\n",
    "#     with open(os.path.join(path, 'browser/fetchers/dmrs/{}.php'.format(track_name)), 'w') as f_php:\n",
    "#         f_php.write(gen_php_template(track_name, TEMPLATE_DMR_PHP))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create php files\n",
    "# create index files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mc_split_file\n",
    "output_folder_php = '/cndd/fangming/CEMBA/snmcseq_dev/browsers/php_mc_{}'.format(ens)\n",
    "if not os.path.isdir(output_folder_php):\n",
    "    os.makedirs(output_folder_php)\n",
    "    logging.info('Created directory: {}'.format(output_folder_php))\n",
    "    \n",
    "track_names_mc = np.unique(['_'.join(os.path.basename(file).split('_')[:-1]) for file in mc_split_files])\n",
    "for track_name in track_names_mc:\n",
    "    with open(os.path.join(output_folder_php, '{}.php'.format(track_name)), 'w') as f_php:\n",
    "        f_php.write(gen_php_template(track_name, TEMPLATE_MC_PHP))\n",
    "\n",
    "# dmr files\n",
    "output_folder_php = '/cndd/fangming/CEMBA/snmcseq_dev/browsers/php_dmrs_{}'.format(ens)\n",
    "if not os.path.isdir(output_folder_php):\n",
    "    os.makedirs(output_folder_php)\n",
    "    logging.info('Created directory: {}'.format(output_folder_php))\n",
    "track_names_dmr = [os.path.basename(file)[:-len('.bed')] for file in bed_files]\n",
    "for track_name in track_names_dmr:\n",
    "    with open(os.path.join(output_folder_php, '{}.php'.format(track_name)), 'w') as f_php:\n",
    "        f_php.write(gen_php_template(track_name, TEMPLATE_DMR_PHP))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# index_file\n",
    "TEMPLATE_INDEX_MC = (\n",
    "\"\"\"\n",
    "    {{\n",
    "      id   : '{}',\n",
    "      name : '{}',\n",
    "      type : 'MethTrack',\n",
    "      path : 'DNA Methylation',\n",
    "      data : './browser/fetchers/mc/{}.php',\n",
    "      iconCls : 'salk_meth',\n",
    "      height : 30,\n",
    "      class : 'CG -CHH -CHG',\n",
    "      showControls : true\n",
    "    }},\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "TEMPLATE_INDEX_DMR = (\n",
    "\"\"\"\n",
    "    {{\n",
    "      id : '{}',\n",
    "      name : '{}',\n",
    "      type : 'ReadsTrack',\n",
    "      path : 'DMR',\n",
    "      data : './browser/fetchers/dmrs/{}.php',\n",
    "      iconCls : 'salk_dmr',\n",
    "      height : 5,\n",
    "      scale : 100,\n",
    "      color : {{read : '000099'}},\n",
    "      showControls : 0,\n",
    "      single : true,\n",
    "    }},\n",
    "\"\"\"\n",
    ")\n",
    "TEMPLATE_INDEX_ATAC = (\n",
    "\"\"\"\n",
    "      {{\n",
    "        id : '{}',\n",
    "        name : '{}',\n",
    "        type : 'PairedEndTrack',\n",
    "        path : 'ATAC-Seq',\n",
    "        data : './browser/fetchers/atac/{}.php',\n",
    "        iconCls : 'silk_bricks',\n",
    "        height : 20,\n",
    "        scale : 2,\n",
    "        single : true,\n",
    "      }},\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "def gen_annoj_config(track_name, track_vis_name, template_string):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    return template_string.format(track_name, track_vis_name, track_name)\n",
    "\n",
    "def gen_annoj_configs(track_names, track_vis_names, template_string):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    string = ''\n",
    "    for track_name, track_vis_name in zip(track_names, track_vis_names):\n",
    "        string += template_string.format(track_name, track_vis_name, track_name)\n",
    "    return string\n",
    "\n",
    "def gen_active_configs(lst):\n",
    "    string = ''\n",
    "    for item in lst:\n",
    "        string +=\"\\t'{}',\\n\".format(item)\n",
    "    return string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": [
     2,
     96
    ]
   },
   "outputs": [],
   "source": [
    "TEMPLATE_INDEX_HTML=(\n",
    "\"\"\"<html>\n",
    "<head>\n",
    "  <meta http-equiv='Content-Type' content='text/html; charset=iso-8859-1'>\n",
    "  <title>{}</title>\n",
    "\n",
    "\n",
    "  <link type='text/css' rel='stylesheet' href='aj2/aj2_css_dev/ext-all.css' />\n",
    "  <script type='text/javascript' src='aj2/js/ext-base-3.2.js'></script>\n",
    "  <script type='text/javascript' src='aj2/js/ext-all-3.2.js'></script>\n",
    "\n",
    "  <link type='text/css' rel='stylesheet' href='aj2/aj2_css_dev/viewport.css' />\n",
    "  <link type='text/css' rel='stylesheet' href='aj2/aj2_css_dev/plugins.css' />\n",
    "  <link type='text/css' rel='stylesheet' href='aj2/aj2_css_dev/salk.css' />\n",
    "  <script type='text/javascript' src='aj2/js/excanvas.js'></script>\n",
    "  <script type='text/javascript' src='aj2/js/aj-cndd1-dev.js'></script>\n",
    "\n",
    "  <!-- Favicon -->\n",
    "  <link rel=\"icon\" href=\"/var/www/html/annoj/browser/aj.ico\" type=\"image/x-icon\">\n",
    "  <link rel=\"shortcut icon\" href=\"/var/www/html/annoj/browser/aj.ico\" type=\"image/x-icon\">\n",
    "\n",
    "\n",
    "  <!-- Config -->\n",
    "  <script type='text/javascript'>\n",
    "\n",
    "  AnnoJ.config = {{\n",
    "\n",
    "    info : {{\n",
    "      title  : '{}',\n",
    "      genome  : 'mm10',\n",
    "      contact  : '',\n",
    "      email : '',\n",
    "      institution : ''\n",
    "    }},\n",
    "    tracks : [\n",
    "      //Models\n",
    "      {{\n",
    "        id   : 'gene_model_mm10',\n",
    "        name : 'Gene Models (mm10)',\n",
    "        type : 'ModelsTrack',\n",
    "        path : 'Annotation models',\n",
    "        data : './browser/fetchers/models/genes_mm10.php',\n",
    "        height : 100,\n",
    "        showControls : true\n",
    "      }},\n",
    "\n",
    "\n",
    "      // DNA methylation\n",
    "      {}\n",
    "      // DMRs\n",
    "      {}\n",
    "\n",
    "\n",
    "    ],\n",
    "\n",
    "    active : [// *** Gene models\n",
    "      'gene_model_mm10',\n",
    "      {}\n",
    "\n",
    "    ],\n",
    "    genome    : './browser/fetchers/mus_musculus.php',\n",
    "    bookmarks : './browser/fetchers/mus_musculus.php',\n",
    "    stylesheets : [\n",
    "      {{\n",
    "        id   : 'css1',\n",
    "        name : 'Plugins CSS',\n",
    "        href : 'css/plugins.css',\n",
    "        active : true\n",
    "      }},{{\n",
    "        id   : 'css2',\n",
    "        name : 'SALK CSS',\n",
    "        href : 'css/salk.css',\n",
    "        active : true\n",
    "      }}\n",
    "    ],\n",
    "    location : {{\n",
    "      assembly : '1',\n",
    "      position : '132487763',\n",
    "      bases    : 300,\n",
    "      pixels   : 1\n",
    "    }},\n",
    "    admin : {{\n",
    "      name  : 'Eran Mukamel',\n",
    "      email : 'emukamel@ucsd.edu',\n",
    "      notes : 'University of California, San Diego'\n",
    "    }},\n",
    "  }};\n",
    "\n",
    "\n",
    "\n",
    "  </script>\n",
    "\n",
    "\n",
    "\n",
    "</head>\n",
    "\n",
    "<body>\n",
    "\n",
    "  <noscript>\n",
    "    <table id='noscript'><tr>\n",
    "      <td><img src='hs/img/Anno-J.jpg' /></td>\n",
    "      <td>\n",
    "        <p>Anno-J cannot run because your browser is currently configured to block Javascript.</p>\n",
    "        <p>To use the application please access your browser settings or preferences, turn Javascript support back on, and then refresh this page.</p>\n",
    "        <p>Thankyou, and enjoy the application!<br /></p>\n",
    "      </td>\n",
    "    </tr></table>\n",
    "  </noscript>\n",
    "  \n",
    "  <!-- Enable URL queries -->\n",
    "  <script type='text/javascript'> var queryPost; </script>\n",
    "  <script type='text/javascript' src='./browser/js/urlinit.js'></script>\n",
    "  \n",
    "  <!-- Google Analytics -->\n",
    "  <script type=\"text/javascript\">\n",
    "  var gaJsHost = ((\"https:\" == document.location.protocol) ? \"https://ssl.\" : \"http://www.\");\n",
    "  document.write(unescape(\"%3Cscript src='\" + gaJsHost + \"google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E\"));\n",
    "  </script>\n",
    "\n",
    "  <script type=\"text/javascript\">\n",
    "  var pageTracker = _gat._getTracker(\"UA-4150298-1\");\n",
    "  pageTracker._initData();\n",
    "  pageTracker._trackPageview();\n",
    "  </script>\n",
    "  <script type=\"text/javascript\">\n",
    "  var gaJsHost = ((\"https:\" == document.location.protocol) ? \"https://ssl.\" : \"http://www.\");\n",
    "  document.write(unescape(\"%3Cscript src='\" + gaJsHost + \"google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E\"));\n",
    "  </script>\n",
    "  <script type=\"text/javascript\">\n",
    "  try {{\n",
    "    var pageTracker = _gat._getTracker(\"UA-4150298-1\");\n",
    "    pageTracker._trackPageview();\n",
    "  }} catch(err) {{}}</script>\n",
    "</body>\n",
    "\n",
    "</html>\n",
    "\"\"\" # .format(title, title, mc, dmr, active)\n",
    ")\n",
    "\n",
    "def gen_index_html(template_string, title, mc, dmr, active):\n",
    "    return (template_string.format(title, title, mc, dmr, active))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "track_vis_names_mc = [file.replace('merged_', '').replace('mc_', '') for file in track_names_mc]\n",
    "track_vis_names_dmr = track_names_dmr \n",
    "\n",
    "title = 'CEMBA browser ({})'.format(ens)\n",
    "mc_configs = gen_annoj_configs(track_names_mc, track_vis_names_mc, TEMPLATE_INDEX_MC) \n",
    "dmr_configs = gen_annoj_configs(track_names_dmr, track_vis_names_dmr, TEMPLATE_INDEX_DMR) \n",
    "\n",
    "# active order (natsorted mc track and dmr tracks)\n",
    "from natsort import natsorted\n",
    "active_orders = np.vstack((np.asarray(natsorted(track_names_mc)), \n",
    "                           np.asarray(natsorted(track_names_dmr)))).flatten('F')\n",
    "active_configs = gen_active_configs(active_orders) \n",
    "\n",
    "index_html = gen_index_html(TEMPLATE_INDEX_HTML, title, mc_configs, dmr_configs, active_configs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_html = '/cndd/fangming/CEMBA/snmcseq_dev/browsers/index_{}.html'.format(ens)\n",
    "with open(output_html, 'w') as f:\n",
    "    f.write(index_html)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<html>\n",
      "<head>\n",
      "  <meta http-equiv='Content-Type' content='text/html; charset=iso-8859-1'>\n",
      "  <title>CEMBA browser (Ens2)</title>\n",
      "\n",
      "\n",
      "  <link type='text/css' rel='stylesheet' href='aj2/aj2_css_dev/ext-all.css' />\n",
      "  <script type='text/javascript' src='aj2/js/ext-base-3.2.js'></script>\n",
      "  <script type='text/javascript' src='aj2/js/ext-all-3.2.js'></script>\n",
      "\n",
      "  <link type='text/css' rel='stylesheet' href='aj2/aj2_css_dev/viewport.css' />\n",
      "  <link type='text/css' rel='stylesheet' href='aj2/aj2_css_dev/plugins.css' />\n",
      "  <link type='text/css' rel='stylesheet' href='aj2/aj2_css_dev/salk.css' />\n",
      "  <script type='text/javascript' src='aj2/js/excanvas.js'></script>\n",
      "  <script type='text/javascript' src='aj2/js/aj-cndd1-dev.js'></script>\n",
      "\n",
      "  <!-- Favicon -->\n",
      "  <link rel=\"icon\" href=\"/var/www/html/annoj/browser/aj.ico\" type=\"image/x-icon\">\n",
      "  <link rel=\"shortcut icon\" href=\"/var/www/html/annoj/browser/aj.ico\" type=\"image/x-icon\">\n",
      "\n",
      "\n",
      "  <!-- Config -->\n",
      "  <script type='text/javascript'>\n",
      "\n",
      "  AnnoJ.config = {\n",
      "\n",
      "    info : {\n",
      "      title  : 'CEMBA browser (Ens2)',\n",
      "      genome  : 'mm10',\n",
      "      contact  : '',\n",
      "      email : '',\n",
      "      institution : ''\n",
      "    },\n",
      "    tracks : [\n",
      "      //Models\n",
      "      {\n",
      "        id   : 'gene_model_mm10',\n",
      "        name : 'Gene Models (mm10)',\n",
      "        type : 'ModelsTrack',\n",
      "        path : 'Annotation models',\n",
      "        data : './browser/fetchers/models/genes_mm10.php',\n",
      "        height : 100,\n",
      "        showControls : true\n",
      "      },\n",
      "\n",
      "\n",
      "      // DNA methylation\n",
      "      \n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id   : 'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens2',\n",
      "      name : 'mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens2',\n",
      "      type : 'MethTrack',\n",
      "      path : 'DNA Methylation',\n",
      "      data : './browser/fetchers/mc/mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens2.php',\n",
      "      iconCls : 'salk_meth',\n",
      "      height : 30,\n",
      "      class : 'CG -CHH -CHG',\n",
      "      showControls : true\n",
      "    },\n",
      "\n",
      "      // DMRs\n",
      "      \n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "    {\n",
      "      id : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens2',\n",
      "      name : 'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens2',\n",
      "      type : 'ReadsTrack',\n",
      "      path : 'DMR',\n",
      "      data : './browser/fetchers/dmrs/dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens2.php',\n",
      "      iconCls : 'salk_dmr',\n",
      "      height : 5,\n",
      "      scale : 100,\n",
      "      color : {read : '000099'},\n",
      "      showControls : 0,\n",
      "      single : true,\n",
      "    },\n",
      "\n",
      "\n",
      "\n",
      "    ],\n",
      "\n",
      "    active : [// *** Gene models\n",
      "      'gene_model_mm10',\n",
      "      \t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens2',\n",
      "\t'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens2',\n",
      "\t'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens2',\n",
      "\n",
      "\n",
      "    ],\n",
      "    genome    : './browser/fetchers/mus_musculus.php',\n",
      "    bookmarks : './browser/fetchers/mus_musculus.php',\n",
      "    stylesheets : [\n",
      "      {\n",
      "        id   : 'css1',\n",
      "        name : 'Plugins CSS',\n",
      "        href : 'css/plugins.css',\n",
      "        active : true\n",
      "      },{\n",
      "        id   : 'css2',\n",
      "        name : 'SALK CSS',\n",
      "        href : 'css/salk.css',\n",
      "        active : true\n",
      "      }\n",
      "    ],\n",
      "    location : {\n",
      "      assembly : '1',\n",
      "      position : '132487763',\n",
      "      bases    : 300,\n",
      "      pixels   : 1\n",
      "    },\n",
      "    admin : {\n",
      "      name  : 'Eran Mukamel',\n",
      "      email : 'emukamel@ucsd.edu',\n",
      "      notes : 'University of California, San Diego'\n",
      "    },\n",
      "  };\n",
      "\n",
      "\n",
      "\n",
      "  </script>\n",
      "\n",
      "\n",
      "\n",
      "</head>\n",
      "\n",
      "<body>\n",
      "\n",
      "  <noscript>\n",
      "    <table id='noscript'><tr>\n",
      "      <td><img src='hs/img/Anno-J.jpg' /></td>\n",
      "      <td>\n",
      "        <p>Anno-J cannot run because your browser is currently configured to block Javascript.</p>\n",
      "        <p>To use the application please access your browser settings or preferences, turn Javascript support back on, and then refresh this page.</p>\n",
      "        <p>Thankyou, and enjoy the application!<br /></p>\n",
      "      </td>\n",
      "    </tr></table>\n",
      "  </noscript>\n",
      "\n",
      "  <!-- Google Analytics -->\n",
      "  <script type=\"text/javascript\">\n",
      "  var gaJsHost = ((\"https:\" == document.location.protocol) ? \"https://ssl.\" : \"http://www.\");\n",
      "  document.write(unescape(\"%3Cscript src='\" + gaJsHost + \"google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E\"));\n",
      "  </script>\n",
      "\n",
      "  <script type=\"text/javascript\">\n",
      "  var pageTracker = _gat._getTracker(\"UA-4150298-1\");\n",
      "  pageTracker._initData();\n",
      "  pageTracker._trackPageview();\n",
      "  </script>\n",
      "  <script type=\"text/javascript\">\n",
      "  var gaJsHost = ((\"https:\" == document.location.protocol) ? \"https://ssl.\" : \"http://www.\");\n",
      "  document.write(unescape(\"%3Cscript src='\" + gaJsHost + \"google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E\"));\n",
      "  </script>\n",
      "  <script type=\"text/javascript\">\n",
      "  try {\n",
      "    var pageTracker = _gat._getTracker(\"UA-4150298-1\");\n",
      "    pageTracker._trackPageview();\n",
      "  } catch(err) {}</script>\n",
      "</body>\n",
      "\n",
      "</html>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(index_html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CEMBA_3C_171206_merged_cluster_C1'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lump together\n",
    "os.path.splitext('CEMBA_3C_171206_merged.cluster.C1.bed')[0].replace('.', '_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "04/06/2018 10:36:18 AM Uploading bed files (ATAC readq) to mySQL databases...\n"
     ]
    }
   ],
   "source": [
    "# bed to mysql # adding atac readend tracks\n",
    "\n",
    "logging.info(\"Uploading bed files (ATAC readq) to mySQL databases...\")\n",
    "dirc = '/cndd/projects/Public_Datasets/CEMBA/snATACSeq/Datasets/CEMBA_3C_171206/uniq_chrs_readends'\n",
    "table_prefixes = []\n",
    "with cd(dirc):\n",
    "    bed_files = sorted(glob.glob('CEMBA_*_merged.cluster.C*.bed'))\n",
    "    for bed_file in bed_files:\n",
    "        # fn, db, table_prefix\n",
    "        table_prefix = os.path.splitext(bed_file)[0].replace('.', '_')\n",
    "        table_prefixes.append(table_prefix)\n",
    "        cmd = '/cndd/fangming/lab_scripts/bed_load_mysql.sh {} {} {}'.format(bed_file, \n",
    "                                                                             'CEMBA_annoj', \n",
    "                                                                             table_prefix)\n",
    "#         print(cmd)\n",
    "#         sp.call(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CEMBA_3C_171206_merged_cluster_C1',\n",
       " 'CEMBA_3C_171206_merged_cluster_C10',\n",
       " 'CEMBA_3C_171206_merged_cluster_C11',\n",
       " 'CEMBA_3C_171206_merged_cluster_C12',\n",
       " 'CEMBA_3C_171206_merged_cluster_C13',\n",
       " 'CEMBA_3C_171206_merged_cluster_C14',\n",
       " 'CEMBA_3C_171206_merged_cluster_C15',\n",
       " 'CEMBA_3C_171206_merged_cluster_C16',\n",
       " 'CEMBA_3C_171206_merged_cluster_C17',\n",
       " 'CEMBA_3C_171206_merged_cluster_C18',\n",
       " 'CEMBA_3C_171206_merged_cluster_C19',\n",
       " 'CEMBA_3C_171206_merged_cluster_C2',\n",
       " 'CEMBA_3C_171206_merged_cluster_C3',\n",
       " 'CEMBA_3C_171206_merged_cluster_C4',\n",
       " 'CEMBA_3C_171206_merged_cluster_C5',\n",
       " 'CEMBA_3C_171206_merged_cluster_C6',\n",
       " 'CEMBA_3C_171206_merged_cluster_C7',\n",
       " 'CEMBA_3C_171206_merged_cluster_C8',\n",
       " 'CEMBA_3C_171206_merged_cluster_C9']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# atac php\n",
    "\n",
    "ens = 'Ens10'\n",
    "output_folder_php = '/cndd/fangming/CEMBA/snmcseq_dev/browsers/php_atac_{}'.format(ens)\n",
    "    \n",
    "if not os.path.isdir(output_folder_php):\n",
    "    os.makedirs(output_folder_php)\n",
    "    logging.info('Created directory: {}'.format(output_folder_php))\n",
    "    \n",
    "for table_prefix in table_prefixes:\n",
    "    with open(os.path.join(output_folder_php, '{}.php'.format(table_prefix)), 'w') as f_php:\n",
    "        f_php.write(gen_php_template(table_prefix, TEMPLATE_ATAC_PHP))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C6</th>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C18</th>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C8</th>\n",
       "      <td>mL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C12</th>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C9</th>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           annotation\n",
       "cluster_ID           \n",
       "C6              mL2/3\n",
       "C18             mL2/3\n",
       "C8                mL4\n",
       "C12             mL5-1\n",
       "C9              mL5-1"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# atac annotation\n",
    "f = '/cndd/fangming/integration/XW45_draft_annotation.txt'\n",
    "atac_annot = [\n",
    "              {'cluster_ID': 'C6', 'annotation': 'mL2/3'},\n",
    "              {'cluster_ID': 'C18', 'annotation': 'mL2/3'},\n",
    "              {'cluster_ID': 'C8', 'annotation': 'mL4'},\n",
    "              {'cluster_ID': 'C12', 'annotation': 'mL5-1'},\n",
    "              {'cluster_ID': 'C9', 'annotation': 'mL5-1'},\n",
    "              {'cluster_ID': 'C7', 'annotation': 'mL5-2'},\n",
    "              {'cluster_ID': 'C3', 'annotation': 'mL6-1'},\n",
    "              {'cluster_ID': 'C17', 'annotation': 'mL6-2'},\n",
    "              {'cluster_ID': 'C5', 'annotation': 'mL6-2'},\n",
    "              {'cluster_ID': 'C19', 'annotation': 'mDL-2'},\n",
    "              \n",
    "              {'cluster_ID': 'C11', 'annotation': 'mSst'},\n",
    "              {'cluster_ID': 'C13', 'annotation': 'mVip/Ndnf'},\n",
    "              {'cluster_ID': 'C15', 'annotation': 'mPv'},\n",
    "    \n",
    "              {'cluster_ID': 'C1', 'annotation': 'mGlia'},\n",
    "              {'cluster_ID': 'C2', 'annotation': 'mGlia'},\n",
    "              {'cluster_ID': 'C14', 'annotation': 'mGlia'},\n",
    "              {'cluster_ID': 'C16', 'annotation': 'mGlia'},\n",
    "              {'cluster_ID': 'C4', 'annotation': 'mGlia'},\n",
    "              {'cluster_ID': 'C10', 'annotation': 'mGlia'},\n",
    "             ] \n",
    "atac_annot = pd.DataFrame(atac_annot) \n",
    "# atac_annot[['cluster_ID', 'annotation']].to_csv(f, sep=\"\\t\", header=True, index=False)\n",
    "atac_annot = atac_annot.set_index('cluster_ID')\n",
    "atac_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATAC_C1_mGlia', 'ATAC_C10_mGlia', 'ATAC_C11_mSst', 'ATAC_C12_mL5-1', 'ATAC_C13_mVip/Ndnf', 'ATAC_C14_mGlia', 'ATAC_C15_mPv', 'ATAC_C16_mGlia', 'ATAC_C17_mL6-2', 'ATAC_C18_mL2/3', 'ATAC_C19_mDL-2', 'ATAC_C2_mGlia', 'ATAC_C3_mL6-1', 'ATAC_C4_mGlia', 'ATAC_C5_mL6-2', 'ATAC_C6_mL2/3', 'ATAC_C7_mL5-2', 'ATAC_C8_mL4', 'ATAC_C9_mL5-1']\n"
     ]
    }
   ],
   "source": [
    "# atac paragraph in html\n",
    "\n",
    "track_vis_names_atac = [(lambda clst: 'ATAC_{}_{}'.format(clst, atac_annot.loc[clst, 'annotation']))\n",
    "                        (table_prefix.split('_')[-1]) for table_prefix in table_prefixes]\n",
    "print(track_vis_names_atac)\n",
    "\n",
    "atac_configs = gen_annoj_configs(table_prefixes, track_vis_names_atac, TEMPLATE_INDEX_ATAC) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C1',\n",
      "        name : 'ATAC_C1_mGlia',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C1.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C10',\n",
      "        name : 'ATAC_C10_mGlia',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C10.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C11',\n",
      "        name : 'ATAC_C11_mSst',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C11.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C12',\n",
      "        name : 'ATAC_C12_mL5-1',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C12.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C13',\n",
      "        name : 'ATAC_C13_mVip/Ndnf',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C13.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C14',\n",
      "        name : 'ATAC_C14_mGlia',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C14.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C15',\n",
      "        name : 'ATAC_C15_mPv',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C15.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C16',\n",
      "        name : 'ATAC_C16_mGlia',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C16.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C17',\n",
      "        name : 'ATAC_C17_mL6-2',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C17.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C18',\n",
      "        name : 'ATAC_C18_mL2/3',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C18.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C19',\n",
      "        name : 'ATAC_C19_mDL-2',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C19.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C2',\n",
      "        name : 'ATAC_C2_mGlia',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C2.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C3',\n",
      "        name : 'ATAC_C3_mL6-1',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C3.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C4',\n",
      "        name : 'ATAC_C4_mGlia',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C4.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C5',\n",
      "        name : 'ATAC_C5_mL6-2',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C5.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C6',\n",
      "        name : 'ATAC_C6_mL2/3',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C6.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C7',\n",
      "        name : 'ATAC_C7_mL5-2',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C7.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C8',\n",
      "        name : 'ATAC_C8_mL4',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C8.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n",
      "      {\n",
      "        id : 'CEMBA_3C_171206_merged_cluster_C9',\n",
      "        name : 'ATAC_C9_mL5-1',\n",
      "        type : 'PairedEndTrack',\n",
      "        path : 'ATAC-Seq',\n",
      "        data : './browser/fetchers/atac/CEMBA_3C_171206_merged_cluster_C9.php',\n",
      "        iconCls : 'silk_bricks',\n",
      "        height : 20,\n",
      "        scale : 2,\n",
      "        single : true,\n",
      "      },\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(atac_configs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'ATAC_C1_mGlia',\n",
      "'ATAC_C10_mGlia',\n",
      "'ATAC_C11_mSst',\n",
      "'ATAC_C12_mL5-1',\n",
      "'ATAC_C13_mVip/Ndnf',\n",
      "'ATAC_C14_mGlia',\n",
      "'ATAC_C15_mPv',\n",
      "'ATAC_C16_mGlia',\n",
      "'ATAC_C17_mL6-2',\n",
      "'ATAC_C18_mL2/3',\n",
      "'ATAC_C19_mDL-2',\n",
      "'ATAC_C2_mGlia',\n",
      "'ATAC_C3_mL6-1',\n",
      "'ATAC_C4_mGlia',\n",
      "'ATAC_C5_mL6-2',\n",
      "'ATAC_C6_mL2/3',\n",
      "'ATAC_C7_mL5-2',\n",
      "'ATAC_C8_mL4',\n",
      "'ATAC_C9_mL5-1',\n"
     ]
    }
   ],
   "source": [
    "# active order\n",
    "for i in track_vis_names_atac:\n",
    "    print('\\'{}\\','.format(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'CEMBA_3C_171206_merged_cluster_C1',\n",
    "'CEMBA_3C_171206_merged_cluster_C10',\n",
    "'CEMBA_3C_171206_merged_cluster_C11',\n",
    "'CEMBA_3C_171206_merged_cluster_C12',\n",
    "'CEMBA_3C_171206_merged_cluster_C13',\n",
    "'CEMBA_3C_171206_merged_cluster_C14',\n",
    "'CEMBA_3C_171206_merged_cluster_C15',\n",
    "'CEMBA_3C_171206_merged_cluster_C16',\n",
    "'CEMBA_3C_171206_merged_cluster_C17',\n",
    "'CEMBA_3C_171206_merged_cluster_C18',\n",
    "'CEMBA_3C_171206_merged_cluster_C19',\n",
    "'CEMBA_3C_171206_merged_cluster_C2',\n",
    "'CEMBA_3C_171206_merged_cluster_C3',\n",
    "'CEMBA_3C_171206_merged_cluster_C4',\n",
    "'CEMBA_3C_171206_merged_cluster_C5',\n",
    "'CEMBA_3C_171206_merged_cluster_C6',\n",
    "'CEMBA_3C_171206_merged_cluster_C7',\n",
    "'CEMBA_3C_171206_merged_cluster_C8',\n",
    "'CEMBA_3C_171206_merged_cluster_C9',\n",
    "\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst1_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst2_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst3_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst4_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst5_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst6_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst7_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst8_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst9_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst10_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst11_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst12_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst13_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst14_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst15_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst16_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst17_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst18_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst19_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst19_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst20_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst20_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst21_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst21_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst22_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst22_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst23_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst23_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst24_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst24_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst25_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst25_Ens10',\n",
    "'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst26_Ens10',\n",
    "'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst26_Ens10',\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_ID</th>\n",
       "      <th>cluster_annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cluster_8</td>\n",
       "      <td>mDL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cluster_7</td>\n",
       "      <td>mDL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>cluster_26</td>\n",
       "      <td>mDL-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cluster_2</td>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cluster_3</td>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>cluster_19</td>\n",
       "      <td>mL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cluster_1</td>\n",
       "      <td>mL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cluster_4</td>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cluster_6</td>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cluster_10</td>\n",
       "      <td>mL5-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cluster_16</td>\n",
       "      <td>mL6-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cluster_5</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cluster_13</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cluster_14</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>cluster_20</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>cluster_22</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cluster_12</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>cluster_17</td>\n",
       "      <td>mNdnf-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cluster_11</td>\n",
       "      <td>mPv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>cluster_24</td>\n",
       "      <td>mPv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cluster_9</td>\n",
       "      <td>mSst-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>cluster_25</td>\n",
       "      <td>mSst-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>cluster_15</td>\n",
       "      <td>mVip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cluster_18</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>cluster_21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>cluster_23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cluster_ID cluster_annotation\n",
       "7    cluster_8              mDL-2\n",
       "6    cluster_7              mDL-2\n",
       "25  cluster_26              mDL-3\n",
       "1    cluster_2              mL2/3\n",
       "2    cluster_3              mL2/3\n",
       "18  cluster_19                mL4\n",
       "0    cluster_1                mL4\n",
       "3    cluster_4              mL5-1\n",
       "5    cluster_6              mL5-1\n",
       "9   cluster_10              mL5-2\n",
       "15  cluster_16              mL6-1\n",
       "4    cluster_5              mL6-2\n",
       "12  cluster_13              mL6-2\n",
       "13  cluster_14              mL6-2\n",
       "19  cluster_20              mL6-2\n",
       "21  cluster_22              mL6-2\n",
       "11  cluster_12              mL6-2\n",
       "16  cluster_17            mNdnf-2\n",
       "10  cluster_11                mPv\n",
       "23  cluster_24                mPv\n",
       "8    cluster_9             mSst-1\n",
       "24  cluster_25             mSst-1\n",
       "14  cluster_15               mVip\n",
       "17  cluster_18                NaN\n",
       "20  cluster_21                NaN\n",
       "22  cluster_23                NaN"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mc cluster type\n",
    "\n",
    "f = os.path.join(PATH_ENSEMBLES, 'Ens10', 'cluster', 'cluster_lv_npc50_k30_binc_mCHmCG_100000_nmcc_Ens10.tsv.annot')\n",
    "mc_annot = pd.read_table(f)\n",
    "mc_annot = mc_annot.sort_values('cluster_annotation')\n",
    "mc_annot\n",
    "# for idx, row in mc_annot.iterrows():\n",
    "#     clst = int(row['cluster_ID'].split('_')[-1])\n",
    "#     annot = row['cluster_annotation']\n",
    "#     print('\\'mc_merged_mCG_cluster_mCHmCG_lv_npc50_k30_clst{}_Ens10\\','.format(clst))\n",
    "#     print('\\'dmr_hypo_mCG_cluster_mCHmCG_lv_npc50_k30_clst{}_Ens10\\','.format(clst))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C6</th>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C18</th>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C8</th>\n",
       "      <td>mL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C12</th>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C9</th>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C7</th>\n",
       "      <td>mL5-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3</th>\n",
       "      <td>mL6-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C17</th>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C5</th>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C19</th>\n",
       "      <td>mDL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C11</th>\n",
       "      <td>mSst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C13</th>\n",
       "      <td>mVip/Ndnf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C15</th>\n",
       "      <td>mPv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1</th>\n",
       "      <td>mGlia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2</th>\n",
       "      <td>mGlia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C14</th>\n",
       "      <td>mGlia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C16</th>\n",
       "      <td>mGlia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C4</th>\n",
       "      <td>mGlia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10</th>\n",
       "      <td>mGlia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           annotation\n",
       "cluster_ID           \n",
       "C6              mL2/3\n",
       "C18             mL2/3\n",
       "C8                mL4\n",
       "C12             mL5-1\n",
       "C9              mL5-1\n",
       "C7              mL5-2\n",
       "C3              mL6-1\n",
       "C17             mL6-2\n",
       "C5              mL6-2\n",
       "C19             mDL-2\n",
       "C11              mSst\n",
       "C13         mVip/Ndnf\n",
       "C15               mPv\n",
       "C1              mGlia\n",
       "C2              mGlia\n",
       "C14             mGlia\n",
       "C16             mGlia\n",
       "C4              mGlia\n",
       "C10             mGlia"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_ID</th>\n",
       "      <th>cluster_annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cluster_1</td>\n",
       "      <td>mL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cluster_2</td>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cluster_3</td>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cluster_4</td>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cluster_5</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cluster_6</td>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cluster_7</td>\n",
       "      <td>mDL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cluster_8</td>\n",
       "      <td>mDL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cluster_9</td>\n",
       "      <td>mSst-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cluster_10</td>\n",
       "      <td>mL5-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cluster_11</td>\n",
       "      <td>mPv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cluster_12</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cluster_13</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cluster_14</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>cluster_15</td>\n",
       "      <td>mVip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cluster_16</td>\n",
       "      <td>mL6-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>cluster_17</td>\n",
       "      <td>mNdnf-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cluster_18</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>cluster_19</td>\n",
       "      <td>mL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>cluster_20</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>cluster_21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>cluster_22</td>\n",
       "      <td>mL6-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>cluster_23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>cluster_24</td>\n",
       "      <td>mPv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>cluster_25</td>\n",
       "      <td>mSst-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>cluster_26</td>\n",
       "      <td>mDL-3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cluster_ID cluster_annotation\n",
       "0    cluster_1                mL4\n",
       "1    cluster_2              mL2/3\n",
       "2    cluster_3              mL2/3\n",
       "3    cluster_4              mL5-1\n",
       "4    cluster_5              mL6-2\n",
       "5    cluster_6              mL5-1\n",
       "6    cluster_7              mDL-2\n",
       "7    cluster_8              mDL-2\n",
       "8    cluster_9             mSst-1\n",
       "9   cluster_10              mL5-2\n",
       "10  cluster_11                mPv\n",
       "11  cluster_12              mL6-2\n",
       "12  cluster_13              mL6-2\n",
       "13  cluster_14              mL6-2\n",
       "14  cluster_15               mVip\n",
       "15  cluster_16              mL6-1\n",
       "16  cluster_17            mNdnf-2\n",
       "17  cluster_18                NaN\n",
       "18  cluster_19                mL4\n",
       "19  cluster_20              mL6-2\n",
       "20  cluster_21                NaN\n",
       "21  cluster_22              mL6-2\n",
       "22  cluster_23                NaN\n",
       "23  cluster_24                mPv\n",
       "24  cluster_25             mSst-1\n",
       "25  cluster_26              mDL-3"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mc_annot.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
