{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlalchemy as sa\n",
    "from sqlalchemy import update\n",
    "from sqlalchemy.dialects.mysql import INTEGER\n",
    "from sqlalchemy.dialects.mysql import SMALLINT\n",
    "from collections import OrderedDict\n",
    "from itertools import product \n",
    "import re\n",
    "\n",
    "\n",
    "from __init__ import *\n",
    "from snmcseq_utils import cd \n",
    "from snmcseq_utils import get_mouse_chromosomes\n",
    "from snmcseq_utils import compute_global_mC \n",
    "from snmcseq_utils import create_logger\n",
    "from CEMBA_run_tsne import run_tsne_CEMBA\n",
    "from CEMBA_update_mysql import insert_into\n",
    "from CEMBA_update_mysql import insert_into_worker\n",
    "from CEMBA_update_mysql import gene_id_to_table_name \n",
    "from CEMBA_update_mysql import connect_sql "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "log = create_logger()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# An engine connects to a mysql database\n",
    "database = 'CEMBA_RNA'\n",
    "engine = connect_sql(database) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating tables\n",
    "# \"genes\" table (gene annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_genes_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('genes', metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('gene_id', sa.String(50), primary_key=True, autoincrement=False), \n",
    "                        sa.Column('gene_name', sa.String(100), nullable=False), \n",
    "                        sa.Column('chr', sa.String(5), nullable=True), \n",
    "                        sa.Column('start', sa.Integer, nullable=True), \n",
    "                        sa.Column('end', sa.Integer, nullable=True), \n",
    "                        sa.Column('strand', sa.CHAR(1), nullable=True), \n",
    "                        sa.Column('gene_type', sa.String(100), nullable=True), \n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create genes table\n",
    "metadata = sa.MetaData(engine)\n",
    "genes_table = define_genes_table(metadata)\n",
    "metadata.create_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/06/2018 04:41:23 PM                 gene_id      gene_name   chr    start      end strand  \\\n",
      "0  ENSMUSG00000102693.1  4933401J01Rik  chr1  3073253  3074322      +   \n",
      "1  ENSMUSG00000064842.1        Gm26206  chr1  3102016  3102125      +   \n",
      "2  ENSMUSG00000051951.5           Xkr4  chr1  3205901  3671498      -   \n",
      "3  ENSMUSG00000102851.1        Gm18956  chr1  3252757  3253236      +   \n",
      "4  ENSMUSG00000103377.1        Gm37180  chr1  3365731  3368549      -   \n",
      "\n",
      "              gene_type  \n",
      "0                   TEC  \n",
      "1                 snRNA  \n",
      "2        protein_coding  \n",
      "3  processed_pseudogene  \n",
      "4                   TEC  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<sqlalchemy.engine.result.ResultProxy at 0x7f7735161eb8>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update to genes table\n",
    "table_name = 'genes'\n",
    "\n",
    "path_genebody = PATH_GENEBODY_ANNOTATION\n",
    "df_genes = pd.read_table(path_genebody, index_col='gene_id')\n",
    "\n",
    "df_sql = df_genes.reset_index()\n",
    "insert_into(engine, 'genes', df_sql, ignore=False, verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# \"ensembles\" table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_enss_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('ensembles', metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('ensemble_id', sa.Integer, primary_key=True, autoincrement=False), \n",
    "                        sa.Column('ensemble_name', sa.String(255), nullable=False, unique=True),\n",
    "                        sa.Column('public_access', sa.Boolean, nullable=False, default=False),\n",
    "                        sa.Column('datasets', sa.String(4000), nullable=False),\n",
    "                        sa.Column('description', sa.String(255), nullable=True),\n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create enss table\n",
    "metadata = sa.MetaData(engine)\n",
    "enss_table = define_enss_table(metadata)\n",
    "metadata.create_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update to enss table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# \"cells\" table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_cells_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('cells', metadata,\n",
    "                sa.Column('cell_id', sa.Integer, primary_key=True, autoincrement=True), # auto increment and not null are implicitly defined\n",
    "                sa.Column('cell_name', sa.String(255), nullable=False, unique=True), # unique but not primary key\n",
    "                sa.Column('dataset', sa.String(40), nullable=False), # not in mapping summary\n",
    "#                 sa.Column('cell_type', sa.String(20), nullable=True), # not in mapping summary\n",
    "                     \n",
    "#                 sa.Column('global_mCH', sa.Float, nullable=True),\n",
    "#                 sa.Column('global_mCG', sa.Float, nullable=True),\n",
    "#                 sa.Column('global_mCA', sa.Float, nullable=True), # not in mapping summary\n",
    "#                 sa.Column('global_mCCC', sa.Float, nullable=True), \n",
    "#                 sa.Column('estimated_mCH', sa.Float, nullable=True),\n",
    "#                 sa.Column('estimated_mCG', sa.Float, nullable=True),\n",
    "#                 sa.Column('percent_genome_covered', sa.Float, nullable=True),\n",
    "                     \n",
    "#                 sa.Column('total_reads', sa.Integer, nullable=True),\n",
    "#                 sa.Column('mapped_reads', sa.Integer, nullable=True),\n",
    "#                 sa.Column('mapping_rate', sa.Float, nullable=True),\n",
    "#                 sa.Column('nonclonal_reads', sa.Integer, nullable=True),\n",
    "#                 sa.Column('percent_nonclonal_rate', sa.Float, nullable=True),\n",
    "#                 sa.Column('filtered_reads', sa.Integer, nullable=True),\n",
    "#                 sa.Column('filtered_rate', sa.Float, nullable=True),\n",
    "#                 sa.Column('lambda_mC', sa.Float, nullable=True),\n",
    "                )\n",
    "    return table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create cells table\n",
    "metadata = sa.MetaData(engine)\n",
    "cells_table = define_cells_table(metadata)\n",
    "metadata.create_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update to cells table (dataset level)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Individual gene tables (\"gene_ensemble_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_gene_table(metadata, gene_id):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table_name = gene_id_to_table_name(gene_id)\n",
    "    table = sa.Table(table_name, metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('cell_id', sa.Integer, sa.ForeignKey('cells.cell_id'), primary_key=True, autoincrement=False), \n",
    "                        sa.Column('normalized_counts', sa.Float, nullable=True),\n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/06/2018 04:43:00 PM Creating gene tables... (53379 in total)\n"
     ]
    }
   ],
   "source": [
    "# create individual gene tables (takes time)\n",
    "ti = time.time()\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "# load specfic tables needed to be referenced as foreign key\n",
    "sa.Table('cells', metadata, autoload=True)\n",
    "# load all genes \n",
    "sql = 'SELECT * FROM genes'\n",
    "df_genes = pd.read_sql(sql, engine, index_col='gene_id')\n",
    "# define gene tables \n",
    "for i, gene_id in enumerate(df_genes.index):\n",
    "    gene_table = define_gene_table(metadata, gene_id)\n",
    "# create them\n",
    "logging.info(\"Creating gene tables... ({} in total)\".format(df_genes.index.shape[0]))\n",
    "metadata.create_all(engine)\n",
    "logging.info(time.time() - ti)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update to gene tables (dataset level)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Individual ensemble tables (\"Ens1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_ens_table(metadata, ens):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # prototypes\n",
    "#   sa.Column('tsne_x_mCH', sa.Float, nullable=True),\n",
    "#   sa.Column('tsne_y_mCH', sa.Float, nullable=True),\n",
    "#   sa.Column('cluster', SMALLINT(unsigned=True), nullable=True),\n",
    "#   sa.Column('annotation', sa.String(20), nullable=True),\n",
    "\n",
    "    # tsnes\n",
    "    contexts = ['ATAC'] \n",
    "    perps = PERPLEXITIES\n",
    "#     tsne_types = ['{}_ndim2_perp{}'.format(context, p) for (context, p) in product(contexts, perps)]\n",
    "    tsne_types = ['ATAC']\n",
    "    \n",
    "    # prepare columns\n",
    "    args_tsne = ([sa.Column('tsne_x_{}'.format(tsne_type), sa.Float, nullable=True) \n",
    "            for tsne_type in tsne_types] \n",
    "            + [sa.Column('tsne_y_{}'.format(tsne_type), sa.Float, nullable=True) \n",
    "            for tsne_type in tsne_types]\n",
    "           )\n",
    "    \n",
    "    # clusters\n",
    "    ks = K_NN\n",
    "#     cluster_types = ['{}_lv_npc50_k{}'.format(context, k) for (context, k) in product(contexts, ks)]\n",
    "    cluster_types = ['ATAC']\n",
    "    # prepare columns\n",
    "    args_cluster = ([sa.Column('cluster_{}'.format(cluster_type), SMALLINT(unsigned=True), nullable=True) \n",
    "            for cluster_type in cluster_types] \n",
    "            + [sa.Column('annotation_{}'.format(cluster_type), sa.String(20), nullable=True) \n",
    "            for cluster_type in cluster_types]\n",
    "           )\n",
    "    \n",
    "    # combine cols\n",
    "    args = args_tsne + args_cluster\n",
    "    # create the table\n",
    "    table = sa.Table(ens, metadata,\n",
    "                        # NOT auto increment not null primary key\n",
    "                        sa.Column('cell_id', sa.Integer, sa.ForeignKey('cells.cell_id'), primary_key=True, autoincrement=False), \n",
    "                        *args\n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create individual ensemble tables (ensemble level)\n",
    "# demo here\n",
    "ens = 'Ens0'\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "# load specfic tables needed to be referenced as foreign key\n",
    "sa.Table('cells', metadata, autoload=True)\n",
    "# define and create \n",
    "ens_table = define_ens_table(metadata, ens)\n",
    "metadata.create_all(engine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update to ensemble tables (ensemble level)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to mysql\n",
    "\n",
    "### Dataset level\n",
    "\n",
    "- upload to dataset table\n",
    "- upload to cells table\n",
    "- upload to gene tables (bottle-neck)\n",
    "\n",
    "### Ensemble level\n",
    "\n",
    "- upload to enss table\n",
    "- create ens table && upload to ens table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSMUSG00000051951.5'"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def gene_id_to_name(gene_id, df_genes):\n",
    "    \"\"\"df_genes\n",
    "    \"\"\"\n",
    "    return df_genes.loc[gene_id, 'gene_name']\n",
    "\n",
    "def gene_name_to_id(gene_name, df_genes):\n",
    "    \"\"\"df_genes\n",
    "    \"\"\"\n",
    "    return df_genes[df_genes['gene_name'] == gene_name].index.values[0]\n",
    "\n",
    "gene_name_to_id('Xkr4', df_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001.4</th>\n",
       "      <td>Gnai3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>108107280</td>\n",
       "      <td>108146146</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003.15</th>\n",
       "      <td>Pbsn</td>\n",
       "      <td>chrX</td>\n",
       "      <td>77837901</td>\n",
       "      <td>77853623</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.14</th>\n",
       "      <td>Cdc45</td>\n",
       "      <td>chr16</td>\n",
       "      <td>18780447</td>\n",
       "      <td>18811987</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031.16</th>\n",
       "      <td>H19</td>\n",
       "      <td>chr7</td>\n",
       "      <td>142575529</td>\n",
       "      <td>142578143</td>\n",
       "      <td>-</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037.16</th>\n",
       "      <td>Scml2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>161117193</td>\n",
       "      <td>161258213</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      gene_name    chr      start        end strand  \\\n",
       "gene_id                                                               \n",
       "ENSMUSG00000000001.4      Gnai3   chr3  108107280  108146146      -   \n",
       "ENSMUSG00000000003.15      Pbsn   chrX   77837901   77853623      -   \n",
       "ENSMUSG00000000028.14     Cdc45  chr16   18780447   18811987      -   \n",
       "ENSMUSG00000000031.16       H19   chr7  142575529  142578143      -   \n",
       "ENSMUSG00000000037.16     Scml2   chrX  161117193  161258213      +   \n",
       "\n",
       "                            gene_type  \n",
       "gene_id                                \n",
       "ENSMUSG00000000001.4   protein_coding  \n",
       "ENSMUSG00000000003.15  protein_coding  \n",
       "ENSMUSG00000000028.14  protein_coding  \n",
       "ENSMUSG00000000031.16         lincRNA  \n",
       "ENSMUSG00000000037.16  protein_coding  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "engine = connect_sql(database)\n",
    "sql = \"\"\"SELECT * FROM genes\"\"\"\n",
    "df_genes = pd.read_sql(sql, engine, index_col='gene_id')\n",
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gene*cell matrix\n",
    "\n",
    "# dataset = 'CEMBA_3C_171207'\n",
    "# with cd('/cndd/Public_Datasets/CEMBA/snATACSeq/MiniBrain_fromRongxin'):\n",
    "#     df_clst = pd.read_table('XW46_cluster.txt', header=None, names=['sample', 'cluster_ID'], index_col='sample')\n",
    "#     df_tsne = pd.read_table('XW46_tsne.txt', header=None, names=['sample', 'tsne_x', 'tsne_y'], index_col='sample')\n",
    "#     df = pd.read_table('XW46_gene.txt.gz') \n",
    "\n",
    "# df = df.T\n",
    "    \n",
    "# print(df_clst.shape)\n",
    "# print(df_tsne.shape)\n",
    "# print(df.shape)\n",
    "# df.head()\n",
    "from scipy import sparse\n",
    "from collections import namedtuple\n",
    "\n",
    "\n",
    "fdir = '/cndd/Public_Datasets/CEMBA/snATACSeq/Datasets/1B/CEMBA180118_1B_rep1/counts/genebody/'\n",
    "fdata = fdir + 'CEMBA180118_1B_rep1.genebody.count_matrix.npz'\n",
    "frow = fdir + 'CEMBA180118_1B_rep1.genebody.count_matrix.row.index' # gene\n",
    "fcol = fdir + 'CEMBA180118_1B_rep1.genebody.count_matrix.col.index' # col\n",
    "dataset = 'CEMBA_1B_180118'\n",
    "\n",
    "GC_matrix = namedtuple('GC_matrix', ['gene', 'cell', 'data'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ids = pd.read_table(frow, header=None)[0].values\n",
    "cells = pd.read_table(fcol, header=None)[0].values\n",
    "cells = cells + '_' + dataset \n",
    "data = sparse.load_npz(fdata)\n",
    "\n",
    "gc_matrix_raw = GC_matrix(\n",
    "    gene_ids, \n",
    "    cells, \n",
    "    data,\n",
    ")\n",
    "\n",
    "assert gc_matrix_raw.data.shape[0] == len(gc_matrix_raw.gene)\n",
    "assert gc_matrix_raw.data.shape[1] == len(gc_matrix_raw.cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "# logcpm\n",
    "\n",
    "def sparse_logcpm(gc_matrix):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    lib_size_inv = sparse.diags(np.ravel(1.0/gc_matrix.data.sum(axis=0)))\n",
    "    logcpm = (gc_matrix.data).dot(lib_size_inv*1e6).tocoo()\n",
    "    logcpm.data = np.log10(logcpm.data + 1)\n",
    "\n",
    "    gc_logcpm = GC_matrix(\n",
    "        gc_matrix.gene, \n",
    "        gc_matrix.cell, \n",
    "        logcpm,\n",
    "    )\n",
    "    \n",
    "    return gc_logcpm\n",
    "    \n",
    "gc_logcpm = sparse_logcpm(gc_matrix_raw)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to dataset table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/07/2018 01:51:47 PM Update dataset: CEMBA_1B_180118 to datasets table in CEMBA_ATAC database\n",
      "08/07/2018 01:51:47 PM May not be a valid dataset: CEMBA_1B_180118 (not found in snmCSeq datasets directory)\n"
     ]
    }
   ],
   "source": [
    "from CEMBA_update_mysql import upload_to_datasets\n",
    "upload_to_datasets(dataset, database=database, strict=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to cells table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9063, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_name</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGCGATAGAACCAGGTAACGAACGCAGGACGT_CEMBA_1B_180118</td>\n",
       "      <td>CEMBA_1B_180118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGCGATAGAACCAGGTAACGAACGGGCTCTGA_CEMBA_1B_180118</td>\n",
       "      <td>CEMBA_1B_180118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGCGATAGAACCAGGTAAGAGATGCAGGACGT_CEMBA_1B_180118</td>\n",
       "      <td>CEMBA_1B_180118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGCGATAGAACCAGGTATAGCCTTCAGGACGT_CEMBA_1B_180118</td>\n",
       "      <td>CEMBA_1B_180118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGCGATAGAACCAGGTCGAATTCCCAGGACGT_CEMBA_1B_180118</td>\n",
       "      <td>CEMBA_1B_180118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          cell_name          dataset\n",
       "0  AGCGATAGAACCAGGTAACGAACGCAGGACGT_CEMBA_1B_180118  CEMBA_1B_180118\n",
       "1  AGCGATAGAACCAGGTAACGAACGGGCTCTGA_CEMBA_1B_180118  CEMBA_1B_180118\n",
       "2  AGCGATAGAACCAGGTAAGAGATGCAGGACGT_CEMBA_1B_180118  CEMBA_1B_180118\n",
       "3  AGCGATAGAACCAGGTATAGCCTTCAGGACGT_CEMBA_1B_180118  CEMBA_1B_180118\n",
       "4  AGCGATAGAACCAGGTCGAATTCCCAGGACGT_CEMBA_1B_180118  CEMBA_1B_180118"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cells = pd.DataFrame()\n",
    "df_cells['cell_name'] = gc_matrix.cell \n",
    "df_cells['dataset'] = dataset \n",
    "\n",
    "print(df_cells.shape)\n",
    "df_cells.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/07/2018 01:33:07 PM                                           cell_name          dataset\n",
      "0  AGCGATAGAACCAGGTAACGAACGCAGGACGT_CEMBA_1B_180118  CEMBA_1B_180118\n",
      "1  AGCGATAGAACCAGGTAACGAACGGGCTCTGA_CEMBA_1B_180118  CEMBA_1B_180118\n",
      "2  AGCGATAGAACCAGGTAAGAGATGCAGGACGT_CEMBA_1B_180118  CEMBA_1B_180118\n",
      "3  AGCGATAGAACCAGGTATAGCCTTCAGGACGT_CEMBA_1B_180118  CEMBA_1B_180118\n",
      "4  AGCGATAGAACCAGGTCGAATTCCCAGGACGT_CEMBA_1B_180118  CEMBA_1B_180118\n",
      "08/07/2018 01:33:10 PM Upload complete.\n"
     ]
    }
   ],
   "source": [
    "# insert into\n",
    "insert_into(engine, 'cells', df_cells, ignore=False, verbose='True')\n",
    "logging.info(\"Upload complete.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to gene tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/07/2018 03:17:42 PM Progress on genes: 1 gene_ENSMUSG00000046334_4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSMUSG00000102693.1    cell_id  normalized_counts\n",
      "0     4951           2.513277\n",
      "ENSMUSG00000064842.1    cell_id  normalized_counts\n",
      "0     6774           1.896742\n",
      "ENSMUSG00000051951.5       cell_id  normalized_counts\n",
      "0        9061           3.609613\n",
      "1        9060           3.211215\n",
      "2        9058           3.158629\n",
      "3        9056           2.886287\n",
      "4        9048           3.213160\n",
      "5        9040           3.979891\n",
      "6        9035           2.673257\n",
      "7        9030           3.333683\n",
      "8        9024           3.643308\n",
      "9        9022           2.952043\n",
      "10       9016           3.282972\n",
      "11       9009           3.006551\n",
      "12       9006           3.370012\n",
      "13       8979           3.384755\n",
      "14       8975           2.904327\n",
      "15       8957           3.180250\n",
      "16       8955           3.115255\n",
      "17       8954           3.018372\n",
      "18       8946           3.269855\n",
      "19       8939           3.226106\n",
      "20       8933           3.071205\n",
      "21       8932           3.155671\n",
      "22       8931           3.131090\n",
      "23       8927           3.484600\n",
      "24       8926           3.099842\n",
      "25       8924           3.205086\n",
      "26       8917           2.986979\n",
      "27       8887           2.887289\n",
      "28       8884           2.724702\n",
      "29       8876           3.373587\n",
      "...       ...                ...\n",
      "3627       62           2.951656\n",
      "3628       61           2.772377\n",
      "3629       58           3.191731\n",
      "3630       57           2.771331\n",
      "3631       56           3.113449\n",
      "3632       51           2.740429\n",
      "3633       50           2.757763\n",
      "3634       47           3.170186\n",
      "3635       46           3.019844\n",
      "3636       43           2.959973\n",
      "3637       39           2.724416\n",
      "3638       35           3.205782\n",
      "3639       32           2.409518\n",
      "3640       29           2.400878\n",
      "3641       28           3.148439\n",
      "3642       26           3.039179\n",
      "3643       25           2.614468\n",
      "3644       24           2.926125\n",
      "3645       23           2.532891\n",
      "3646       22           2.910629\n",
      "3647       20           2.308523\n",
      "3648       17           2.153579\n",
      "3649       16           3.065213\n",
      "3650       15           2.763415\n",
      "3651       13           2.304061\n",
      "3652        9           2.728970\n",
      "3653        6           2.789452\n",
      "3654        5           2.373707\n",
      "3655        4           2.926247\n",
      "3656        3           2.304321\n",
      "\n",
      "[3657 rows x 2 columns]\n",
      "ENSMUSG00000102851.1    cell_id  normalized_counts\n",
      "0     4606           1.893186\n",
      "1     1923           2.085447\n",
      "2     1896           1.915561\n",
      "3      505           2.541593\n",
      "4       71           2.464556\n",
      "ENSMUSG00000103377.1     cell_id  normalized_counts\n",
      "0      8637           2.551358\n",
      "1      8623           2.689735\n",
      "2      8513           2.779048\n",
      "3      8446           2.359339\n",
      "4      8397           2.199907\n",
      "5      8372           2.234382\n",
      "6      8352           2.094487\n",
      "7      8308           2.363545\n",
      "8      8186           2.428576\n",
      "9      8023           2.720988\n",
      "10     7921           2.245345\n",
      "11     7888           2.354444\n",
      "12     7828           2.544572\n",
      "13     7690           2.386364\n",
      "14     7638           2.889970\n",
      "15     7276           1.788408\n",
      "16     7259           3.499078\n",
      "17     7087           2.703319\n",
      "18     7008           2.232032\n",
      "19     6838           2.443100\n",
      "20     6524           2.751572\n",
      "21     6238           2.688713\n",
      "22     6233           3.074538\n",
      "23     5958           3.040044\n",
      "24     5950           2.769652\n",
      "25     5780           2.688607\n",
      "26     5613           2.618041\n",
      "27     5568           2.614780\n",
      "28     5296           1.951232\n",
      "29     5267           2.253967\n",
      "..      ...                ...\n",
      "69     1880           2.564942\n",
      "70     1873           2.021341\n",
      "71     1690           2.364723\n",
      "72     1675           2.157500\n",
      "73     1649           2.737396\n",
      "74     1339           2.610572\n",
      "75     1183           2.186825\n",
      "76     1111           2.400161\n",
      "77      987           2.240127\n",
      "78      976           2.210815\n",
      "79      957           2.562648\n",
      "80      862           2.497762\n",
      "81      823           2.243536\n",
      "82      780           2.437279\n",
      "83      767           2.663086\n",
      "84      750           2.182508\n",
      "85      748           1.891073\n",
      "86      735           2.009948\n",
      "87      732           2.644291\n",
      "88      594           2.688396\n",
      "89      506           2.214349\n",
      "90      499           1.866228\n",
      "91      484           1.888856\n",
      "92      472           2.194405\n",
      "93      153           2.026609\n",
      "94      133           3.081805\n",
      "95      120           3.345298\n",
      "96      114           3.640264\n",
      "97       32           2.409518\n",
      "98        6           2.789452\n",
      "\n",
      "[99 rows x 2 columns]\n",
      "ENSMUSG00000104017.1     cell_id  normalized_counts\n",
      "0      8794           2.704030\n",
      "1      8581           2.460926\n",
      "2      7999           2.560249\n",
      "3      7973           2.596268\n",
      "4      7692           3.091599\n",
      "5      6199           2.834207\n",
      "6      6040           2.546011\n",
      "7      5307           2.321064\n",
      "8      5137           2.270065\n",
      "9      5132           1.862457\n",
      "10     5031           2.324433\n",
      "11     4875           2.261882\n",
      "12     4766           2.361635\n",
      "13     4566           2.221864\n",
      "14     4148           1.988660\n",
      "15     4145           2.853860\n",
      "16     4128           2.670009\n",
      "17     4023           2.799020\n",
      "18     3933           2.329709\n",
      "19     3657           2.818721\n",
      "20     2650           1.913942\n",
      "21     2579           2.718832\n",
      "22     2555           2.263632\n",
      "23     2302           2.511732\n",
      "24     2293           2.048145\n",
      "25     1972           1.507663\n",
      "26     1958           2.108197\n",
      "27     1910           1.807948\n",
      "28     1525           2.086358\n",
      "29     1441           2.124493\n",
      "30     1333           2.291474\n",
      "31      976           1.912450\n",
      "32      487           2.421124\n",
      "33      105           1.975339\n",
      "ENSMUSG00000103025.1     cell_id  normalized_counts\n",
      "0      8957           3.180250\n",
      "1      8939           3.226106\n",
      "2      8732           2.374778\n",
      "3      7679           3.665305\n",
      "4      7454           2.492131\n",
      "5      7319           2.881645\n",
      "6      7247           2.639600\n",
      "7      7181           2.213823\n",
      "8      6886           3.062107\n",
      "9      6702           2.773617\n",
      "10     6656           2.467900\n",
      "11     6646           2.378674\n",
      "12     6267           2.649951\n",
      "13     6150           3.445263\n",
      "14     5632           3.043505\n",
      "15     5608           1.866663\n",
      "16     5527           2.797456\n",
      "17     5482           2.505882\n",
      "18     5419           2.519728\n",
      "19     5412           2.399133\n",
      "20     5375           2.448213\n",
      "21     5358           1.717840\n",
      "22     5331           1.962344\n",
      "23     5191           2.353471\n",
      "24     5154           2.536700\n",
      "25     4929           2.265410\n",
      "26     4676           2.109956\n",
      "27     4563           2.284260\n",
      "28     4554           1.868673\n",
      "29     4306           2.415619\n",
      "..      ...                ...\n",
      "36     3135           3.194437\n",
      "37     2892           2.511031\n",
      "38     2719           2.741255\n",
      "39     2670           2.461612\n",
      "40     2518           1.977433\n",
      "41     2499           2.289418\n",
      "42     2442           1.879236\n",
      "43     2384           2.220151\n",
      "44     2336           2.484165\n",
      "45     2170           2.644323\n",
      "46     2036           2.211827\n",
      "47     2034           1.927809\n",
      "48     1987           2.598062\n",
      "49     1908           2.919356\n",
      "50     1861           2.405526\n",
      "51     1849           2.436042\n",
      "52     1815           2.770161\n",
      "53     1673           2.689136\n",
      "54     1527           2.259241\n",
      "55     1526           2.901210\n",
      "56     1396           3.355999\n",
      "57     1337           2.417897\n",
      "58     1144           2.550974\n",
      "59      975           2.352549\n",
      "60      768           2.443040\n",
      "61      732           2.344245\n",
      "62      288           2.910805\n",
      "63      243           2.997839\n",
      "64      207           2.368596\n",
      "65      153           2.325592\n",
      "\n",
      "[66 rows x 2 columns]\n",
      "ENSMUSG00000089699.1      cell_id  normalized_counts\n",
      "0       8876           3.373587\n",
      "1       8791           2.404240\n",
      "2       8761           2.618759\n",
      "3       8753           2.755374\n",
      "4       8742           2.794277\n",
      "5       8698           2.367641\n",
      "6       8669           2.548066\n",
      "7       8606           2.021363\n",
      "8       8595           1.921305\n",
      "9       8581           2.460926\n",
      "10      8579           2.122929\n",
      "11      8563           2.129018\n",
      "12      8547           2.639411\n",
      "13      8535           2.340365\n",
      "14      8518           2.563780\n",
      "15      8471           2.260315\n",
      "16      8444           2.701901\n",
      "17      8419           2.474924\n",
      "18      8417           2.637203\n",
      "19      8411           2.348281\n",
      "20      8390           2.145933\n",
      "21      8372           2.234382\n",
      "22      8361           2.308151\n",
      "23      8337           2.732578\n",
      "24      8306           3.093746\n",
      "25      8294           2.359487\n",
      "26      8219           2.202053\n",
      "27      8216           2.035029\n",
      "28      8184           2.698107\n",
      "29      8180           2.541255\n",
      "..       ...                ...\n",
      "418      428           2.517946\n",
      "419      424           2.292273\n",
      "420      390           2.734863\n",
      "421      376           3.003045\n",
      "422      369           1.752867\n",
      "423      344           2.164450\n",
      "424      322           2.487864\n",
      "425      320           2.769143\n",
      "426      307           2.195648\n",
      "427      293           2.257800\n",
      "428      291           1.972536\n",
      "429      275           2.387046\n",
      "430      252           2.220971\n",
      "431      249           2.095817\n",
      "432      247           2.668091\n",
      "433      243           2.697245\n",
      "434      225           2.127696\n",
      "435      221           2.895213\n",
      "436      215           1.878948\n",
      "437      213           2.124065\n",
      "438      207           2.668696\n",
      "439      203           1.820271\n",
      "440      193           2.077767\n",
      "441      164           2.063402\n",
      "442       99           2.797524\n",
      "443       95           2.401136\n",
      "444       74           2.317475\n",
      "445       63           2.923060\n",
      "446       39           2.248930\n",
      "447       24           2.450033\n",
      "\n",
      "[448 rows x 2 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSMUSG00000103201.1     cell_id  normalized_counts\n",
      "0      8361           2.308151\n",
      "1      8337           2.732578\n",
      "2      7357           2.613048\n",
      "3      7043           1.950322\n",
      "4      6822           2.496441\n",
      "5      6708           2.920076\n",
      "6      5796           2.675198\n",
      "7      5058           2.883463\n",
      "8      4676           2.109956\n",
      "9      4535           1.611768\n",
      "10     4471           2.918368\n",
      "11     4454           2.296869\n",
      "12     4184           2.326002\n",
      "13     4148           1.988660\n",
      "14     3734           2.450580\n",
      "15     3362           2.367892\n",
      "16     2945           2.754636\n",
      "17     2329           1.681193\n",
      "18     1749           1.925660\n",
      "19      347           2.526822\n",
      "20      344           2.463990\n",
      "21       39           2.248930\n",
      "22       22           2.610132\n",
      "ENSMUSG00000103147.1     cell_id  normalized_counts\n",
      "0      8592           2.236930\n",
      "1      8243           2.379499\n",
      "2      7507           3.005672\n",
      "3      6903           2.613225\n",
      "4      6857           2.645436\n",
      "5      6202           2.903633\n",
      "6      4130           2.433402\n",
      "7      3923           2.227842\n",
      "8      3468           2.276230\n",
      "9      2809           2.602452\n",
      "10     2509           2.566611\n",
      "11     2240           2.349218\n",
      "12     2227           1.803335\n",
      "13     2184           1.808395\n",
      "14     2150           2.255202\n",
      "15     2085           2.680346\n",
      "16     2056           1.798587\n",
      "17     1666           2.107950\n",
      "18     1254           2.717532\n",
      "19      767           2.838863\n",
      "20      515           2.097419\n",
      "21      472           2.194405\n",
      "22      428           2.218232\n",
      "23      398           1.912415\n",
      "24      389           2.347826\n",
      "ENSMUSG00000103161.1     cell_id  normalized_counts\n",
      "0      8700           2.888627\n",
      "1      7472           2.978029\n",
      "2      7281           1.976083\n",
      "3      7064           2.299067\n",
      "4      6971           2.660209\n",
      "5      6928           2.331643\n",
      "6      6561           2.578312\n",
      "7      6467           2.598748\n",
      "8      6271           3.300380\n",
      "9      5748           2.501067\n",
      "10     5493           1.805266\n",
      "11     5182           2.544382\n",
      "12     5127           2.726389\n",
      "13     4904           2.076314\n",
      "14     4267           2.649661\n",
      "15     4231           2.963047\n",
      "16     4119           2.792260\n",
      "17     3475           1.994461\n",
      "18     3302           2.218622\n",
      "19     3246           2.730479\n",
      "20     3132           2.078406\n",
      "21     2619           1.952125\n",
      "22     2328           2.219120\n",
      "23     2213           2.596652\n",
      "24     1977           2.640024\n",
      "25     1847           2.095950\n",
      "26     1668           2.019733\n",
      "27      756           2.592363\n",
      "28      414           1.596549\n",
      "29      273           2.452929\n",
      "30      151           1.881564\n",
      "31       20           2.009622\n"
     ]
    }
   ],
   "source": [
    "sql = \"\"\"SELECT * FROM cells\"\"\"\n",
    "df_cells = pd.read_sql(sql, engine)\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "\n",
    "for i, gene in enumerate(gc_logcpm.gene):\n",
    "    if i > 10:\n",
    "        break\n",
    "    if (i%1000 == 0):\n",
    "        logging.info(\"Progress on genes: {} {}\".format(i+1, gene_table_name))\n",
    "    \n",
    "    data_gene = gc_logcpm.data.getrow(i).tocoo()\n",
    "    gene_table_name = gene_id_to_table_name(gene)\n",
    "    \n",
    "    df_gene = pd.DataFrame()\n",
    "    df_gene['cell_name'] = [gc_logcpm.cell[col] for col in data_gene.col]\n",
    "    df_gene['normalized_counts'] = data_gene.data\n",
    "    \n",
    "    df_gene = pd.merge(df_gene, df_cells, on='cell_name', how='left')\n",
    "    df_gene = df_gene[['cell_id', 'normalized_counts']]\n",
    "    print(gene, df_gene)\n",
    "    \n",
    "#     insert_into(engine, gene_table_name, df_gene, ignore=True, verbose=True)\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      cell_id  normalized_counts\n",
      "8567     8568             1.4893\n",
      "8568     8569             2.6315\n",
      "8569     8570             1.4147\n",
      "8570     8571             4.2155\n",
      "8571     8572             2.3922\n"
     ]
    }
   ],
   "source": [
    "sql = \"\"\"SELECT * FROM cells\"\"\"\n",
    "df_cells = pd.read_sql(sql, engine)\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "\n",
    "for i, (gene_id, row) in enumerate(df_res.iterrows()):\n",
    "    gene_table_name = gene_id_to_table_name(gene_id)\n",
    "#     if (i%10==0):\n",
    "#         logging.info(\"Progress on genes: {} {}\".format(i+1, gene_table_name))\n",
    "    if i == 0:\n",
    "        df_gene = row.to_frame('normalized_counts')\n",
    "        df_gene.index = dataset + '_' + df_gene.index.values\n",
    "        df_gene = pd.merge(df_gene, df_cells, left_index=True, right_on='cell_name', how='left')\n",
    "        df_gene = df_gene[['cell_id', 'normalized_counts']]\n",
    "        print(df_gene.head())\n",
    "\n",
    "#     if i == 0:\n",
    "#         df_gene = row.to_frame('normalized_counts')\n",
    "#         df_gene = pd.merge(df_gene, df_cells, left_index=True, right_on='cell_name', how='left')\n",
    "#         df_gene = df_gene[['cell_id', 'normalized_counts']]\n",
    "#         print(df_gene.head())\n",
    "#         insert_into(engine, gene_table_name, df_gene, ignore=True, verbose=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to ensembles table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlalchemy.engine.result.ResultProxy at 0x7fcc89f1aa58>"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Upload to ensembles table\n",
    "ens = 'Ens2'\n",
    "ens_name = 'CEMBA_3C_171207'\n",
    "ens_datasets=['CEMBA_3C_171207']\n",
    "\n",
    "\n",
    "table_name = 'ensembles'\n",
    "ens_id = int(ens[len('Ens'):])\n",
    "dict_list = [{'ensemble_id': ens_id, 'ensemble_name': ens_name, 'public_access': False, 'datasets': ','.join(ens_datasets)}]\n",
    "insert_into_worker(engine, table_name, dict_list, ignore=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create and upload to ensemble table (2 ensembles)\n",
    "\n",
    "def create_and_upload_to_ens(ens, df_ens, database):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    \n",
    "    logging.info(\"Creating {} table...\".format(ens))\n",
    "    engine = connect_sql(database)\n",
    "    # create ens table\n",
    "    metadata = sa.MetaData(engine)\n",
    "    # load specfic tables needed to referenced as foreign key\n",
    "    sa.Table('cells', metadata, autoload=True)\n",
    "    # define ensemble table\n",
    "    ens_table = define_ens_table(metadata, ens)\n",
    "    metadata.create_all()\n",
    "    logging.info(\"Done creating {} table!\".format(ens))\n",
    "\n",
    "    # upload to ens table\n",
    "    logging.info(\"Uploading to {} table...\".format(ens))\n",
    "\n",
    "    engine = connect_sql(database)\n",
    "    table_name = ens \n",
    "    insert_into(engine, table_name, df_ens, ignore=False, verbose=True)\n",
    "\n",
    "    logging.info(\"Done uploading to {} table!\".format(ens))\n",
    "\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/20/2018 09:59:55 PM Creating Ens2 table...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ens2\n",
      "      cell_id  cluster_ATAC  tsne_x_ATAC  tsne_y_ATAC\n",
      "8567     8568            13    -9.351734    30.328056\n",
      "8568     8569            19   -32.812104     4.672624\n",
      "8569     8570             8     3.887879    -2.134793\n",
      "8570     8571            17    32.098704    19.152496\n",
      "8571     8572            20    -1.280459    20.562413\n",
      "CEMBA_snATAC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/20/2018 09:59:56 PM Done creating Ens2 table!\n",
      "03/20/2018 09:59:56 PM Uploading to Ens2 table...\n",
      "03/20/2018 09:59:56 PM       cell_id  cluster_ATAC  tsne_x_ATAC  tsne_y_ATAC\n",
      "8567     8568            13    -9.351734    30.328056\n",
      "8568     8569            19   -32.812104     4.672624\n",
      "8569     8570             8     3.887879    -2.134793\n",
      "8570     8571            17    32.098704    19.152496\n",
      "8571     8572            20    -1.280459    20.562413\n",
      "03/20/2018 09:59:57 PM Done uploading to Ens2 table!\n"
     ]
    }
   ],
   "source": [
    "df_ens = pd.merge(df_clst, df_tsne, left_index=True, right_index=True)\n",
    "df_ens.columns = ['cluster_ATAC', 'tsne_x_ATAC', 'tsne_y_ATAC']\n",
    "df_ens.index = dataset + '_' + df_ens.index.values\n",
    "df_ens['cluster_ATAC'] = [int(cluster.strip('C')) for cluster in df_ens['cluster_ATAC']]\n",
    "\n",
    "# get cells\n",
    "sql = \"\"\"SELECT * FROM cells\"\"\"\n",
    "df_cells = pd.read_sql(sql, engine)[['cell_id', 'cell_name']]\n",
    "\n",
    "# get cell_id\n",
    "df_ens = pd.merge(df_cells, df_ens, left_on='cell_name', right_index=True, how='right')\n",
    "df_ens = df_ens.drop('cell_name', axis=1)\n",
    "\n",
    "print(ens)\n",
    "print(df_ens.head())\n",
    "print(database)\n",
    "create_and_upload_to_ens(ens, df_ens, database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # update a table\n",
    "# sql = \"\"\"SELECT * FROM cells\"\"\"\n",
    "# df_cells = pd.read_sql(sql, engine)\n",
    "# df_cells['cell_name'] = df_cells['dataset'].values[0] + '_' + df_cells['cell_name']\n",
    "\n",
    "# metadata = sa.MetaData(engine)\n",
    "# table = sa.Table('cells', metadata, autoload=True)\n",
    "# for i, row in df_cells.iterrows():\n",
    "#     updater = update(table).where(table.c.cell_id==row.cell_id).values(cell_name=row.cell_name)\n",
    "#     engine.execute(updater)\n",
    "\n",
    "\n",
    "\n",
    "# delete a table\n",
    "#     gene_table = sa.Table(gene_table_name, metadata, autoload=True)\n",
    "#     d = gene_table.delete()\n",
    "#     d.execute()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Delete zero rows (in gene table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "04/20/2018 11:37:15 AM Progress: 1\n",
      "04/20/2018 11:42:02 AM Progress: 1001\n",
      "04/20/2018 11:49:09 AM Progress: 2001\n",
      "04/20/2018 11:55:28 AM Progress: 3001\n",
      "04/20/2018 12:02:12 PM Progress: 4001\n",
      "04/20/2018 12:09:01 PM Progress: 5001\n",
      "04/20/2018 12:15:35 PM Progress: 6001\n",
      "04/20/2018 12:21:51 PM Progress: 7001\n",
      "04/20/2018 12:28:24 PM Progress: 8001\n",
      "04/20/2018 12:35:53 PM Progress: 9001\n",
      "04/20/2018 12:42:28 PM Progress: 10001\n",
      "04/20/2018 12:49:07 PM Progress: 11001\n",
      "04/20/2018 12:55:29 PM Progress: 12001\n",
      "04/20/2018 01:02:18 PM Progress: 13001\n"
     ]
    }
   ],
   "source": [
    "sql = \"SELECT * from genes\"\n",
    "df_genes = pd.read_sql(sql, engine)\n",
    "gene_tables = df_genes['gene_id'].apply(gene_id_to_table_name)\n",
    "\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "for i, gene_table in enumerate(gene_tables):\n",
    "    if i % 1000 == 0:\n",
    "        logging.info('Progress: {}'.format(i+1))\n",
    "        \n",
    "    table = sa.Table(gene_table, metadata, autoload=True)\n",
    "    d = table.delete(table.c.normalized_counts == 0)\n",
    "    d.execute()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Delete from a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/07/2018 04:00:39 PM Progress: 1\n",
      "08/07/2018 04:01:01 PM Progress: 1001\n",
      "08/07/2018 04:01:19 PM Progress: 2001\n",
      "08/07/2018 04:01:36 PM Progress: 3001\n",
      "08/07/2018 04:01:52 PM Progress: 4001\n",
      "08/07/2018 04:02:08 PM Progress: 5001\n",
      "08/07/2018 04:02:23 PM Progress: 6001\n",
      "08/07/2018 04:02:40 PM Progress: 7001\n",
      "08/07/2018 04:02:55 PM Progress: 8001\n",
      "08/07/2018 04:03:12 PM Progress: 9001\n",
      "08/07/2018 04:03:27 PM Progress: 10001\n",
      "08/07/2018 04:03:45 PM Progress: 11001\n",
      "08/07/2018 04:04:00 PM Progress: 12001\n",
      "08/07/2018 04:04:12 PM Progress: 13001\n",
      "08/07/2018 04:04:24 PM Progress: 14001\n",
      "08/07/2018 04:04:36 PM Progress: 15001\n",
      "08/07/2018 04:04:48 PM Progress: 16001\n",
      "08/07/2018 04:04:59 PM Progress: 17001\n",
      "08/07/2018 04:05:11 PM Progress: 18001\n",
      "08/07/2018 04:05:22 PM Progress: 19001\n",
      "08/07/2018 04:05:34 PM Progress: 20001\n",
      "08/07/2018 04:05:45 PM Progress: 21001\n",
      "08/07/2018 04:05:56 PM Progress: 22001\n",
      "08/07/2018 04:06:07 PM Progress: 23001\n",
      "08/07/2018 04:06:18 PM Progress: 24001\n",
      "08/07/2018 04:06:29 PM Progress: 25001\n",
      "08/07/2018 04:06:40 PM Progress: 26001\n",
      "08/07/2018 04:06:51 PM Progress: 27001\n",
      "08/07/2018 04:07:02 PM Progress: 28001\n",
      "08/07/2018 04:07:12 PM Progress: 29001\n",
      "08/07/2018 04:07:25 PM Progress: 30001\n",
      "08/07/2018 04:07:37 PM Progress: 31001\n",
      "08/07/2018 04:07:51 PM Progress: 32001\n",
      "08/07/2018 04:08:05 PM Progress: 33001\n",
      "08/07/2018 04:08:19 PM Progress: 34001\n",
      "08/07/2018 04:08:35 PM Progress: 35001\n",
      "08/07/2018 04:08:50 PM Progress: 36001\n",
      "08/07/2018 04:09:04 PM Progress: 37001\n",
      "08/07/2018 04:09:19 PM Progress: 38001\n",
      "08/07/2018 04:09:34 PM Progress: 39001\n",
      "08/07/2018 04:09:48 PM Progress: 40001\n",
      "08/07/2018 04:10:04 PM Progress: 41001\n",
      "08/07/2018 04:10:19 PM Progress: 42001\n",
      "08/07/2018 04:10:34 PM Progress: 43001\n",
      "08/07/2018 04:10:49 PM Progress: 44001\n",
      "08/07/2018 04:11:06 PM Progress: 45001\n",
      "08/07/2018 04:11:23 PM Progress: 46001\n",
      "08/07/2018 04:11:38 PM Progress: 47001\n",
      "08/07/2018 04:11:53 PM Progress: 48001\n",
      "08/07/2018 04:12:08 PM Progress: 49001\n",
      "08/07/2018 04:12:24 PM Progress: 50001\n",
      "08/07/2018 04:12:40 PM Progress: 51001\n",
      "08/07/2018 04:12:55 PM Progress: 52001\n",
      "08/07/2018 04:13:10 PM Progress: 53001\n"
     ]
    }
   ],
   "source": [
    "sql = \"SELECT * from genes\"\n",
    "df_genes = pd.read_sql(sql, engine)\n",
    "gene_tables = df_genes['gene_id'].apply(gene_id_to_table_name)\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "for i, gene_table in enumerate(gene_tables):\n",
    "    if i % 1000 == 0:\n",
    "        logging.info('Progress: {}'.format(i+1))\n",
    "        \n",
    "    table = sa.Table(gene_table, metadata, autoload=True)\n",
    "    d = table.delete()\n",
    "    d.execute()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Update ensemble (old)\n",
    "\n",
    "- update to enss\n",
    "- create and update Ens "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensemble_id</th>\n",
       "      <th>ensemble_name</th>\n",
       "      <th>public_access</th>\n",
       "      <th>datasets</th>\n",
       "      <th>description</th>\n",
       "      <th>snmc_ensemble_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>CEMBA_3C_171206</td>\n",
       "      <td>0</td>\n",
       "      <td>CEMBA_3C_171206</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>CEMBA_3C_171207</td>\n",
       "      <td>0</td>\n",
       "      <td>CEMBA_3C_171207</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100</td>\n",
       "      <td>Multimodal_test</td>\n",
       "      <td>0</td>\n",
       "      <td>CEMBA_3C_171206,CEMBA_3C_171207,CEMBA_4B_17121...</td>\n",
       "      <td>Test multimodal visualization</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensemble_id    ensemble_name  public_access  \\\n",
       "0            1  CEMBA_3C_171206              0   \n",
       "1            2  CEMBA_3C_171207              0   \n",
       "2          100  Multimodal_test              0   \n",
       "\n",
       "                                            datasets  \\\n",
       "0                                    CEMBA_3C_171206   \n",
       "1                                    CEMBA_3C_171207   \n",
       "2  CEMBA_3C_171206,CEMBA_3C_171207,CEMBA_4B_17121...   \n",
       "\n",
       "                     description  snmc_ensemble_id  \n",
       "0                           None                 1  \n",
       "1                           None                 2  \n",
       "2  Test multimodal visualization               100  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "engine_old = connect_sql('CEMBA_snATAC')\n",
    "\n",
    "sql = 'SELECT * from ensembles'\n",
    "df = pd.read_sql(sql, engine_old)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/08/2018 02:28:23 PM    ensemble_id    ensemble_name  public_access  \\\n",
      "0            1  CEMBA_3C_171206              0   \n",
      "1            2  CEMBA_3C_171207              0   \n",
      "2          100  Multimodal_test              0   \n",
      "\n",
      "                                            datasets  \\\n",
      "0                                    CEMBA_3C_171206   \n",
      "1                                    CEMBA_3C_171207   \n",
      "2  CEMBA_3C_171206,CEMBA_3C_171207,CEMBA_4B_17121...   \n",
      "\n",
      "                     description  snmc_ensemble_id  \n",
      "0                           None                 1  \n",
      "1                           None                 2  \n",
      "2  Test multimodal visualization               100  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<sqlalchemy.engine.result.ResultProxy at 0x7f7727e7e0b8>"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "insert_into(engine, 'ensembles', df, ignore=False, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/10/2018 09:58:01 AM Creating Ens2 table...\n",
      "08/10/2018 09:58:01 AM Done creating Ens2 table!\n",
      "08/10/2018 09:58:01 AM Uploading to Ens2 table...\n",
      "08/10/2018 09:58:01 AM    tsne_x_ATAC  tsne_y_ATAC  cluster_ATAC annotation_ATAC  cell_id\n",
      "0     -9.35173     30.32810            13            None    92411\n",
      "1    -32.81210      4.67262            19            None    92412\n",
      "2      3.88788     -2.13479             8            None    92413\n",
      "3     32.09870     19.15250            17            None    92415\n",
      "4     -1.28046     20.56240            20            None    92416\n",
      "08/10/2018 09:58:02 AM Done uploading to Ens2 table!\n"
     ]
    }
   ],
   "source": [
    "engine_old = connect_sql('CEMBA_snATAC')\n",
    "engine= connect_sql('CEMBA_ATAC')\n",
    "\n",
    "sql = 'SELECT * FROM Ens2 JOIN cells ON cells.cell_id = Ens2.cell_id'\n",
    "df = pd.read_sql(sql, engine_old).drop(['cell_id', 'dataset'], axis=1)\n",
    "df['cell_name'] = [cell_name[16:]+'_'+cell_name[:15] for cell_name in df['cell_name']]\n",
    "\n",
    "sql = 'SELECT * FROM cells'\n",
    "df2 = pd.read_sql(sql, engine) \n",
    "df = pd.merge(df, df2, on='cell_name').drop(['cell_name', 'dataset'], axis=1)\n",
    "df.head()\n",
    "\n",
    "create_and_upload_to_ens('Ens2', df, 'CEMBA_ATAC')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Update ensemble (new)\n",
    "\n",
    "- update to enss\n",
    "- create and update Ens "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[51]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "ens_ids = sorted([int(item.split('_')[2][3:-4]) for item in \n",
    " filter(lambda x: (re.search(r'^tsne', x)), os.listdir('/cndd/fangming/CEMBA/ATAC'))])\n",
    "ens_ids\n",
    "\n",
    "engine = connect_sql('CEMBA_ATAC')\n",
    "sql = 'SELECT * FROM ensembles'\n",
    "ensemble_info = pd.read_sql(sql, engine)\n",
    "ens_ids = [ens_id for ens_id in ens_ids if ens_id not in ensemble_info['ensemble_id'].values]\n",
    "ens_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# ens_ids = [\n",
    "#     3, \n",
    "#     4, \n",
    "#     5, \n",
    "#     6, \n",
    "#     7, \n",
    "#     8, \n",
    "#     9, \n",
    "#     10, \n",
    "# ]\n",
    "\n",
    "\n",
    "engine = connect_sql('CEMBA')\n",
    "sql = 'SELECT * FROM ensembles'\n",
    "ensemble_info = pd.read_sql(sql, engine)\n",
    "ensemble_info = ensemble_info.set_index('ensemble_id').loc[ens_ids, :].reset_index()\n",
    "ensemble_info['ens'] = ensemble_info['ensemble_id'].apply(lambda x: 'Ens{}'.format(x)) \n",
    "ensemble_info['snmc_ensemble_id'] = ensemble_info['ensemble_id'] \n",
    "ensemble_info = ensemble_info[['ens', 'ensemble_name', 'datasets', 'snmc_ensemble_id']]\n",
    "ensemble_info\n",
    "\n",
    "\n",
    "\n",
    "for idx, (ens, ens_name, ens_datasets, snmc_ens_id) in ensemble_info.iterrows():\n",
    "\n",
    "    # update to ensembles\n",
    "    table_name = 'ensembles'\n",
    "    ens_id = int(ens[len('Ens'):])\n",
    "    dict_list = [{'ensemble_id': ens_id, \n",
    "                  'ensemble_name': ens_name, \n",
    "                  'public_access': False, \n",
    "                  'datasets': ens_datasets, \n",
    "                  'snmc_ensemble_id': snmc_ens_id, \n",
    "                 }]\n",
    "    engine= connect_sql('CEMBA_ATAC')\n",
    "    insert_into_worker(engine, table_name, dict_list, ignore=False)\n",
    "\n",
    "    \n",
    "    # create and update Ens\n",
    "    dirc = '/cndd/fangming/CEMBA/ATAC/'\n",
    "    f1 = dirc + 'tsne_atac_{}.tsv'.format(ens)\n",
    "    df1 = pd.read_table(f1)\n",
    "    f2 = dirc + 'cluster_atac_{}.tsv'.format(ens)\n",
    "    df2 = pd.read_table(f2)\n",
    "    df = pd.merge(df1, df2, on='sample').rename(columns={'sample': 'cell_name'})\n",
    "\n",
    "\n",
    "    sql = 'SELECT * FROM cells'\n",
    "    df2 = pd.read_sql(sql, engine) \n",
    "    df = (pd.merge(df, df2, on='cell_name')\n",
    "          .drop(['cell_name', 'dataset'], axis=1)\n",
    "          .rename(columns={\n",
    "              'tsne_x': 'tsne_x_ATAC', \n",
    "              'tsne_y': 'tsne_y_ATAC', \n",
    "              'cluster_ID': 'cluster_ATAC', \n",
    "          }))\n",
    "    df['cluster_ATAC'] = df['cluster_ATAC'].apply(lambda x: int(x[len('cluster_'):]))\n",
    "\n",
    "    create_and_upload_to_ens(ens, df, 'CEMBA_ATAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_ATAC</th>\n",
       "      <th>tsne_x_ATAC</th>\n",
       "      <th>tsne_y_ATAC</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cluster_15</td>\n",
       "      <td>-10.438904</td>\n",
       "      <td>-34.038421</td>\n",
       "      <td>82477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cluster_2</td>\n",
       "      <td>23.816641</td>\n",
       "      <td>19.874356</td>\n",
       "      <td>82479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cluster_13</td>\n",
       "      <td>23.799659</td>\n",
       "      <td>4.783634</td>\n",
       "      <td>82480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cluster_1</td>\n",
       "      <td>-36.239155</td>\n",
       "      <td>5.543818</td>\n",
       "      <td>82484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cluster_10</td>\n",
       "      <td>-19.669854</td>\n",
       "      <td>-25.038445</td>\n",
       "      <td>82485</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cluster_ATAC  tsne_x_ATAC  tsne_y_ATAC  cell_id\n",
       "0   cluster_15   -10.438904   -34.038421    82477\n",
       "1    cluster_2    23.816641    19.874356    82479\n",
       "2   cluster_13    23.799659     4.783634    82480\n",
       "3    cluster_1   -36.239155     5.543818    82484\n",
       "4   cluster_10   -19.669854   -25.038445    82485"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TEMPORARY!!!!!!!!\n",
    "\n",
    "# f = '/cndd/fangming/CEMBA/annoj_browser/multimodal_v2/cell_info.tsv'\n",
    "# ens = 'Ens100'\n",
    "# df = pd.read_table(f)\n",
    "# df = df[df['modality']=='atac']\n",
    "# df = df.drop('modality', axis=1).rename(columns={'tsne_x': 'tsne_x_ATAC', \n",
    "#                                             'tsne_y': 'tsne_y_ATAC',\n",
    "#                                             'cluster_ID': 'cluster_ATAC',\n",
    "#                                             'sample': 'cell_name',\n",
    "#                                            })\n",
    "# sql = 'SELECT * FROM cells'\n",
    "# df2 = pd.read_sql(sql, engine)\n",
    "# df = pd.merge(df, df2, on='cell_name').drop(['dataset', 'cell_name'], axis=1)\n",
    "\n",
    "# create_and_upload_to_ens(ens, df, 'CEMBA_ATAC')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## for RNA seq "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __init__ import *\n",
    "import snmcseq_utils\n",
    "# import importlib\n",
    "# importlib.reload(snmcseq_utils)\n",
    "from collections import namedtuple\n",
    "from scipy import sparse\n",
    "\n",
    "GC_matrix = namedtuple('GC_matrix', ['gene', 'cell', 'data'])\n",
    "\n",
    "def sparse_logtpm(gc_matrix, gene_lengths):\n",
    "    \"\"\"\n",
    "    gene_lengths: array like \n",
    "    \n",
    "    \"\"\"\n",
    "    gene_lengths = np.array(gene_lengths)\n",
    "    gene_length_inv = sparse.diags(np.ravel(1.0/gene_lengths))\n",
    "    tmp = (gene_length_inv).dot(gc_matrix.data).tocoo()\n",
    "    lib_size_inv = sparse.diags(np.ravel(1.0/tmp.sum(axis=0)))\n",
    "    \n",
    "    logtpm = tmp.dot(lib_size_inv*1e6).tocoo()\n",
    "    logtpm.data = np.log10(logtpm.data + 1)\n",
    "\n",
    "    gc_logtpm = GC_matrix(\n",
    "        gc_matrix.gene, \n",
    "        gc_matrix.cell, \n",
    "        logtpm,\n",
    "    )\n",
    "    \n",
    "    return gc_logtpm\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102693.1</th>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>+</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064842.1</th>\n",
       "      <td>Gm26206</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>+</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951.5</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102851.1</th>\n",
       "      <td>Gm18956</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3252757</td>\n",
       "      <td>3253236</td>\n",
       "      <td>+</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000103377.1</th>\n",
       "      <td>Gm37180</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3365731</td>\n",
       "      <td>3368549</td>\n",
       "      <td>-</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_name   chr    start      end strand  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000102693.1  4933401J01Rik  chr1  3073253  3074322      +   \n",
       "ENSMUSG00000064842.1        Gm26206  chr1  3102016  3102125      +   \n",
       "ENSMUSG00000051951.5           Xkr4  chr1  3205901  3671498      -   \n",
       "ENSMUSG00000102851.1        Gm18956  chr1  3252757  3253236      +   \n",
       "ENSMUSG00000103377.1        Gm37180  chr1  3365731  3368549      -   \n",
       "\n",
       "                                 gene_type  \n",
       "gene_id                                     \n",
       "ENSMUSG00000102693.1                   TEC  \n",
       "ENSMUSG00000064842.1                 snRNA  \n",
       "ENSMUSG00000051951.5        protein_coding  \n",
       "ENSMUSG00000102851.1  processed_pseudogene  \n",
       "ENSMUSG00000103377.1                   TEC  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes = pd.read_table(PATH_GENEBODY_ANNOTATION, index_col='gene_id')\n",
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53278, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>ENSMUSG00000109644.1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>45567794</td>\n",
       "      <td>45575327</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>ENSMUSG00000108652.1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>74818818</td>\n",
       "      <td>74853813</td>\n",
       "      <td>-</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>ENSMUSG00000007777.9</td>\n",
       "      <td>chr11</td>\n",
       "      <td>51685386</td>\n",
       "      <td>51688874</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>ENSMUSG00000086714.1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>26445696</td>\n",
       "      <td>26459390</td>\n",
       "      <td>+</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>ENSMUSG00000043644.4</td>\n",
       "      <td>chr11</td>\n",
       "      <td>120348678</td>\n",
       "      <td>120351190</td>\n",
       "      <td>+</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            gene_id    chr      start        end strand  \\\n",
       "gene_name                                                                 \n",
       "0610005C13Rik  ENSMUSG00000109644.1   chr7   45567794   45575327      -   \n",
       "0610006L08Rik  ENSMUSG00000108652.1   chr7   74818818   74853813      -   \n",
       "0610009B22Rik  ENSMUSG00000007777.9  chr11   51685386   51688874      -   \n",
       "0610009E02Rik  ENSMUSG00000086714.1   chr2   26445696   26459390      +   \n",
       "0610009L18Rik  ENSMUSG00000043644.4  chr11  120348678  120351190      +   \n",
       "\n",
       "                                   gene_type  \n",
       "gene_name                                     \n",
       "0610005C13Rik                  antisense_RNA  \n",
       "0610006L08Rik                        lincRNA  \n",
       "0610009B22Rik                 protein_coding  \n",
       "0610009E02Rik           processed_transcript  \n",
       "0610009L18Rik  bidirectional_promoter_lncRNA  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes_v2 = df_genes.reset_index().groupby('gene_name').first()\n",
    "print(df_genes_v2.shape)\n",
    "df_genes_v2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = '/cndd/Public_Datasets/CEMBA/BICCN_minibrain_data/Zeng/SMARTer_nuclei_MOp/exon_counts.csv.gz'\n",
    "df = pd.read_csv(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample_id</th>\n",
       "      <th>SM-GE653_S113_E1-50</th>\n",
       "      <th>SM-GE653_S114_E1-50</th>\n",
       "      <th>SM-GE653_S115_E1-50</th>\n",
       "      <th>SM-GE653_S116_E1-50</th>\n",
       "      <th>SM-GE653_S117_E1-50</th>\n",
       "      <th>SM-GE653_S118_E1-50</th>\n",
       "      <th>SM-GE653_S119_E1-50</th>\n",
       "      <th>SM-GE653_S120_E1-50</th>\n",
       "      <th>SM-GE653_S121_E1-50</th>\n",
       "      <th>SM-GE653_S122_E1-50</th>\n",
       "      <th>...</th>\n",
       "      <th>SM-GE66H_S079_E1-50</th>\n",
       "      <th>SM-GE66H_S080_E1-50</th>\n",
       "      <th>SM-GE66H_S081_E1-50</th>\n",
       "      <th>SM-GE66H_S082_E1-50</th>\n",
       "      <th>SM-GE66H_S083_E1-50</th>\n",
       "      <th>SM-GE66H_S084_E1-50</th>\n",
       "      <th>SM-GE66H_S085_E1-50</th>\n",
       "      <th>SM-GE66H_S086_E1-50</th>\n",
       "      <th>SM-GE66H_S087_E1-50</th>\n",
       "      <th>SM-GE66H_S088_E1-50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "      <td>1</td>\n",
       "      <td>301</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6848 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample_id      SM-GE653_S113_E1-50  SM-GE653_S114_E1-50  SM-GE653_S115_E1-50  \\\n",
       "0610005C13Rik                    0                    0                    0   \n",
       "0610006L08Rik                    0                    0                    0   \n",
       "0610007P14Rik                  136                    0                    0   \n",
       "0610009B22Rik                    0                    0                  136   \n",
       "0610009E02Rik                    0                    0                    0   \n",
       "\n",
       "sample_id      SM-GE653_S116_E1-50  SM-GE653_S117_E1-50  SM-GE653_S118_E1-50  \\\n",
       "0610005C13Rik                    0                    0                    0   \n",
       "0610006L08Rik                    0                    0                    0   \n",
       "0610007P14Rik                    0                   25                    0   \n",
       "0610009B22Rik                    0                    0                    0   \n",
       "0610009E02Rik                    0                    0                    0   \n",
       "\n",
       "sample_id      SM-GE653_S119_E1-50  SM-GE653_S120_E1-50  SM-GE653_S121_E1-50  \\\n",
       "0610005C13Rik                    0                    0                    0   \n",
       "0610006L08Rik                    0                    0                    0   \n",
       "0610007P14Rik                  125                    1                  301   \n",
       "0610009B22Rik                    0                    0                    0   \n",
       "0610009E02Rik                    0                    0                    0   \n",
       "\n",
       "sample_id      SM-GE653_S122_E1-50         ...           SM-GE66H_S079_E1-50  \\\n",
       "0610005C13Rik                    0         ...                             0   \n",
       "0610006L08Rik                    0         ...                             0   \n",
       "0610007P14Rik                    0         ...                             0   \n",
       "0610009B22Rik                    0         ...                             0   \n",
       "0610009E02Rik                    0         ...                             0   \n",
       "\n",
       "sample_id      SM-GE66H_S080_E1-50  SM-GE66H_S081_E1-50  SM-GE66H_S082_E1-50  \\\n",
       "0610005C13Rik                    0                    0                    0   \n",
       "0610006L08Rik                    0                    0                    0   \n",
       "0610007P14Rik                    0                    0                    0   \n",
       "0610009B22Rik                    0                    0                    0   \n",
       "0610009E02Rik                    0                    0                    0   \n",
       "\n",
       "sample_id      SM-GE66H_S083_E1-50  SM-GE66H_S084_E1-50  SM-GE66H_S085_E1-50  \\\n",
       "0610005C13Rik                    0                   70                    0   \n",
       "0610006L08Rik                    0                    0                    0   \n",
       "0610007P14Rik                  102                    0                    1   \n",
       "0610009B22Rik                    0                   74                    0   \n",
       "0610009E02Rik                    0                    0                    0   \n",
       "\n",
       "sample_id      SM-GE66H_S086_E1-50  SM-GE66H_S087_E1-50  SM-GE66H_S088_E1-50  \n",
       "0610005C13Rik                    0                    0                    0  \n",
       "0610006L08Rik                    0                    0                    0  \n",
       "0610007P14Rik                   85                    0                  179  \n",
       "0610009B22Rik                    0                    0                    0  \n",
       "0610009E02Rik                    0                    0                    0  \n",
       "\n",
       "[5 rows x 6848 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.set_index('sample_id') \n",
    "df = df.T\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['0610005C13Rik' '0610006L08Rik' '0610007P14Rik' ... 'n-R5s144' 'n-R5s146'\n",
      " 'n-R5s149']\n",
      "['SM-GE653_S113_E1-50' 'SM-GE653_S114_E1-50' 'SM-GE653_S115_E1-50' ...\n",
      " 'SM-GE66H_S086_E1-50' 'SM-GE66H_S087_E1-50' 'SM-GE66H_S088_E1-50']\n"
     ]
    }
   ],
   "source": [
    "gene_names = df.index.values\n",
    "cells = df.columns.values \n",
    "print(gene_names)\n",
    "print(cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get non-duplicated gene list\n",
    "gene_ids = [snmcseq_utils.gene_name_to_id(name, df_genes_v2) for name in gene_names]\n",
    "\n",
    "good_idx = [i for i in range(len(gene_ids)) if gene_ids[i]]\n",
    "gene_names = np.array(gene_names)[good_idx]\n",
    "gene_ids = np.array(gene_ids)[good_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324, 6848)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample_id</th>\n",
       "      <th>SM-GE653_S113_E1-50</th>\n",
       "      <th>SM-GE653_S114_E1-50</th>\n",
       "      <th>SM-GE653_S115_E1-50</th>\n",
       "      <th>SM-GE653_S116_E1-50</th>\n",
       "      <th>SM-GE653_S117_E1-50</th>\n",
       "      <th>SM-GE653_S118_E1-50</th>\n",
       "      <th>SM-GE653_S119_E1-50</th>\n",
       "      <th>SM-GE653_S120_E1-50</th>\n",
       "      <th>SM-GE653_S121_E1-50</th>\n",
       "      <th>SM-GE653_S122_E1-50</th>\n",
       "      <th>...</th>\n",
       "      <th>SM-GE66H_S079_E1-50</th>\n",
       "      <th>SM-GE66H_S080_E1-50</th>\n",
       "      <th>SM-GE66H_S081_E1-50</th>\n",
       "      <th>SM-GE66H_S082_E1-50</th>\n",
       "      <th>SM-GE66H_S083_E1-50</th>\n",
       "      <th>SM-GE66H_S084_E1-50</th>\n",
       "      <th>SM-GE66H_S085_E1-50</th>\n",
       "      <th>SM-GE66H_S086_E1-50</th>\n",
       "      <th>SM-GE66H_S087_E1-50</th>\n",
       "      <th>SM-GE66H_S088_E1-50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000109644.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000108652.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000007777.9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000086714.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000043644.4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6848 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample_id             SM-GE653_S113_E1-50  SM-GE653_S114_E1-50  \\\n",
       "ENSMUSG00000109644.1                    0                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                    0                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id             SM-GE653_S115_E1-50  SM-GE653_S116_E1-50  \\\n",
       "ENSMUSG00000109644.1                    0                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                  136                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id             SM-GE653_S117_E1-50  SM-GE653_S118_E1-50  \\\n",
       "ENSMUSG00000109644.1                    0                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                    0                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id             SM-GE653_S119_E1-50  SM-GE653_S120_E1-50  \\\n",
       "ENSMUSG00000109644.1                    0                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                    0                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id             SM-GE653_S121_E1-50  SM-GE653_S122_E1-50  \\\n",
       "ENSMUSG00000109644.1                    0                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                    0                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id                    ...           SM-GE66H_S079_E1-50  \\\n",
       "ENSMUSG00000109644.1         ...                             0   \n",
       "ENSMUSG00000108652.1         ...                             0   \n",
       "ENSMUSG00000007777.9         ...                             0   \n",
       "ENSMUSG00000086714.1         ...                             0   \n",
       "ENSMUSG00000043644.4         ...                             0   \n",
       "\n",
       "sample_id             SM-GE66H_S080_E1-50  SM-GE66H_S081_E1-50  \\\n",
       "ENSMUSG00000109644.1                    0                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                    0                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id             SM-GE66H_S082_E1-50  SM-GE66H_S083_E1-50  \\\n",
       "ENSMUSG00000109644.1                    0                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                    0                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id             SM-GE66H_S084_E1-50  SM-GE66H_S085_E1-50  \\\n",
       "ENSMUSG00000109644.1                   70                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                   74                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id             SM-GE66H_S086_E1-50  SM-GE66H_S087_E1-50  \\\n",
       "ENSMUSG00000109644.1                    0                    0   \n",
       "ENSMUSG00000108652.1                    0                    0   \n",
       "ENSMUSG00000007777.9                    0                    0   \n",
       "ENSMUSG00000086714.1                    0                    0   \n",
       "ENSMUSG00000043644.4                    0                    0   \n",
       "\n",
       "sample_id             SM-GE66H_S088_E1-50  \n",
       "ENSMUSG00000109644.1                    0  \n",
       "ENSMUSG00000108652.1                    0  \n",
       "ENSMUSG00000007777.9                    0  \n",
       "ENSMUSG00000086714.1                    0  \n",
       "ENSMUSG00000043644.4                    0  \n",
       "\n",
       "[5 rows x 6848 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.loc[gene_names, :]\n",
    "df.index = gene_ids\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_counts = GC_matrix(\n",
    "    gene=df.index, \n",
    "    cell=df.columns,\n",
    "    data=sparse.coo_matrix(df.values),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_lengths = (df_genes['end'] - df_genes['start']).loc[gc_counts.gene].values\n",
    "gc_logtpm = sparse_logtpm(gc_counts, gene_lengths)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA seq ensemble upload "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_and_upload_to_ens(ens, df_ens, database):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    \n",
    "    logging.info(\"Creating {} table...\".format(ens))\n",
    "    engine = connect_sql(database)\n",
    "    # create ens table\n",
    "    metadata = sa.MetaData(engine)\n",
    "    # load specfic tables needed to referenced as foreign key\n",
    "    sa.Table('cells', metadata, autoload=True)\n",
    "    # define ensemble table\n",
    "    ens_table = define_ens_table(metadata, ens)\n",
    "    metadata.create_all()\n",
    "    logging.info(\"Done creating {} table!\".format(ens))\n",
    "\n",
    "    # upload to ens table\n",
    "    logging.info(\"Uploading to {} table...\".format(ens))\n",
    "\n",
    "    engine = connect_sql(database)\n",
    "    table_name = ens \n",
    "    insert_into(engine, table_name, df_ens, ignore=False, verbose=True)\n",
    "\n",
    "    logging.info(\"Done uploading to {} table!\".format(ens))\n",
    "\n",
    "    return\n",
    "\n",
    "def define_ens_table(metadata, ens):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # prototypes\n",
    "#   sa.Column('tsne_x_mCH', sa.Float, nullable=True),\n",
    "#   sa.Column('tsne_y_mCH', sa.Float, nullable=True),\n",
    "#   sa.Column('cluster', SMALLINT(unsigned=True), nullable=True),\n",
    "#   sa.Column('annotation', sa.String(20), nullable=True),\n",
    "\n",
    "    # tsnes\n",
    "    contexts = ['RNA'] \n",
    "    perps = PERPLEXITIES\n",
    "#     tsne_types = ['{}_ndim2_perp{}'.format(context, p) for (context, p) in product(contexts, perps)]\n",
    "    tsne_types = ['RNA']\n",
    "    \n",
    "    # prepare columns\n",
    "    args_tsne = ([sa.Column('tsne_x_{}'.format(tsne_type), sa.Float, nullable=True) \n",
    "            for tsne_type in tsne_types] \n",
    "            + [sa.Column('tsne_y_{}'.format(tsne_type), sa.Float, nullable=True) \n",
    "            for tsne_type in tsne_types]\n",
    "           )\n",
    "    \n",
    "    # clusters\n",
    "    ks = K_NN\n",
    "#     cluster_types = ['{}_lv_npc50_k{}'.format(context, k) for (context, k) in product(contexts, ks)]\n",
    "    cluster_types = ['RNA']\n",
    "    # prepare columns\n",
    "    args_cluster = ([sa.Column('cluster_{}'.format(cluster_type), SMALLINT(unsigned=True), nullable=True) \n",
    "            for cluster_type in cluster_types] \n",
    "            + [sa.Column('annotation_{}'.format(cluster_type), sa.String(20), nullable=True) \n",
    "            for cluster_type in cluster_types]\n",
    "           )\n",
    "    \n",
    "    # combine cols\n",
    "    args = args_tsne + args_cluster\n",
    "    # create the table\n",
    "    table = sa.Table(ens, metadata,\n",
    "                        # NOT auto increment not null primary key\n",
    "                        sa.Column('cell_id', sa.Integer, sa.ForeignKey('cells.cell_id'), primary_key=True, autoincrement=False), \n",
    "                        *args\n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6083, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_RNA</th>\n",
       "      <th>tsne_x_RNA</th>\n",
       "      <th>tsne_y_RNA</th>\n",
       "      <th>modality</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28</td>\n",
       "      <td>25.665414</td>\n",
       "      <td>-24.553544</td>\n",
       "      <td>snrna</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>28</td>\n",
       "      <td>23.874565</td>\n",
       "      <td>-21.141237</td>\n",
       "      <td>snrna</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34</td>\n",
       "      <td>20.783336</td>\n",
       "      <td>-6.606048</td>\n",
       "      <td>snrna</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>28</td>\n",
       "      <td>24.488323</td>\n",
       "      <td>-24.852014</td>\n",
       "      <td>snrna</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>2.924461</td>\n",
       "      <td>-19.134132</td>\n",
       "      <td>snrna</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cluster_RNA  tsne_x_RNA  tsne_y_RNA modality  cell_id\n",
       "0           28   25.665414  -24.553544    snrna        1\n",
       "1           28   23.874565  -21.141237    snrna        2\n",
       "2           34   20.783336   -6.606048    snrna        3\n",
       "3           28   24.488323  -24.852014    snrna        5\n",
       "4           31    2.924461  -19.134132    snrna        6"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# process data (ensemble table)\n",
    "\n",
    "ens = 'Ens100'\n",
    "\n",
    "dirc = '/cndd/fangming/CEMBA/RNA/'\n",
    "f = dirc + '{}/cell_info.tsv'.format(ens)\n",
    "df = pd.read_table(f).rename(columns={'sample': 'cell_name'})\n",
    "df = df[df['modality'].isin(['snrna'])]\n",
    "\n",
    "\n",
    "sql = 'SELECT * FROM cells'\n",
    "engine = connect_sql(DATABASE_RNA)\n",
    "df2 = pd.read_sql(sql, engine) \n",
    "df = (pd.merge(df, df2, on='cell_name')\n",
    "      .drop(['cell_name', 'dataset'], axis=1)\n",
    "      .rename(columns={\n",
    "          'tsne_x': 'tsne_x_RNA', \n",
    "          'tsne_y': 'tsne_y_RNA', \n",
    "          'cluster_ID': 'cluster_RNA', \n",
    "      }))\n",
    "df['cluster_RNA'] = df['cluster_RNA'].apply(lambda x: int(x[len('cluster_'):]))\n",
    "\n",
    "df_100 = df\n",
    "print(df.shape)\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/30/2018 04:18:26 PM Creating Ens100 table...\n",
      "08/30/2018 04:18:26 PM Done creating Ens100 table!\n",
      "08/30/2018 04:18:26 PM Uploading to Ens100 table...\n",
      "08/30/2018 04:18:26 PM    cluster_RNA  tsne_x_RNA  tsne_y_RNA modality  cell_id\n",
      "0           28   25.665414  -24.553544    snrna        1\n",
      "1           28   23.874565  -21.141237    snrna        2\n",
      "2           34   20.783336   -6.606048    snrna        3\n",
      "3           28   24.488323  -24.852014    snrna        5\n",
      "4           31    2.924461  -19.134132    snrna        6\n",
      "08/30/2018 04:18:27 PM Done uploading to Ens100 table!\n"
     ]
    }
   ],
   "source": [
    "ens_ids = [100]\n",
    "new_datasets = [['Zeng_SMARTer_nuclei_MOp']]\n",
    "dfs = [df_100]\n",
    "\n",
    "# check if the ensemble is there in methylation\n",
    "engine = connect_sql(DATABASE)\n",
    "sql = 'SELECT * FROM ensembles'\n",
    "ensemble_info = pd.read_sql(sql, engine)\n",
    "ensemble_info = ensemble_info.set_index('ensemble_id').loc[ens_ids, :].reset_index()\n",
    "ensemble_info['ens'] = ensemble_info['ensemble_id'].apply(lambda x: 'Ens{}'.format(x)) \n",
    "ensemble_info['snmc_ensemble_id'] = ensemble_info['ensemble_id'] \n",
    "ensemble_info = ensemble_info.set_index('ensemble_id')[['ens', 'ensemble_name', 'datasets', 'snmc_ensemble_id']]\n",
    "ensemble_info['data'] = dfs\n",
    "\n",
    "# check if the new datasets are available in CEMBA_RNA\n",
    "engine = connect_sql(DATABASE_RNA)\n",
    "sql = 'SELECT dataset FROM datasets'\n",
    "dataset_info = pd.read_sql(sql, engine)\n",
    "for dataset in np.ravel(new_datasets):\n",
    "    assert dataset in dataset_info['dataset'].values\n",
    "\n",
    "# add new dataset into ensemble information\n",
    "for ens_id, new_dataset in zip(ens_ids, new_datasets):\n",
    "    ensemble_info.loc[ens_id, 'datasets'] += ',{}'.format('_'.join(new_dataset))\n",
    "\n",
    "for idx, (ens, ens_name, ens_datasets, snmc_ens_id, df_data) in ensemble_info.iterrows():\n",
    "\n",
    "    # update to ensembles\n",
    "    table_name = 'ensembles'\n",
    "    ens_id = int(ens[len('Ens'):])\n",
    "    dict_list = [{'ensemble_id': ens_id, \n",
    "                  'ensemble_name': ens_name, \n",
    "                  'public_access': False, \n",
    "                  'datasets': ens_datasets, \n",
    "                  'snmc_ensemble_id': snmc_ens_id, \n",
    "                 }]\n",
    "    engine= connect_sql(DATABASE_RNA)\n",
    "    insert_into_worker(engine, table_name, dict_list, ignore=False)\n",
    "\n",
    "    # create and upload to ensemble table\n",
    "    create_and_upload_to_ens(ens, df_data, DATABASE_RNA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# change table label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_column(engine, table_name, column):\n",
    "    column_name = column.compile(dialect=engine.dialect)\n",
    "    column_type = column.type.compile(engine.dialect)\n",
    "    engine.execute('ALTER TABLE %s ADD COLUMN %s %s' % (table_name, column_name, column_type))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/31/2018 12:38:28 PM Progress: 1 gene_ENSMUSG00000000001_4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_ENSMUSG00000000001_4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08/31/2018 12:46:14 PM Progress: 1001 gene_ENSMUSG00000006423_15\n",
      "08/31/2018 12:53:39 PM Progress: 2001 gene_ENSMUSG00000018820_13\n",
      "08/31/2018 01:02:10 PM Progress: 3001 gene_ENSMUSG00000021253_7\n",
      "08/31/2018 01:10:47 PM Progress: 4001 gene_ENSMUSG00000022978_11\n",
      "08/31/2018 01:17:52 PM Progress: 5001 gene_ENSMUSG00000025357_8\n",
      "08/31/2018 01:26:10 PM Progress: 6001 gene_ENSMUSG00000027248_13\n",
      "08/31/2018 01:34:15 PM Progress: 7001 gene_ENSMUSG00000028954_11\n",
      "08/31/2018 01:41:35 PM Progress: 8001 gene_ENSMUSG00000030745_9\n",
      "08/31/2018 01:49:31 PM Progress: 9001 gene_ENSMUSG00000032494_12\n",
      "08/31/2018 01:57:51 PM Progress: 10001 gene_ENSMUSG00000035325_16\n",
      "08/31/2018 02:06:10 PM Progress: 11001 gene_ENSMUSG00000038152_4\n",
      "08/31/2018 02:14:06 PM Progress: 12001 gene_ENSMUSG00000040658_6\n",
      "08/31/2018 02:21:14 PM Progress: 13001 gene_ENSMUSG00000043986_5\n",
      "08/31/2018 02:27:00 PM Progress: 14001 gene_ENSMUSG00000047495_15\n",
      "08/31/2018 02:32:46 PM Progress: 15001 gene_ENSMUSG00000051185_9\n",
      "08/31/2018 02:39:03 PM Progress: 16001 gene_ENSMUSG00000055700_7\n",
      "08/31/2018 02:44:39 PM Progress: 17001 gene_ENSMUSG00000060201_4\n",
      "08/31/2018 02:49:51 PM Progress: 18001 gene_ENSMUSG00000064442_1\n",
      "08/31/2018 02:51:40 PM Progress: 19001 gene_ENSMUSG00000066389_1\n",
      "08/31/2018 02:55:59 PM Progress: 20001 gene_ENSMUSG00000071203_6\n",
      "08/31/2018 03:00:03 PM Progress: 21001 gene_ENSMUSG00000074885_4\n",
      "08/31/2018 03:02:25 PM Progress: 22001 gene_ENSMUSG00000077433_1\n",
      "08/31/2018 03:05:45 PM Progress: 23001 gene_ENSMUSG00000079871_4\n",
      "08/31/2018 03:07:30 PM Progress: 24001 gene_ENSMUSG00000081590_1\n",
      "08/31/2018 03:09:12 PM Progress: 25001 gene_ENSMUSG00000082726_1\n",
      "08/31/2018 03:11:02 PM Progress: 26001 gene_ENSMUSG00000083884_1\n",
      "08/31/2018 03:13:25 PM Progress: 27001 gene_ENSMUSG00000085090_1\n",
      "08/31/2018 03:17:55 PM Progress: 28001 gene_ENSMUSG00000086322_7\n",
      "08/31/2018 03:22:15 PM Progress: 29001 gene_ENSMUSG00000087554_1\n",
      "08/31/2018 03:24:17 PM Progress: 30001 gene_ENSMUSG00000088856_1\n",
      "08/31/2018 03:26:59 PM Progress: 31001 gene_ENSMUSG00000090207_2\n",
      "08/31/2018 03:29:57 PM Progress: 32001 gene_ENSMUSG00000092309_2\n",
      "08/31/2018 03:32:27 PM Progress: 33001 gene_ENSMUSG00000093827_1\n",
      "08/31/2018 03:34:17 PM Progress: 34001 gene_ENSMUSG00000095042_1\n",
      "08/31/2018 03:36:15 PM Progress: 35001 gene_ENSMUSG00000096295_2\n",
      "08/31/2018 03:40:16 PM Progress: 36001 gene_ENSMUSG00000097464_1\n",
      "08/31/2018 03:44:11 PM Progress: 37001 gene_ENSMUSG00000098572_1\n",
      "08/31/2018 03:46:16 PM Progress: 38001 gene_ENSMUSG00000099674_6\n",
      "08/31/2018 03:48:34 PM Progress: 39001 gene_ENSMUSG00000100731_1\n",
      "08/31/2018 03:50:54 PM Progress: 40001 gene_ENSMUSG00000101782_1\n",
      "08/31/2018 03:53:37 PM Progress: 41001 gene_ENSMUSG00000102855_1\n",
      "08/31/2018 03:56:17 PM Progress: 42001 gene_ENSMUSG00000103944_1\n",
      "08/31/2018 03:58:48 PM Progress: 43001 gene_ENSMUSG00000104988_1\n",
      "08/31/2018 04:01:08 PM Progress: 44001 gene_ENSMUSG00000105996_1\n",
      "08/31/2018 04:03:28 PM Progress: 45001 gene_ENSMUSG00000107094_1\n",
      "08/31/2018 04:06:11 PM Progress: 46001 gene_ENSMUSG00000108204_1\n",
      "08/31/2018 04:08:31 PM Progress: 47001 gene_ENSMUSG00000109254_1\n",
      "08/31/2018 04:12:44 PM Progress: 48001 gene_ENSMUSG00000110272_1\n",
      "08/31/2018 04:15:34 PM Progress: 49001 gene_ENSMUSG00000111375_3\n",
      "08/31/2018 04:19:52 PM Progress: 50001 gene_ENSMUSG00000112456_1\n",
      "08/31/2018 04:22:15 PM Progress: 51001 gene_ENSMUSG00000113459_1\n",
      "08/31/2018 04:25:05 PM Progress: 52001 gene_ENSMUSG00000114463_1\n",
      "08/31/2018 04:27:53 PM Progress: 53001 gene_ENSMUSG00000115468_1\n"
     ]
    }
   ],
   "source": [
    "sql = \"SELECT * from genes\"\n",
    "engine = connect_sql(DATABASE_ATAC)\n",
    "df_genes = pd.read_sql(sql, engine)\n",
    "gene_tables = df_genes['gene_id'].apply(gene_id_to_table_name)\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "for i, gene_table in enumerate(gene_tables):\n",
    "    if i % 1000 == 0:\n",
    "        logging.info('Progress: {} {}'.format(i+1, gene_table))\n",
    "    try:\n",
    "        table = sa.Table(gene_table, metadata, autoload=True)\n",
    "        col = sa.Column('smoothed_normalized_counts', sa.Float, default=0)\n",
    "\n",
    "        add_column(engine, gene_table, col)\n",
    "    except:\n",
    "        print(gene_table)\n",
    "    \n",
    "    \n",
    "#     d = table.delete()\n",
    "#     d.execute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
