{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import sqlalchemy as sa\n",
    "from sqlalchemy import update\n",
    "from sqlalchemy.dialects.mysql import INTEGER\n",
    "from sqlalchemy.dialects.mysql import SMALLINT\n",
    "import time\n",
    "import os\n",
    "import logging\n",
    "import glob\n",
    "from collections import OrderedDict\n",
    "from itertools import product \n",
    "\n",
    "\n",
    "from __init__ import *\n",
    "from snmcseq_utils import get_mouse_chromosomes\n",
    "from snmcseq_utils import compute_global_mC \n",
    "from snmcseq_utils import create_logger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "log = create_logger()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# An engine connects to a mysql database\n",
    "engine = sa.create_engine('mysql://f7xie:3405040212@localhost/CEMBA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_id_to_table_name(gene_id):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table_name = 'gene_' + gene_id.replace('.', '_')\n",
    "    return table_name\n",
    "\n",
    "\n",
    "CELLS_TABLE_COLS = ['cell_id', \n",
    "                     'cell_name', \n",
    "                     'dataset', \n",
    "                     'cell_type',\n",
    "                     'global_mCH', \n",
    "                     'global_mCG',\n",
    "                     'global_mCA',\n",
    "                     'global_mCCC', \n",
    "                     'estimated_mCH', \n",
    "                     'estimated_mCG',\n",
    "                     'percent_genome_covered', \n",
    "                     'total_reads',\n",
    "                     'mapped_reads', \n",
    "                     'mapping_rate', \n",
    "                     'nonclonal_reads', \n",
    "                     'percent_nonclonal_rate',\n",
    "                     'filtered_reads',\n",
    "                     'filtered_rate',\n",
    "                     'lambda_mC']\n",
    "\n",
    "def define_cells_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('cells', metadata,\n",
    "                sa.Column('cell_id', sa.Integer, primary_key=True, autoincrement=True), # auto increment and not null are implicitly defined\n",
    "                sa.Column('cell_name', sa.String(255), nullable=False, unique=True), # unique but not primary key\n",
    "                sa.Column('dataset', sa.String(40), nullable=False), # not in mapping summary\n",
    "                sa.Column('cell_type', sa.String(20), nullable=True), # not in mapping summary\n",
    "                     \n",
    "                sa.Column('global_mCH', sa.Float, nullable=True),\n",
    "                sa.Column('global_mCG', sa.Float, nullable=True),\n",
    "                sa.Column('global_mCA', sa.Float, nullable=True), # not in mapping summary\n",
    "                sa.Column('global_mCCC', sa.Float, nullable=True), \n",
    "                sa.Column('estimated_mCH', sa.Float, nullable=True),\n",
    "                sa.Column('estimated_mCG', sa.Float, nullable=True),\n",
    "                sa.Column('percent_genome_covered', sa.Float, nullable=True),\n",
    "                     \n",
    "                sa.Column('total_reads', sa.Integer, nullable=True),\n",
    "                sa.Column('mapped_reads', sa.Integer, nullable=True),\n",
    "                sa.Column('mapping_rate', sa.Float, nullable=True),\n",
    "                sa.Column('nonclonal_reads', sa.Integer, nullable=True),\n",
    "                sa.Column('percent_nonclonal_rate', sa.Float, nullable=True),\n",
    "                sa.Column('filtered_reads', sa.Integer, nullable=True),\n",
    "                sa.Column('filtered_rate', sa.Float, nullable=True),\n",
    "                sa.Column('lambda_mC', sa.Float, nullable=True),\n",
    "                )\n",
    "    return table\n",
    "\n",
    "\n",
    "def define_gene_table(metadata, gene_id):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table_name = gene_id_to_table_name(gene_id)\n",
    "    table = sa.Table(table_name, metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('cell_id', sa.Integer, sa.ForeignKey('cells.cell_id'), primary_key=True, autoincrement=False), \n",
    "                        sa.Column('mCH', INTEGER(unsigned=True), nullable=True),\n",
    "                        sa.Column('CH', INTEGER(unsigned=True), nullable=True),\n",
    "                        sa.Column('mCG', INTEGER(unsigned=True), nullable=True),\n",
    "                        sa.Column('CG', INTEGER(unsigned=True), nullable=True),\n",
    "                        sa.Column('mCA', INTEGER(unsigned=True), nullable=True),\n",
    "                        sa.Column('CA', INTEGER(unsigned=True), nullable=True),\n",
    "    )\n",
    "    return table\n",
    "\n",
    "def define_ens_table(metadata, ens):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # prototypes\n",
    "#   sa.Column('tsne_x_mCH', sa.Float, nullable=True),\n",
    "#   sa.Column('tsne_y_mCH', sa.Float, nullable=True),\n",
    "#   sa.Column('cluster', SMALLINT(unsigned=True), nullable=True),\n",
    "#   sa.Column('annotation', sa.String(20), nullable=True),\n",
    "\n",
    "    # tsnes\n",
    "    contexts = CONTEXTS + ['m'.join(comb_contexts) for comb_contexts in COMBINED_CONTEXTS_LIST]\n",
    "    perps = PERPLEXITIES\n",
    "    tsne_types = ['m{}_ndim2_perp{}'.format(context, p) for (context, p) in product(contexts, perps)]\n",
    "    \n",
    "    # prepare columns\n",
    "    args_tsne = ([sa.Column('tsne_x_{}'.format(tsne_type), sa.Float, nullable=True) \n",
    "            for tsne_type in tsne_types] \n",
    "            + [sa.Column('tsne_y_{}'.format(tsne_type), sa.Float, nullable=True) \n",
    "            for tsne_type in tsne_types]\n",
    "           )\n",
    "    \n",
    "    # clusters\n",
    "    ks = K_NN\n",
    "    cluster_types = ['m{}_lv_npc50_k{}'.format(context, k) for (context, k) in product(contexts, ks)]\n",
    "    # prepare columns\n",
    "    args_cluster = ([sa.Column('cluster_{}'.format(cluster_type), SMALLINT(unsigned=True), nullable=True) \n",
    "            for cluster_type in cluster_types] \n",
    "            + [sa.Column('annotation_{}'.format(cluster_type), sa.String(20), nullable=True) \n",
    "            for cluster_type in cluster_types]\n",
    "           )\n",
    "    \n",
    "    # combine cols\n",
    "    args = args_tsne + args_cluster\n",
    "    # create the table\n",
    "    table = sa.Table(ens, metadata,\n",
    "                        # NOT auto increment not null primary key\n",
    "                        sa.Column('cell_id', sa.Integer, sa.ForeignKey('cells.cell_id'), primary_key=True, autoincrement=False), \n",
    "                        *args\n",
    "    )\n",
    "    return table\n",
    "\n",
    "def define_enss_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('ensembles', metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('ensemble_id', sa.Integer, primary_key=True, autoincrement=False), \n",
    "                        sa.Column('ensemble_name', sa.String(255), nullable=False, unique=True),\n",
    "                        sa.Column('public_access', sa.Boolean, nullable=False, default=False),\n",
    "                        sa.Column('datasets', sa.String(4000), nullable=False),\n",
    "    )\n",
    "    return table\n",
    "\n",
    "def define_genes_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('genes', metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('gene_id', sa.String(50), primary_key=True, autoincrement=False), \n",
    "                        sa.Column('gene_name', sa.String(100), nullable=False), \n",
    "                        sa.Column('chr', sa.String(5), nullable=True), \n",
    "                        sa.Column('start', sa.Integer, nullable=True), \n",
    "                        sa.Column('end', sa.Integer, nullable=True), \n",
    "                        sa.Column('strand', sa.CHAR(1), nullable=True), \n",
    "                        sa.Column('gene_type', sa.String(100), nullable=True), \n",
    "    )\n",
    "    return table\n",
    "\n",
    "def insert_into_worker(engine, table_name, dict_list, ignore=False):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    metadata = sa.MetaData(engine)\n",
    "    table = sa.Table(table_name, metadata, autoload=True, autoload_with=engine)\n",
    "    if ignore: \n",
    "        inserter = table.insert().prefix_with('IGNORE')\n",
    "    else:\n",
    "        inserter = table.insert()\n",
    "        \n",
    "    return engine.execute(inserter, dict_list)\n",
    "\n",
    "def insert_into(engine, table_name, df_sql, ignore=False, verbose=True):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # change null to none\n",
    "    df_sql = df_sql.where(pd.notnull(df_sql), None)\n",
    "    if verbose:\n",
    "        print(df_sql.head())\n",
    "    # dict_list\n",
    "    dict_list = list(df_sql.T.to_dict().values())\n",
    "    \n",
    "    return insert_into_worker(engine, table_name, dict_list, ignore=ignore)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>tsne_x_mCA_ndim2_perp100</th>\n",
       "      <th>tsne_y_mCA_ndim2_perp100</th>\n",
       "      <th>tsne_x_mCAmCG_ndim2_perp100</th>\n",
       "      <th>tsne_y_mCAmCG_ndim2_perp100</th>\n",
       "      <th>tsne_x_mCG_ndim2_perp100</th>\n",
       "      <th>tsne_y_mCG_ndim2_perp100</th>\n",
       "      <th>tsne_x_mCH_ndim2_perp100</th>\n",
       "      <th>tsne_y_mCH_ndim2_perp100</th>\n",
       "      <th>tsne_x_mCHmCG_ndim2_perp100</th>\n",
       "      <th>...</th>\n",
       "      <th>cluster_mCA_lv_npc50_k5</th>\n",
       "      <th>annotation_mCA_lv_npc50_k5</th>\n",
       "      <th>cluster_mCAmCG_lv_npc50_k5</th>\n",
       "      <th>annotation_mCAmCG_lv_npc50_k5</th>\n",
       "      <th>cluster_mCG_lv_npc50_k5</th>\n",
       "      <th>annotation_mCG_lv_npc50_k5</th>\n",
       "      <th>cluster_mCH_lv_npc50_k5</th>\n",
       "      <th>annotation_mCH_lv_npc50_k5</th>\n",
       "      <th>cluster_mCHmCG_lv_npc50_k5</th>\n",
       "      <th>annotation_mCHmCG_lv_npc50_k5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2768</th>\n",
       "      <td>2769</td>\n",
       "      <td>-5.762300</td>\n",
       "      <td>-15.008065</td>\n",
       "      <td>-37.587276</td>\n",
       "      <td>12.646081</td>\n",
       "      <td>10.866119</td>\n",
       "      <td>11.335617</td>\n",
       "      <td>-7.823882</td>\n",
       "      <td>22.779170</td>\n",
       "      <td>8.615857</td>\n",
       "      <td>...</td>\n",
       "      <td>59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>54</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2769</th>\n",
       "      <td>2770</td>\n",
       "      <td>-0.178485</td>\n",
       "      <td>-9.851624</td>\n",
       "      <td>-29.237764</td>\n",
       "      <td>22.123650</td>\n",
       "      <td>13.289175</td>\n",
       "      <td>6.599739</td>\n",
       "      <td>-1.587324</td>\n",
       "      <td>16.138401</td>\n",
       "      <td>15.223225</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>mL6-1</td>\n",
       "      <td>38</td>\n",
       "      <td>mL6-1</td>\n",
       "      <td>22</td>\n",
       "      <td>mL6-1</td>\n",
       "      <td>22</td>\n",
       "      <td>mL6-1</td>\n",
       "      <td>21</td>\n",
       "      <td>mL6-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2770</th>\n",
       "      <td>2771</td>\n",
       "      <td>-4.812684</td>\n",
       "      <td>4.926068</td>\n",
       "      <td>-15.597104</td>\n",
       "      <td>-37.158940</td>\n",
       "      <td>-13.232499</td>\n",
       "      <td>-2.398929</td>\n",
       "      <td>4.830728</td>\n",
       "      <td>1.114001</td>\n",
       "      <td>-20.163807</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>mL5-1</td>\n",
       "      <td>52</td>\n",
       "      <td>mL5-1</td>\n",
       "      <td>26</td>\n",
       "      <td>mL5-1</td>\n",
       "      <td>33</td>\n",
       "      <td>mL5-1</td>\n",
       "      <td>30</td>\n",
       "      <td>mL5-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2771</th>\n",
       "      <td>2772</td>\n",
       "      <td>4.022660</td>\n",
       "      <td>14.922937</td>\n",
       "      <td>13.459216</td>\n",
       "      <td>-13.775705</td>\n",
       "      <td>-19.178972</td>\n",
       "      <td>-13.628133</td>\n",
       "      <td>-0.024424</td>\n",
       "      <td>-15.978237</td>\n",
       "      <td>-6.252432</td>\n",
       "      <td>...</td>\n",
       "      <td>35</td>\n",
       "      <td>mL2/3</td>\n",
       "      <td>42</td>\n",
       "      <td>mL2/3</td>\n",
       "      <td>4</td>\n",
       "      <td>mL2/3</td>\n",
       "      <td>1</td>\n",
       "      <td>mL2/3</td>\n",
       "      <td>5</td>\n",
       "      <td>mL2/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2772</th>\n",
       "      <td>2773</td>\n",
       "      <td>5.178892</td>\n",
       "      <td>6.409532</td>\n",
       "      <td>5.937645</td>\n",
       "      <td>-32.622080</td>\n",
       "      <td>-9.346658</td>\n",
       "      <td>-8.356455</td>\n",
       "      <td>3.301490</td>\n",
       "      <td>-6.993064</td>\n",
       "      <td>-11.582888</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>mL4</td>\n",
       "      <td>8</td>\n",
       "      <td>mL4</td>\n",
       "      <td>1</td>\n",
       "      <td>mL4</td>\n",
       "      <td>10</td>\n",
       "      <td>mL4</td>\n",
       "      <td>15</td>\n",
       "      <td>mL4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cell_id  tsne_x_mCA_ndim2_perp100  tsne_y_mCA_ndim2_perp100  \\\n",
       "2768     2769                 -5.762300                -15.008065   \n",
       "2769     2770                 -0.178485                 -9.851624   \n",
       "2770     2771                 -4.812684                  4.926068   \n",
       "2771     2772                  4.022660                 14.922937   \n",
       "2772     2773                  5.178892                  6.409532   \n",
       "\n",
       "      tsne_x_mCAmCG_ndim2_perp100  tsne_y_mCAmCG_ndim2_perp100  \\\n",
       "2768                   -37.587276                    12.646081   \n",
       "2769                   -29.237764                    22.123650   \n",
       "2770                   -15.597104                   -37.158940   \n",
       "2771                    13.459216                   -13.775705   \n",
       "2772                     5.937645                   -32.622080   \n",
       "\n",
       "      tsne_x_mCG_ndim2_perp100  tsne_y_mCG_ndim2_perp100  \\\n",
       "2768                 10.866119                 11.335617   \n",
       "2769                 13.289175                  6.599739   \n",
       "2770                -13.232499                 -2.398929   \n",
       "2771                -19.178972                -13.628133   \n",
       "2772                 -9.346658                 -8.356455   \n",
       "\n",
       "      tsne_x_mCH_ndim2_perp100  tsne_y_mCH_ndim2_perp100  \\\n",
       "2768                 -7.823882                 22.779170   \n",
       "2769                 -1.587324                 16.138401   \n",
       "2770                  4.830728                  1.114001   \n",
       "2771                 -0.024424                -15.978237   \n",
       "2772                  3.301490                 -6.993064   \n",
       "\n",
       "      tsne_x_mCHmCG_ndim2_perp100              ...                \\\n",
       "2768                     8.615857              ...                 \n",
       "2769                    15.223225              ...                 \n",
       "2770                   -20.163807              ...                 \n",
       "2771                    -6.252432              ...                 \n",
       "2772                   -11.582888              ...                 \n",
       "\n",
       "      cluster_mCA_lv_npc50_k5  annotation_mCA_lv_npc50_k5  \\\n",
       "2768                       59                         NaN   \n",
       "2769                       38                       mL6-1   \n",
       "2770                        5                       mL5-1   \n",
       "2771                       35                       mL2/3   \n",
       "2772                       24                         mL4   \n",
       "\n",
       "      cluster_mCAmCG_lv_npc50_k5  annotation_mCAmCG_lv_npc50_k5  \\\n",
       "2768                          54                            NaN   \n",
       "2769                          38                          mL6-1   \n",
       "2770                          52                          mL5-1   \n",
       "2771                          42                          mL2/3   \n",
       "2772                           8                            mL4   \n",
       "\n",
       "      cluster_mCG_lv_npc50_k5  annotation_mCG_lv_npc50_k5  \\\n",
       "2768                       20                         NaN   \n",
       "2769                       22                       mL6-1   \n",
       "2770                       26                       mL5-1   \n",
       "2771                        4                       mL2/3   \n",
       "2772                        1                         mL4   \n",
       "\n",
       "      cluster_mCH_lv_npc50_k5  annotation_mCH_lv_npc50_k5  \\\n",
       "2768                       52                         NaN   \n",
       "2769                       22                       mL6-1   \n",
       "2770                       33                       mL5-1   \n",
       "2771                        1                       mL2/3   \n",
       "2772                       10                         mL4   \n",
       "\n",
       "      cluster_mCHmCG_lv_npc50_k5  annotation_mCHmCG_lv_npc50_k5  \n",
       "2768                          55                            NaN  \n",
       "2769                          21                          mL6-1  \n",
       "2770                          30                          mL5-1  \n",
       "2771                           5                          mL2/3  \n",
       "2772                          15                            mL4  \n",
       "\n",
       "[5 rows x 121 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ens\n",
    "ens = 'Ens4'\n",
    "ens_path = os.path.join(PATH_ENSEMBLES, ens)\n",
    "cluster_files = sorted(glob.glob(os.path.join(ens_path, 'cluster/cluster_*.tsv'))) \n",
    "tsne_files = sorted(glob.glob(os.path.join(ens_path, 'tsne/tsne_*.tsv')))\n",
    "\n",
    "# get cells\n",
    "sql = \"\"\"SELECT * FROM cells\"\"\"\n",
    "df_cell = pd.read_sql(sql, engine)[['cell_id', 'cell_name']]\n",
    "\n",
    "dfs_cluster = []\n",
    "for cluster_file in cluster_files:\n",
    "    # filename to cluster_type\n",
    "    paras = os.path.basename(cluster_file).split('_')\n",
    "    method, para1, para2, context = paras[1], paras[2], paras[3], paras[5]\n",
    "    cluster_type = '{}_{}_{}_{}'.format(context, method, para1, para2) \n",
    "    # load data\n",
    "    annot_file = cluster_file + '.annot'\n",
    "    df_cluster = pd.read_table(cluster_file, index_col='sample')\n",
    "    df_annot = pd.read_table(annot_file, index_col='cluster_ID')\n",
    "    # merge \n",
    "    df_cluster = pd.merge(df_cluster, df_annot, left_on='cluster_ID', right_index=True)\n",
    "    # organize format\n",
    "    df_cluster.cluster_ID = [int(cluster_ID[len('cluster_'):])for cluster_ID in df_cluster.cluster_ID]\n",
    "    df_cluster.columns = ['cluster_{}'.format(cluster_type), \n",
    "                         'annotation_{}'.format(cluster_type)]\n",
    "    # append \n",
    "    dfs_cluster.append(df_cluster)\n",
    "# merge \n",
    "df_clusters = pd.concat(dfs_cluster, axis=1)\n",
    "    \n",
    "dfs_tsne = []\n",
    "for tsne_file in tsne_files:\n",
    "    # filename to cluster_type\n",
    "    paras = os.path.basename(tsne_file).split('_')\n",
    "    para1, para2, context = paras[1], paras[2], paras[5]\n",
    "    tsne_type = '{}_{}_{}'.format(context, para1, para2) \n",
    "    # load data\n",
    "    df_tsne = pd.read_table(tsne_file, index_col='sample')\n",
    "    # organize format\n",
    "    df_tsne.columns = [col+'_'+tsne_type for col in df_tsne.columns]\n",
    "    # append\n",
    "    dfs_tsne.append(df_tsne)\n",
    "df_tsnes = pd.concat(dfs_tsne, axis=1)\n",
    "df = pd.concat([df_tsnes, df_clusters], axis=1)\n",
    "\n",
    "df = pd.merge(df_cell, df, left_on='cell_name', right_index=True)\n",
    "df = df.drop('cell_name', axis=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      cell_id  tsne_x_mCA_ndim2_perp100  tsne_y_mCA_ndim2_perp100  \\\n",
      "2768     2769                 -5.762300                -15.008065   \n",
      "2769     2770                 -0.178485                 -9.851624   \n",
      "2770     2771                 -4.812684                  4.926068   \n",
      "2771     2772                  4.022660                 14.922937   \n",
      "2772     2773                  5.178892                  6.409532   \n",
      "\n",
      "      tsne_x_mCAmCG_ndim2_perp100  tsne_y_mCAmCG_ndim2_perp100  \\\n",
      "2768                   -37.587276                    12.646081   \n",
      "2769                   -29.237764                    22.123650   \n",
      "2770                   -15.597104                   -37.158940   \n",
      "2771                    13.459216                   -13.775705   \n",
      "2772                     5.937645                   -32.622080   \n",
      "\n",
      "      tsne_x_mCG_ndim2_perp100  tsne_y_mCG_ndim2_perp100  \\\n",
      "2768                 10.866119                 11.335617   \n",
      "2769                 13.289175                  6.599739   \n",
      "2770                -13.232499                 -2.398929   \n",
      "2771                -19.178972                -13.628133   \n",
      "2772                 -9.346658                 -8.356455   \n",
      "\n",
      "      tsne_x_mCH_ndim2_perp100  tsne_y_mCH_ndim2_perp100  \\\n",
      "2768                 -7.823882                 22.779170   \n",
      "2769                 -1.587324                 16.138401   \n",
      "2770                  4.830728                  1.114001   \n",
      "2771                 -0.024424                -15.978237   \n",
      "2772                  3.301490                 -6.993064   \n",
      "\n",
      "      tsne_x_mCHmCG_ndim2_perp100              ...                \\\n",
      "2768                     8.615857              ...                 \n",
      "2769                    15.223225              ...                 \n",
      "2770                   -20.163807              ...                 \n",
      "2771                    -6.252432              ...                 \n",
      "2772                   -11.582888              ...                 \n",
      "\n",
      "      cluster_mCA_lv_npc50_k5  annotation_mCA_lv_npc50_k5  \\\n",
      "2768                       59                        None   \n",
      "2769                       38                       mL6-1   \n",
      "2770                        5                       mL5-1   \n",
      "2771                       35                       mL2/3   \n",
      "2772                       24                         mL4   \n",
      "\n",
      "      cluster_mCAmCG_lv_npc50_k5  annotation_mCAmCG_lv_npc50_k5  \\\n",
      "2768                          54                           None   \n",
      "2769                          38                          mL6-1   \n",
      "2770                          52                          mL5-1   \n",
      "2771                          42                          mL2/3   \n",
      "2772                           8                            mL4   \n",
      "\n",
      "      cluster_mCG_lv_npc50_k5  annotation_mCG_lv_npc50_k5  \\\n",
      "2768                       20                        None   \n",
      "2769                       22                       mL6-1   \n",
      "2770                       26                       mL5-1   \n",
      "2771                        4                       mL2/3   \n",
      "2772                        1                         mL4   \n",
      "\n",
      "      cluster_mCH_lv_npc50_k5  annotation_mCH_lv_npc50_k5  \\\n",
      "2768                       52                        None   \n",
      "2769                       22                       mL6-1   \n",
      "2770                       33                       mL5-1   \n",
      "2771                        1                       mL2/3   \n",
      "2772                       10                         mL4   \n",
      "\n",
      "      cluster_mCHmCG_lv_npc50_k5  annotation_mCHmCG_lv_npc50_k5  \n",
      "2768                          55                           None  \n",
      "2769                          21                          mL6-1  \n",
      "2770                          30                          mL5-1  \n",
      "2771                           5                          mL2/3  \n",
      "2772                          15                            mL4  \n",
      "\n",
      "[5 rows x 121 columns]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<sqlalchemy.engine.result.ResultProxy at 0x7f66f3477e80>"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_name = ens \n",
    "insert_into(engine, table_name, df, ignore=False, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "# demo: create tables\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "# load specfic tables needed to referenced as foreign key\n",
    "sa.Table('cells', metadata, autoload=True)\n",
    "\n",
    "# define cells table\n",
    "# cells_table = define_cells_table(metadata)\n",
    "\n",
    "# define ensembles table\n",
    "# enss_table = define_enss_table(metadata)\n",
    "\n",
    "# define genes table\n",
    "# genes_table = define_genes_table(metadata)\n",
    "\n",
    "\n",
    "\n",
    "# define ensemble table\n",
    "# ens = 'Ens4'\n",
    "# ens_table = define_ens_table(metadata, ens)\n",
    "\n",
    "# # define gene table\n",
    "# gene_id = 'ENSG1111.1'\n",
    "# gene_table = define_gene_table(metadata, gene_id)\n",
    "\n",
    "metadata.create_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                gene_id      gene_name   chr    start      end strand  \\\n",
      "0  ENSMUSG00000102693.1  4933401J01Rik  chr1  3073253  3074322      +   \n",
      "1  ENSMUSG00000064842.1        Gm26206  chr1  3102016  3102125      +   \n",
      "2  ENSMUSG00000051951.5           Xkr4  chr1  3205901  3671498      -   \n",
      "3  ENSMUSG00000102851.1        Gm18956  chr1  3252757  3253236      +   \n",
      "4  ENSMUSG00000103377.1        Gm37180  chr1  3365731  3368549      -   \n",
      "\n",
      "              gene_type  \n",
      "0                   TEC  \n",
      "1                 snRNA  \n",
      "2        protein_coding  \n",
      "3  processed_pseudogene  \n",
      "4                   TEC  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<sqlalchemy.engine.result.ResultProxy at 0x7f6f76fa2fd0>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update to genes table\n",
    "table_name = 'genes'\n",
    "\n",
    "path_genebody = GENEBODY\n",
    "df_genes = pd.read_table(path_genebody, index_col='gene_id')\n",
    "\n",
    "df_sql = df_genes.reset_index()\n",
    "insert_into(engine, 'genes', df_sql, ignore=False, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_ms_cols(column_names):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    dict_rename = {'Sample': 'cell_name', \n",
    "                   'Total reads': 'total_reads', \n",
    "                   'Mapped reads': 'mapped_reads', \n",
    "                   'Mapping rate': 'mapping_rate',\n",
    "                   'Nonclonal reads': 'nonclonal_reads',\n",
    "                   '% Nonclonal rate': 'percent_nonclonal_rate', \n",
    "                   'Filtered reads': 'filtered_reads', \n",
    "                   'Filtered rate': 'filtered_rate', \n",
    "                   'Lambda mC/C': 'lambda_mC', \n",
    "                   'mCCC/CCC': 'global_mCCC', \n",
    "                   'mCG/CG': 'global_mCG', \n",
    "                   'mCH/CH': 'global_mCH',\n",
    "                   'Estimated mCG/CG': 'estimated_mCG', \n",
    "                   'Estimated mCH/CH': 'estimated_mCH', \n",
    "                   '% Genome covered': 'percent_genome_covered'}\n",
    "    return [dict_rename[col] for col in column_names] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create gene tables \n",
    "ti = time.time()\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "# load specfic tables needed to referenced as foreign key\n",
    "sa.Table('cells', metadata, autoload=True)\n",
    "\n",
    "\n",
    "sql = 'SELECT * FROM genes'\n",
    "df_genes = pd.read_sql(sql, engine, index_col='gene_id')\n",
    "# print(df_genes.shape)\n",
    "# df_genes.head()\n",
    "# df_genes.sort_index().head()\n",
    "\n",
    "# create gene tables \n",
    "for i, gene_id in enumerate(df_genes.index):\n",
    "    logging.info(\"Creating gene tables... ({}/{})\".format(i+1, df_genes.index.shape[0]))\n",
    "    gene_table = define_gene_table(metadata, gene_id)\n",
    "\n",
    "metadata.create_all(engine)\n",
    "\n",
    "tf = time.time()\n",
    "print(tf - ti)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlalchemy.engine.result.ResultProxy at 0x7f66f8333080>"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update to enss table\n",
    "dict_list = [{'ensemble_id': 1, 'ensemble_name': 'CEMBA_3C_171206', 'public_access': False, 'datasets': 'CEMBA_3C_171206'},\n",
    "             {'ensemble_id': 2, 'ensemble_name': 'CEMBA_3C_171207', 'public_access': False, 'datasets': 'CEMBA_3C_171207'},\n",
    "             {'ensemble_id': 3, 'ensemble_name': 'CEMBA_4B_171212', 'public_access': False, 'datasets': 'CEMBA_4B_171212'},\n",
    "             {'ensemble_id': 4, 'ensemble_name': 'CEMBA_4B_171213', 'public_access': False, 'datasets': 'CEMBA_4B_171213'},\n",
    "            ]\n",
    "table_name = 'ensembles'\n",
    "insert_into_worker(engine, table_name, dict_list, ignore=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chunks(l, n):\n",
    "    \"\"\"Yield successive n-sized chunks from l.\"\"\"\n",
    "    for i in range(0, len(l), n):\n",
    "        yield l[i:i + n]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "02/09/2018 08:02:24 PM Compute global methylation levels...(CEMBA_3C_171207, ['CH', 'CG', 'CA'])\n",
      "02/09/2018 08:02:24 PM Progress: 1/1270\n",
      "02/09/2018 08:03:03 PM Progress: 101/1270\n",
      "02/09/2018 08:03:42 PM Progress: 201/1270\n",
      "02/09/2018 08:04:21 PM Progress: 301/1270\n",
      "02/09/2018 08:05:00 PM Progress: 401/1270\n",
      "02/09/2018 08:05:39 PM Progress: 501/1270\n",
      "02/09/2018 08:06:17 PM Progress: 601/1270\n",
      "02/09/2018 08:06:56 PM Progress: 701/1270\n",
      "02/09/2018 08:07:35 PM Progress: 801/1270\n",
      "02/09/2018 08:08:13 PM Progress: 901/1270\n",
      "02/09/2018 08:08:50 PM Progress: 1001/1270\n",
      "02/09/2018 08:09:26 PM Progress: 1101/1270\n",
      "02/09/2018 08:10:01 PM Progress: 1201/1270\n",
      "02/09/2018 08:10:25 PM Done computing global methylation levels, total time spent: 480.4605016708374 seconds\n"
     ]
    }
   ],
   "source": [
    "# stmt = users.update().where(users.c.id==5).values(name='user #5')\n",
    "\n",
    "# dataset = 'CEMBA_3C_171207'\n",
    "# df_res = compute_global_mC(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # update to cells table \n",
    "# metadata = sa.MetaData(engine)\n",
    "# table = sa.Table('cells', metadata, autoload=True)\n",
    "\n",
    "# sql = \"\"\"SELECT * FROM cells WHERE dataset='{}'\"\"\".format(dataset)\n",
    "# df_cells = pd.read_sql(sql, engine)\n",
    "# df_cells = pd.merge(df_cells[['cell_id', 'cell_name']], df_res, left_on='cell_name', right_index=True)\n",
    "# for i, row in df_cells.iterrows():\n",
    "#     updater = update(table).where(table.c.cell_id==row.cell_id).values(global_mCA=row.global_mCA)\n",
    "#     engine.execute(updater)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update mapping summary file\n",
    "# dataset_path = os.path.join(PATH_DATASETS, dataset)\n",
    "# meta_path = os.path.join(dataset_path, 'mapping_summary_{}.tsv'.format(dataset)) \n",
    "\n",
    "# df_meta = pd.read_table(meta_path)\n",
    "# df_meta['mCA/CA'] = df_res.loc[df_meta.Sample, 'global_mCA'].values\n",
    "# df_meta.to_csv(meta_path, sep='\\t', na_rep='NA', header=True, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "02/09/2018 08:41:07 PM Compute global methylation levels...(CEMBA_3C_171206, ['CH', 'CG', 'CA'])\n",
      "02/09/2018 08:41:07 PM Progress: 1/1202\n",
      "02/09/2018 08:41:40 PM Progress: 101/1202\n",
      "02/09/2018 08:42:12 PM Progress: 201/1202\n",
      "02/09/2018 08:42:45 PM Progress: 301/1202\n",
      "02/09/2018 08:43:18 PM Progress: 401/1202\n",
      "02/09/2018 08:43:50 PM Progress: 501/1202\n",
      "02/09/2018 08:44:23 PM Progress: 601/1202\n",
      "02/09/2018 08:44:56 PM Progress: 701/1202\n",
      "02/09/2018 08:45:29 PM Progress: 801/1202\n",
      "02/09/2018 08:46:02 PM Progress: 901/1202\n",
      "02/09/2018 08:46:35 PM Progress: 1001/1202\n",
      "02/09/2018 08:47:08 PM Progress: 1101/1202\n",
      "02/09/2018 08:47:41 PM Progress: 1201/1202\n",
      "02/09/2018 08:47:42 PM Done computing global methylation levels, total time spent: 394.8638460636139 seconds\n"
     ]
    }
   ],
   "source": [
    "# datasets = ['CEMBA_3C_171206']\n",
    "# for dataset in datasets:\n",
    "#     df_res = compute_global_mC(dataset)\n",
    "\n",
    "#     # update to cells table \n",
    "#     metadata = sa.MetaData(engine)\n",
    "#     table = sa.Table('cells', metadata, autoload=True)\n",
    "\n",
    "#     sql = \"\"\"SELECT * FROM cells WHERE dataset='{}'\"\"\".format(dataset)\n",
    "#     df_cells = pd.read_sql(sql, engine)\n",
    "#     df_cells = pd.merge(df_cells[['cell_id', 'cell_name']], df_res, left_on='cell_name', right_index=True)\n",
    "#     for i, row in df_cells.iterrows():\n",
    "#         updater = update(table).where(table.c.cell_id==row.cell_id).values(global_mCA=row.global_mCA)\n",
    "#         engine.execute(updater)\n",
    "\n",
    "#     # update mapping summary file\n",
    "#     dataset_path = os.path.join(PATH_DATASETS, dataset)\n",
    "#     meta_path = os.path.join(dataset_path, 'mapping_summary_{}.tsv'.format(dataset)) \n",
    "\n",
    "#     df_meta = pd.read_table(meta_path)\n",
    "#     df_meta['mCA/CA'] = df_res.loc[df_meta.Sample, 'global_mCA'].values\n",
    "#     df_meta.to_csv(meta_path, sep='\\t', na_rep='NA', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
