{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlalchemy as sa\n",
    "from sqlalchemy import update\n",
    "from sqlalchemy.dialects.mysql import INTEGER\n",
    "from sqlalchemy.dialects.mysql import SMALLINT\n",
    "from collections import OrderedDict\n",
    "from itertools import product \n",
    "\n",
    "\n",
    "from __init__ import *\n",
    "from snmcseq_utils import cd \n",
    "from snmcseq_utils import get_mouse_chromosomes\n",
    "from snmcseq_utils import compute_global_mC \n",
    "from snmcseq_utils import create_logger\n",
    "from CEMBA_run_tsne import run_tsne_CEMBA\n",
    "from CEMBA_update_mysql import insert_into\n",
    "from CEMBA_update_mysql import insert_into_worker\n",
    "from CEMBA_update_mysql import gene_id_to_table_name \n",
    "from CEMBA_update_mysql import connect_sql "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "log = create_logger()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# An engine connects to a mysql database\n",
    "database = 'human_snmcseq'\n",
    "engine = connect_sql(database) # sa.create_engine('mysql://f7xie:3405040212@localhost/{}'.format(database))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating tables\n",
    "# \"genes\" table (gene annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_genes_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('genes', metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('gene_id', sa.String(50), primary_key=True, autoincrement=False), \n",
    "                        sa.Column('gene_name', sa.String(100), nullable=False), \n",
    "                        sa.Column('chr', sa.String(5), nullable=True), \n",
    "                        sa.Column('start', sa.Integer, nullable=True), \n",
    "                        sa.Column('end', sa.Integer, nullable=True), \n",
    "                        sa.Column('strand', sa.CHAR(1), nullable=True), \n",
    "                        sa.Column('gene_type', sa.String(100), nullable=True), \n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create genes table\n",
    "metadata = sa.MetaData(engine)\n",
    "genes_table = define_genes_table(metadata)\n",
    "metadata.create_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/19/2018 11:58:05 AM                 gene_id      gene_name   chr    start      end strand  \\\n",
      "0  ENSMUSG00000102693.1  4933401J01Rik  chr1  3073253  3074322      +   \n",
      "1  ENSMUSG00000064842.1        Gm26206  chr1  3102016  3102125      +   \n",
      "2  ENSMUSG00000051951.5           Xkr4  chr1  3205901  3671498      -   \n",
      "3  ENSMUSG00000102851.1        Gm18956  chr1  3252757  3253236      +   \n",
      "4  ENSMUSG00000103377.1        Gm37180  chr1  3365731  3368549      -   \n",
      "\n",
      "              gene_type  \n",
      "0                   TEC  \n",
      "1                 snRNA  \n",
      "2        protein_coding  \n",
      "3  processed_pseudogene  \n",
      "4                   TEC  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<sqlalchemy.engine.result.ResultProxy at 0x7fccaca4ad68>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update to genes table\n",
    "table_name = 'genes'\n",
    "\n",
    "path_genebody = PATH_GENEBODY_ANNOTATION\n",
    "df_genes = pd.read_table(path_genebody, index_col='gene_id')\n",
    "\n",
    "df_sql = df_genes.reset_index()\n",
    "insert_into(engine, 'genes', df_sql, ignore=False, verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# \"ensembles\" table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_enss_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('ensembles', metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('ensemble_id', sa.Integer, primary_key=True, autoincrement=False), \n",
    "                        sa.Column('ensemble_name', sa.String(255), nullable=False, unique=True),\n",
    "                        sa.Column('public_access', sa.Boolean, nullable=False, default=False),\n",
    "                        sa.Column('datasets', sa.String(4000), nullable=False),\n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create enss table\n",
    "metadata = sa.MetaData(engine)\n",
    "enss_table = define_enss_table(metadata)\n",
    "metadata.create_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update to enss table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# \"cells\" table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_cells_table(metadata):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table = sa.Table('cells', metadata,\n",
    "                sa.Column('cell_id', sa.Integer, primary_key=True, autoincrement=True), # auto increment and not null are implicitly defined\n",
    "                sa.Column('cell_name', sa.String(255), nullable=False, unique=True), # unique but not primary key\n",
    "                sa.Column('dataset', sa.String(40), nullable=False), # not in mapping summary\n",
    "#                 sa.Column('cell_type', sa.String(20), nullable=True), # not in mapping summary\n",
    "                     \n",
    "#                 sa.Column('global_mCH', sa.Float, nullable=True),\n",
    "#                 sa.Column('global_mCG', sa.Float, nullable=True),\n",
    "#                 sa.Column('global_mCA', sa.Float, nullable=True), # not in mapping summary\n",
    "#                 sa.Column('global_mCCC', sa.Float, nullable=True), \n",
    "#                 sa.Column('estimated_mCH', sa.Float, nullable=True),\n",
    "#                 sa.Column('estimated_mCG', sa.Float, nullable=True),\n",
    "#                 sa.Column('percent_genome_covered', sa.Float, nullable=True),\n",
    "                     \n",
    "#                 sa.Column('total_reads', sa.Integer, nullable=True),\n",
    "#                 sa.Column('mapped_reads', sa.Integer, nullable=True),\n",
    "#                 sa.Column('mapping_rate', sa.Float, nullable=True),\n",
    "#                 sa.Column('nonclonal_reads', sa.Integer, nullable=True),\n",
    "#                 sa.Column('percent_nonclonal_rate', sa.Float, nullable=True),\n",
    "#                 sa.Column('filtered_reads', sa.Integer, nullable=True),\n",
    "#                 sa.Column('filtered_rate', sa.Float, nullable=True),\n",
    "#                 sa.Column('lambda_mC', sa.Float, nullable=True),\n",
    "                )\n",
    "    return table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create cells table\n",
    "metadata = sa.MetaData(engine)\n",
    "cells_table = define_cells_table(metadata)\n",
    "metadata.create_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update to cells table (dataset level)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Individual gene tables (\"gene_ensemble_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_gene_table(metadata, gene_id):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table_name = gene_id_to_table_name(gene_id)\n",
    "    table = sa.Table(table_name, metadata,\n",
    "                          # NOT auto increment not null primary key\n",
    "                        sa.Column('cell_id', sa.Integer, sa.ForeignKey('cells.cell_id'), primary_key=True, autoincrement=False), \n",
    "                        sa.Column('normalized_counts', sa.Float, nullable=True),\n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/19/2018 12:18:24 PM Creating gene tables... (53379 in total)\n",
      "03/19/2018 01:34:49 PM 4603.019817352295\n"
     ]
    }
   ],
   "source": [
    "# create individual gene tables (takes time)\n",
    "ti = time.time()\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "# load specfic tables needed to be referenced as foreign key\n",
    "sa.Table('cells', metadata, autoload=True)\n",
    "# load all genes \n",
    "sql = 'SELECT * FROM genes'\n",
    "df_genes = pd.read_sql(sql, engine, index_col='gene_id')\n",
    "# define gene tables \n",
    "for i, gene_id in enumerate(df_genes.index):\n",
    "    gene_table = define_gene_table(metadata, gene_id)\n",
    "# create them\n",
    "logging.info(\"Creating gene tables... ({} in total)\".format(df_genes.index.shape[0]))\n",
    "metadata.create_all(engine)\n",
    "logging.info(time.time() - ti)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update to gene tables (dataset level)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Individual ensemble tables (\"Ens1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "def define_ens_table(metadata, ens):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # prototypes\n",
    "#   sa.Column('tsne_x_mCH', sa.Float, nullable=True),\n",
    "#   sa.Column('tsne_y_mCH', sa.Float, nullable=True),\n",
    "#   sa.Column('cluster', SMALLINT(unsigned=True), nullable=True),\n",
    "#   sa.Column('annotation', sa.String(20), nullable=True),\n",
    "\n",
    "    # tsnes\n",
    "    contexts = ['ATAC'] \n",
    "    perps = PERPLEXITIES\n",
    "#     tsne_types = ['{}_ndim2_perp{}'.format(context, p) for (context, p) in product(contexts, perps)]\n",
    "    tsne_types = ['ATAC']\n",
    "    \n",
    "    # prepare columns\n",
    "    args_tsne = ([sa.Column('tsne_x_{}'.format(tsne_type), sa.Float, nullable=True) \n",
    "            for tsne_type in tsne_types] \n",
    "            + [sa.Column('tsne_y_{}'.format(tsne_type), sa.Float, nullable=True) \n",
    "            for tsne_type in tsne_types]\n",
    "           )\n",
    "    \n",
    "    # clusters\n",
    "    ks = K_NN\n",
    "#     cluster_types = ['{}_lv_npc50_k{}'.format(context, k) for (context, k) in product(contexts, ks)]\n",
    "    cluster_types = ['ATAC']\n",
    "    # prepare columns\n",
    "    args_cluster = ([sa.Column('cluster_{}'.format(cluster_type), SMALLINT(unsigned=True), nullable=True) \n",
    "            for cluster_type in cluster_types] \n",
    "            + [sa.Column('annotation_{}'.format(cluster_type), sa.String(20), nullable=True) \n",
    "            for cluster_type in cluster_types]\n",
    "           )\n",
    "    \n",
    "    # combine cols\n",
    "    args = args_tsne + args_cluster\n",
    "    # create the table\n",
    "    table = sa.Table(ens, metadata,\n",
    "                        # NOT auto increment not null primary key\n",
    "                        sa.Column('cell_id', sa.Integer, sa.ForeignKey('cells.cell_id'), primary_key=True, autoincrement=False), \n",
    "                        *args\n",
    "    )\n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create individual ensemble tables (ensemble level)\n",
    "# demo here\n",
    "ens = 'Ens0'\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "# load specfic tables needed to be referenced as foreign key\n",
    "sa.Table('cells', metadata, autoload=True)\n",
    "# define and create \n",
    "ens_table = define_ens_table(metadata, ens)\n",
    "metadata.create_all(engine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update to ensemble tables (ensemble level)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to mysql\n",
    "### Dataset level\n",
    "- upload to cells table\n",
    "- upload to gene tables (bottle-neck)\n",
    "\n",
    "### Ensemble level\n",
    "- upload to enss table\n",
    "- create ens table\n",
    "- upload to ens table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSMUSG00000051951.5'"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def gene_id_to_name(gene_id, df_genes):\n",
    "    \"\"\"df_genes\n",
    "    \"\"\"\n",
    "    return df_genes.loc[gene_id, 'gene_name']\n",
    "\n",
    "def gene_name_to_id(gene_name, df_genes):\n",
    "    \"\"\"df_genes\n",
    "    \"\"\"\n",
    "    return df_genes[df_genes['gene_name'] == gene_name].index.values[0]\n",
    "\n",
    "gene_name_to_id('Xkr4', df_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001.4</th>\n",
       "      <td>Gnai3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>108107280</td>\n",
       "      <td>108146146</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003.15</th>\n",
       "      <td>Pbsn</td>\n",
       "      <td>chrX</td>\n",
       "      <td>77837901</td>\n",
       "      <td>77853623</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.14</th>\n",
       "      <td>Cdc45</td>\n",
       "      <td>chr16</td>\n",
       "      <td>18780447</td>\n",
       "      <td>18811987</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031.16</th>\n",
       "      <td>H19</td>\n",
       "      <td>chr7</td>\n",
       "      <td>142575529</td>\n",
       "      <td>142578143</td>\n",
       "      <td>-</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037.16</th>\n",
       "      <td>Scml2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>161117193</td>\n",
       "      <td>161258213</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      gene_name    chr      start        end strand  \\\n",
       "gene_id                                                               \n",
       "ENSMUSG00000000001.4      Gnai3   chr3  108107280  108146146      -   \n",
       "ENSMUSG00000000003.15      Pbsn   chrX   77837901   77853623      -   \n",
       "ENSMUSG00000000028.14     Cdc45  chr16   18780447   18811987      -   \n",
       "ENSMUSG00000000031.16       H19   chr7  142575529  142578143      -   \n",
       "ENSMUSG00000000037.16     Scml2   chrX  161117193  161258213      +   \n",
       "\n",
       "                            gene_type  \n",
       "gene_id                                \n",
       "ENSMUSG00000000001.4   protein_coding  \n",
       "ENSMUSG00000000003.15  protein_coding  \n",
       "ENSMUSG00000000028.14  protein_coding  \n",
       "ENSMUSG00000000031.16         lincRNA  \n",
       "ENSMUSG00000000037.16  protein_coding  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "engine = connect_sql(database)\n",
    "sql = \"\"\"SELECT * FROM genes\"\"\"\n",
    "df_genes = pd.read_sql(sql, engine, index_col='gene_id')\n",
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9621, 1)\n",
      "(9621, 2)\n",
      "(24848, 9621)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGCGATAGAACCAGGTAACGAACGCCTATCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTAAGAGATGTAATCTTA</th>\n",
       "      <th>AGCGATAGAACCAGGTAATGACGTCCTATCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTAGGATAACATAGAGGC</th>\n",
       "      <th>AGCGATAGAACCAGGTAGGATAACCCTATCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTAGGATAACTATAGCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTATAGCCTTATAGAGGC</th>\n",
       "      <th>AGCGATAGAACCAGGTATAGCCTTCCTATCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTATTCGTTGATAGAGGC</th>\n",
       "      <th>AGCGATAGAACCAGGTATTCGTTGGTACTGAC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCTCGCGCTTAGCCTCAAGAGATGTATAGCCT</th>\n",
       "      <th>TCTCGCGCTTAGCCTCATAGCCTTCAGGACGT</th>\n",
       "      <th>TCTCGCGCTTAGCCTCATTCGTTGGTACTGAC</th>\n",
       "      <th>TCTCGCGCTTCCATCCAGGATAACAGGCGAAG</th>\n",
       "      <th>TCTCGCGCTTCCATCCATTCGTTGCAGGACGT</th>\n",
       "      <th>TCTCGCGCTTCCATCCGGTTAGACCAGGACGT</th>\n",
       "      <th>TCTCGCGCTTCCATCCTAAGATCCAGGCGAAG</th>\n",
       "      <th>TCTCGCGCTTCCATCCTTACGACCCAGGACGT</th>\n",
       "      <th>TCTCGCGCTTCCATCCTTCATCCAAGGCGAAG</th>\n",
       "      <th>TCTCGCGCTTCCATCCTTGGAAGTTAATCTTA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>1.2309</td>\n",
       "      <td>2.2035</td>\n",
       "      <td>1.4904</td>\n",
       "      <td>2.7199</td>\n",
       "      <td>1.9671</td>\n",
       "      <td>0.6435</td>\n",
       "      <td>1.1516</td>\n",
       "      <td>1.2949</td>\n",
       "      <td>1.0551</td>\n",
       "      <td>1.3511</td>\n",
       "      <td>...</td>\n",
       "      <td>1.5819</td>\n",
       "      <td>2.5104</td>\n",
       "      <td>1.1699</td>\n",
       "      <td>0.8651</td>\n",
       "      <td>1.8328</td>\n",
       "      <td>0.8675</td>\n",
       "      <td>0.6887</td>\n",
       "      <td>0.1175</td>\n",
       "      <td>0.5993</td>\n",
       "      <td>0.7535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ncaph2</th>\n",
       "      <td>0.6129</td>\n",
       "      <td>0.8373</td>\n",
       "      <td>0.3274</td>\n",
       "      <td>1.0381</td>\n",
       "      <td>0.6996</td>\n",
       "      <td>0.4499</td>\n",
       "      <td>0.3647</td>\n",
       "      <td>0.3174</td>\n",
       "      <td>0.7224</td>\n",
       "      <td>0.4432</td>\n",
       "      <td>...</td>\n",
       "      <td>0.7459</td>\n",
       "      <td>0.1373</td>\n",
       "      <td>0.2042</td>\n",
       "      <td>0.4358</td>\n",
       "      <td>0.9827</td>\n",
       "      <td>0.9145</td>\n",
       "      <td>0.9568</td>\n",
       "      <td>0.3386</td>\n",
       "      <td>0.4358</td>\n",
       "      <td>0.3914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eif4enif1</th>\n",
       "      <td>0.2243</td>\n",
       "      <td>0.5336</td>\n",
       "      <td>0.5550</td>\n",
       "      <td>0.3044</td>\n",
       "      <td>0.3740</td>\n",
       "      <td>0.9099</td>\n",
       "      <td>0.3129</td>\n",
       "      <td>0.2693</td>\n",
       "      <td>0.4159</td>\n",
       "      <td>0.3219</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5529</td>\n",
       "      <td>0.1557</td>\n",
       "      <td>0.0879</td>\n",
       "      <td>0.5844</td>\n",
       "      <td>0.7681</td>\n",
       "      <td>0.2178</td>\n",
       "      <td>1.4043</td>\n",
       "      <td>0.8182</td>\n",
       "      <td>0.5844</td>\n",
       "      <td>0.1908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rgr</th>\n",
       "      <td>0.1747</td>\n",
       "      <td>0.6393</td>\n",
       "      <td>0.3970</td>\n",
       "      <td>0.1009</td>\n",
       "      <td>0.1228</td>\n",
       "      <td>0.0726</td>\n",
       "      <td>0.2763</td>\n",
       "      <td>0.0862</td>\n",
       "      <td>0.2224</td>\n",
       "      <td>0.2941</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4020</td>\n",
       "      <td>0.6971</td>\n",
       "      <td>0.3574</td>\n",
       "      <td>0.1004</td>\n",
       "      <td>0.2802</td>\n",
       "      <td>0.0441</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1004</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mapkapk2</th>\n",
       "      <td>1.7354</td>\n",
       "      <td>0.9501</td>\n",
       "      <td>1.4129</td>\n",
       "      <td>0.5049</td>\n",
       "      <td>1.0696</td>\n",
       "      <td>0.5063</td>\n",
       "      <td>0.4804</td>\n",
       "      <td>1.0927</td>\n",
       "      <td>0.8737</td>\n",
       "      <td>1.0037</td>\n",
       "      <td>...</td>\n",
       "      <td>2.1037</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>0.4215</td>\n",
       "      <td>1.1965</td>\n",
       "      <td>1.7008</td>\n",
       "      <td>1.2285</td>\n",
       "      <td>1.8480</td>\n",
       "      <td>0.2920</td>\n",
       "      <td>1.1965</td>\n",
       "      <td>0.4760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9621 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           AGCGATAGAACCAGGTAACGAACGCCTATCCT  AGCGATAGAACCAGGTAAGAGATGTAATCTTA  \\\n",
       "Xkr4                                 1.2309                            2.2035   \n",
       "Ncaph2                               0.6129                            0.8373   \n",
       "Eif4enif1                            0.2243                            0.5336   \n",
       "Rgr                                  0.1747                            0.6393   \n",
       "Mapkapk2                             1.7354                            0.9501   \n",
       "\n",
       "           AGCGATAGAACCAGGTAATGACGTCCTATCCT  AGCGATAGAACCAGGTAGGATAACATAGAGGC  \\\n",
       "Xkr4                                 1.4904                            2.7199   \n",
       "Ncaph2                               0.3274                            1.0381   \n",
       "Eif4enif1                            0.5550                            0.3044   \n",
       "Rgr                                  0.3970                            0.1009   \n",
       "Mapkapk2                             1.4129                            0.5049   \n",
       "\n",
       "           AGCGATAGAACCAGGTAGGATAACCCTATCCT  AGCGATAGAACCAGGTAGGATAACTATAGCCT  \\\n",
       "Xkr4                                 1.9671                            0.6435   \n",
       "Ncaph2                               0.6996                            0.4499   \n",
       "Eif4enif1                            0.3740                            0.9099   \n",
       "Rgr                                  0.1228                            0.0726   \n",
       "Mapkapk2                             1.0696                            0.5063   \n",
       "\n",
       "           AGCGATAGAACCAGGTATAGCCTTATAGAGGC  AGCGATAGAACCAGGTATAGCCTTCCTATCCT  \\\n",
       "Xkr4                                 1.1516                            1.2949   \n",
       "Ncaph2                               0.3647                            0.3174   \n",
       "Eif4enif1                            0.3129                            0.2693   \n",
       "Rgr                                  0.2763                            0.0862   \n",
       "Mapkapk2                             0.4804                            1.0927   \n",
       "\n",
       "           AGCGATAGAACCAGGTATTCGTTGATAGAGGC  AGCGATAGAACCAGGTATTCGTTGGTACTGAC  \\\n",
       "Xkr4                                 1.0551                            1.3511   \n",
       "Ncaph2                               0.7224                            0.4432   \n",
       "Eif4enif1                            0.4159                            0.3219   \n",
       "Rgr                                  0.2224                            0.2941   \n",
       "Mapkapk2                             0.8737                            1.0037   \n",
       "\n",
       "                         ...                 TCTCGCGCTTAGCCTCAAGAGATGTATAGCCT  \\\n",
       "Xkr4                     ...                                           1.5819   \n",
       "Ncaph2                   ...                                           0.7459   \n",
       "Eif4enif1                ...                                           0.5529   \n",
       "Rgr                      ...                                           0.4020   \n",
       "Mapkapk2                 ...                                           2.1037   \n",
       "\n",
       "           TCTCGCGCTTAGCCTCATAGCCTTCAGGACGT  TCTCGCGCTTAGCCTCATTCGTTGGTACTGAC  \\\n",
       "Xkr4                                 2.5104                            1.1699   \n",
       "Ncaph2                               0.1373                            0.2042   \n",
       "Eif4enif1                            0.1557                            0.0879   \n",
       "Rgr                                  0.6971                            0.3574   \n",
       "Mapkapk2                             0.8516                            0.4215   \n",
       "\n",
       "           TCTCGCGCTTCCATCCAGGATAACAGGCGAAG  TCTCGCGCTTCCATCCATTCGTTGCAGGACGT  \\\n",
       "Xkr4                                 0.8651                            1.8328   \n",
       "Ncaph2                               0.4358                            0.9827   \n",
       "Eif4enif1                            0.5844                            0.7681   \n",
       "Rgr                                  0.1004                            0.2802   \n",
       "Mapkapk2                             1.1965                            1.7008   \n",
       "\n",
       "           TCTCGCGCTTCCATCCGGTTAGACCAGGACGT  TCTCGCGCTTCCATCCTAAGATCCAGGCGAAG  \\\n",
       "Xkr4                                 0.8675                            0.6887   \n",
       "Ncaph2                               0.9145                            0.9568   \n",
       "Eif4enif1                            0.2178                            1.4043   \n",
       "Rgr                                  0.0441                            0.0000   \n",
       "Mapkapk2                             1.2285                            1.8480   \n",
       "\n",
       "           TCTCGCGCTTCCATCCTTACGACCCAGGACGT  TCTCGCGCTTCCATCCTTCATCCAAGGCGAAG  \\\n",
       "Xkr4                                 0.1175                            0.5993   \n",
       "Ncaph2                               0.3386                            0.4358   \n",
       "Eif4enif1                            0.8182                            0.5844   \n",
       "Rgr                                  0.0000                            0.1004   \n",
       "Mapkapk2                             0.2920                            1.1965   \n",
       "\n",
       "           TCTCGCGCTTCCATCCTTGGAAGTTAATCTTA  \n",
       "Xkr4                                 0.7535  \n",
       "Ncaph2                               0.3914  \n",
       "Eif4enif1                            0.1908  \n",
       "Rgr                                  0.0000  \n",
       "Mapkapk2                             0.4760  \n",
       "\n",
       "[5 rows x 9621 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get gene*cell matrix\n",
    "dataset = 'CEMBA_3C_171207'\n",
    "with cd('/cndd/Public_Datasets/CEMBA/snATACSeq/MiniBrain_fromRongxin'):\n",
    "    df_clst = pd.read_table('XW46_cluster.txt', header=None, names=['sample', 'cluster_ID'], index_col='sample')\n",
    "    df_tsne = pd.read_table('XW46_tsne.txt', header=None, names=['sample', 'tsne_x', 'tsne_y'], index_col='sample')\n",
    "    df = pd.read_table('XW46_gene.txt.gz') \n",
    "\n",
    "df = df.T\n",
    "    \n",
    "print(df_clst.shape)\n",
    "print(df_tsne.shape)\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24266, 9621)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGCGATAGAACCAGGTAACGAACGCCTATCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTAAGAGATGTAATCTTA</th>\n",
       "      <th>AGCGATAGAACCAGGTAATGACGTCCTATCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTAGGATAACATAGAGGC</th>\n",
       "      <th>AGCGATAGAACCAGGTAGGATAACCCTATCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTAGGATAACTATAGCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTATAGCCTTATAGAGGC</th>\n",
       "      <th>AGCGATAGAACCAGGTATAGCCTTCCTATCCT</th>\n",
       "      <th>AGCGATAGAACCAGGTATTCGTTGATAGAGGC</th>\n",
       "      <th>AGCGATAGAACCAGGTATTCGTTGGTACTGAC</th>\n",
       "      <th>...</th>\n",
       "      <th>TCTCGCGCTTAGCCTCAAGAGATGTATAGCCT</th>\n",
       "      <th>TCTCGCGCTTAGCCTCATAGCCTTCAGGACGT</th>\n",
       "      <th>TCTCGCGCTTAGCCTCATTCGTTGGTACTGAC</th>\n",
       "      <th>TCTCGCGCTTCCATCCAGGATAACAGGCGAAG</th>\n",
       "      <th>TCTCGCGCTTCCATCCATTCGTTGCAGGACGT</th>\n",
       "      <th>TCTCGCGCTTCCATCCGGTTAGACCAGGACGT</th>\n",
       "      <th>TCTCGCGCTTCCATCCTAAGATCCAGGCGAAG</th>\n",
       "      <th>TCTCGCGCTTCCATCCTTACGACCCAGGACGT</th>\n",
       "      <th>TCTCGCGCTTCCATCCTTCATCCAAGGCGAAG</th>\n",
       "      <th>TCTCGCGCTTCCATCCTTGGAAGTTAATCTTA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001.4</th>\n",
       "      <td>1.4893</td>\n",
       "      <td>2.6315</td>\n",
       "      <td>1.4147</td>\n",
       "      <td>4.2155</td>\n",
       "      <td>2.3922</td>\n",
       "      <td>1.953</td>\n",
       "      <td>1.1681</td>\n",
       "      <td>2.0360</td>\n",
       "      <td>1.6638</td>\n",
       "      <td>1.4493</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5976</td>\n",
       "      <td>1.2706</td>\n",
       "      <td>0.6035</td>\n",
       "      <td>1.8261</td>\n",
       "      <td>3.3573</td>\n",
       "      <td>3.3080</td>\n",
       "      <td>2.0872</td>\n",
       "      <td>0.7567</td>\n",
       "      <td>1.8261</td>\n",
       "      <td>1.2383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003.15</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.14</th>\n",
       "      <td>0.3212</td>\n",
       "      <td>0.8244</td>\n",
       "      <td>0.2193</td>\n",
       "      <td>0.0973</td>\n",
       "      <td>0.4030</td>\n",
       "      <td>0.139</td>\n",
       "      <td>0.7995</td>\n",
       "      <td>0.2772</td>\n",
       "      <td>0.2051</td>\n",
       "      <td>0.2167</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2490</td>\n",
       "      <td>0.5776</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2935</td>\n",
       "      <td>0.3072</td>\n",
       "      <td>0.4671</td>\n",
       "      <td>0.5941</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2232</td>\n",
       "      <td>0.4370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031.16</th>\n",
       "      <td>0.1613</td>\n",
       "      <td>0.2949</td>\n",
       "      <td>0.1389</td>\n",
       "      <td>0.0830</td>\n",
       "      <td>0.1155</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.1784</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0868</td>\n",
       "      <td>0.3343</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4269</td>\n",
       "      <td>0.3581</td>\n",
       "      <td>0.1957</td>\n",
       "      <td>0.2701</td>\n",
       "      <td>0.6216</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1998</td>\n",
       "      <td>0.1974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037.16</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9621 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AGCGATAGAACCAGGTAACGAACGCCTATCCT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             1.4893   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.3212   \n",
       "ENSMUSG00000000031.16                            0.1613   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTAAGAGATGTAATCTTA  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             2.6315   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.8244   \n",
       "ENSMUSG00000000031.16                            0.2949   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTAATGACGTCCTATCCT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             1.4147   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.2193   \n",
       "ENSMUSG00000000031.16                            0.1389   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTAGGATAACATAGAGGC  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             4.2155   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.0973   \n",
       "ENSMUSG00000000031.16                            0.0830   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTAGGATAACCCTATCCT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             2.3922   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.4030   \n",
       "ENSMUSG00000000031.16                            0.1155   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTAGGATAACTATAGCCT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                              1.953   \n",
       "ENSMUSG00000000003.15                             0.000   \n",
       "ENSMUSG00000000028.14                             0.139   \n",
       "ENSMUSG00000000031.16                             0.000   \n",
       "ENSMUSG00000000037.16                             0.000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTATAGCCTTATAGAGGC  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             1.1681   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.7995   \n",
       "ENSMUSG00000000031.16                            0.1784   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTATAGCCTTCCTATCCT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             2.0360   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.2772   \n",
       "ENSMUSG00000000031.16                            0.0000   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTATTCGTTGATAGAGGC  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             1.6638   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.2051   \n",
       "ENSMUSG00000000031.16                            0.0868   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       AGCGATAGAACCAGGTATTCGTTGGTACTGAC  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             1.4493   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.2167   \n",
       "ENSMUSG00000000031.16                            0.3343   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                                     ...                 \\\n",
       "gene_id                              ...                  \n",
       "ENSMUSG00000000001.4                 ...                  \n",
       "ENSMUSG00000000003.15                ...                  \n",
       "ENSMUSG00000000028.14                ...                  \n",
       "ENSMUSG00000000031.16                ...                  \n",
       "ENSMUSG00000000037.16                ...                  \n",
       "\n",
       "                       TCTCGCGCTTAGCCTCAAGAGATGTATAGCCT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             3.5976   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.2490   \n",
       "ENSMUSG00000000031.16                            0.4269   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTAGCCTCATAGCCTTCAGGACGT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             1.2706   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.5776   \n",
       "ENSMUSG00000000031.16                            0.3581   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTAGCCTCATTCGTTGGTACTGAC  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             0.6035   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.0000   \n",
       "ENSMUSG00000000031.16                            0.1957   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTCCATCCAGGATAACAGGCGAAG  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             1.8261   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.2935   \n",
       "ENSMUSG00000000031.16                            0.2701   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTCCATCCATTCGTTGCAGGACGT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             3.3573   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.3072   \n",
       "ENSMUSG00000000031.16                            0.6216   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTCCATCCGGTTAGACCAGGACGT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             3.3080   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.4671   \n",
       "ENSMUSG00000000031.16                            0.9398   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTCCATCCTAAGATCCAGGCGAAG  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             2.0872   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.5941   \n",
       "ENSMUSG00000000031.16                            0.0000   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTCCATCCTTACGACCCAGGACGT  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             0.7567   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.0000   \n",
       "ENSMUSG00000000031.16                            0.0000   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTCCATCCTTCATCCAAGGCGAAG  \\\n",
       "gene_id                                                   \n",
       "ENSMUSG00000000001.4                             1.8261   \n",
       "ENSMUSG00000000003.15                            0.0000   \n",
       "ENSMUSG00000000028.14                            0.2232   \n",
       "ENSMUSG00000000031.16                            0.1998   \n",
       "ENSMUSG00000000037.16                            0.0000   \n",
       "\n",
       "                       TCTCGCGCTTCCATCCTTGGAAGTTAATCTTA  \n",
       "gene_id                                                  \n",
       "ENSMUSG00000000001.4                             1.2383  \n",
       "ENSMUSG00000000003.15                            0.0000  \n",
       "ENSMUSG00000000028.14                            0.4370  \n",
       "ENSMUSG00000000031.16                            0.1974  \n",
       "ENSMUSG00000000037.16                            0.0000  \n",
       "\n",
       "[5 rows x 9621 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_res = pd.merge(df_genes[['gene_name']], df, left_on='gene_name', right_index=True)\n",
    "df_res = df_res.drop('gene_name', axis=1)\n",
    "print(df_res.shape)\n",
    "df_res.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to cells table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9621, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_name</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CEMBA_3C_171207_AGCGATAGAACCAGGTAACGAACGCCTATCCT</td>\n",
       "      <td>CEMBA_3C_171207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CEMBA_3C_171207_AGCGATAGAACCAGGTAAGAGATGTAATCTTA</td>\n",
       "      <td>CEMBA_3C_171207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CEMBA_3C_171207_AGCGATAGAACCAGGTAATGACGTCCTATCCT</td>\n",
       "      <td>CEMBA_3C_171207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CEMBA_3C_171207_AGCGATAGAACCAGGTAGGATAACATAGAGGC</td>\n",
       "      <td>CEMBA_3C_171207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CEMBA_3C_171207_AGCGATAGAACCAGGTAGGATAACCCTATCCT</td>\n",
       "      <td>CEMBA_3C_171207</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          cell_name          dataset\n",
       "0  CEMBA_3C_171207_AGCGATAGAACCAGGTAACGAACGCCTATCCT  CEMBA_3C_171207\n",
       "1  CEMBA_3C_171207_AGCGATAGAACCAGGTAAGAGATGTAATCTTA  CEMBA_3C_171207\n",
       "2  CEMBA_3C_171207_AGCGATAGAACCAGGTAATGACGTCCTATCCT  CEMBA_3C_171207\n",
       "3  CEMBA_3C_171207_AGCGATAGAACCAGGTAGGATAACATAGAGGC  CEMBA_3C_171207\n",
       "4  CEMBA_3C_171207_AGCGATAGAACCAGGTAGGATAACCCTATCCT  CEMBA_3C_171207"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cells = pd.DataFrame()\n",
    "df_cells['cell_name'] = dataset + '_' + df_clst.index.values\n",
    "df_cells['dataset'] = dataset \n",
    "\n",
    "print(df_cells.shape)\n",
    "df_cells.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/20/2018 09:52:14 PM                                           cell_name          dataset\n",
      "0  CEMBA_3C_171207_AGCGATAGAACCAGGTAACGAACGCCTATCCT  CEMBA_3C_171207\n",
      "1  CEMBA_3C_171207_AGCGATAGAACCAGGTAAGAGATGTAATCTTA  CEMBA_3C_171207\n",
      "2  CEMBA_3C_171207_AGCGATAGAACCAGGTAATGACGTCCTATCCT  CEMBA_3C_171207\n",
      "3  CEMBA_3C_171207_AGCGATAGAACCAGGTAGGATAACATAGAGGC  CEMBA_3C_171207\n",
      "4  CEMBA_3C_171207_AGCGATAGAACCAGGTAGGATAACCCTATCCT  CEMBA_3C_171207\n",
      "03/20/2018 09:52:15 PM Upload complete.\n"
     ]
    }
   ],
   "source": [
    "# insert into\n",
    "insert_into(engine, 'cells', df_cells, ignore=False, verbose='True')\n",
    "logging.info(\"Upload complete.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to gene tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      cell_id  normalized_counts\n",
      "8567     8568             1.4893\n",
      "8568     8569             2.6315\n",
      "8569     8570             1.4147\n",
      "8570     8571             4.2155\n",
      "8571     8572             2.3922\n"
     ]
    }
   ],
   "source": [
    "sql = \"\"\"SELECT * FROM cells\"\"\"\n",
    "df_cells = pd.read_sql(sql, engine)\n",
    "\n",
    "metadata = sa.MetaData(engine)\n",
    "\n",
    "for i, (gene_id, row) in enumerate(df_res.iterrows()):\n",
    "    gene_table_name = gene_id_to_table_name(gene_id)\n",
    "#     if (i%10==0):\n",
    "#         logging.info(\"Progress on genes: {} {}\".format(i+1, gene_table_name))\n",
    "    if i == 0:\n",
    "        df_gene = row.to_frame('normalized_counts')\n",
    "        df_gene.index = dataset + '_' + df_gene.index.values\n",
    "        df_gene = pd.merge(df_gene, df_cells, left_index=True, right_on='cell_name', how='left')\n",
    "        df_gene = df_gene[['cell_id', 'normalized_counts']]\n",
    "        print(df_gene.head())\n",
    "\n",
    "#     if i == 0:\n",
    "#         df_gene = row.to_frame('normalized_counts')\n",
    "#         df_gene = pd.merge(df_gene, df_cells, left_index=True, right_on='cell_name', how='left')\n",
    "#         df_gene = df_gene[['cell_id', 'normalized_counts']]\n",
    "#         print(df_gene.head())\n",
    "#         insert_into(engine, gene_table_name, df_gene, ignore=True, verbose=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlalchemy.engine.result.ResultProxy at 0x7fcc89f1aa58>"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Upload to ensembles table\n",
    "ens = 'Ens2'\n",
    "ens_name = 'CEMBA_3C_171207'\n",
    "ens_datasets=['CEMBA_3C_171207']\n",
    "\n",
    "\n",
    "table_name = 'ensembles'\n",
    "ens_id = int(ens[len('Ens'):])\n",
    "dict_list = [{'ensemble_id': ens_id, 'ensemble_name': ens_name, 'public_access': False, 'datasets': ','.join(ens_datasets)}]\n",
    "insert_into_worker(engine, table_name, dict_list, ignore=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create and upload to ensemble table (2 ensembles)\n",
    "\n",
    "def create_and_upload_to_ens(ens, df_ens, database):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    \n",
    "    logging.info(\"Creating {} table...\".format(ens))\n",
    "    engine = connect_sql(database)\n",
    "    # create ens table\n",
    "    metadata = sa.MetaData(engine)\n",
    "    # load specfic tables needed to referenced as foreign key\n",
    "    sa.Table('cells', metadata, autoload=True)\n",
    "    # define ensemble table\n",
    "    ens_table = define_ens_table(metadata, ens)\n",
    "    metadata.create_all()\n",
    "    logging.info(\"Done creating {} table!\".format(ens))\n",
    "\n",
    "    # upload to ens table\n",
    "    logging.info(\"Uploading to {} table...\".format(ens))\n",
    "\n",
    "    engine = connect_sql(database)\n",
    "    table_name = ens \n",
    "    insert_into(engine, table_name, df_ens, ignore=False, verbose=True)\n",
    "\n",
    "    logging.info(\"Done uploading to {} table!\".format(ens))\n",
    "\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/20/2018 09:59:55 PM Creating Ens2 table...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ens2\n",
      "      cell_id  cluster_ATAC  tsne_x_ATAC  tsne_y_ATAC\n",
      "8567     8568            13    -9.351734    30.328056\n",
      "8568     8569            19   -32.812104     4.672624\n",
      "8569     8570             8     3.887879    -2.134793\n",
      "8570     8571            17    32.098704    19.152496\n",
      "8571     8572            20    -1.280459    20.562413\n",
      "CEMBA_snATAC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03/20/2018 09:59:56 PM Done creating Ens2 table!\n",
      "03/20/2018 09:59:56 PM Uploading to Ens2 table...\n",
      "03/20/2018 09:59:56 PM       cell_id  cluster_ATAC  tsne_x_ATAC  tsne_y_ATAC\n",
      "8567     8568            13    -9.351734    30.328056\n",
      "8568     8569            19   -32.812104     4.672624\n",
      "8569     8570             8     3.887879    -2.134793\n",
      "8570     8571            17    32.098704    19.152496\n",
      "8571     8572            20    -1.280459    20.562413\n",
      "03/20/2018 09:59:57 PM Done uploading to Ens2 table!\n"
     ]
    }
   ],
   "source": [
    "df_ens = pd.merge(df_clst, df_tsne, left_index=True, right_index=True)\n",
    "df_ens.columns = ['cluster_ATAC', 'tsne_x_ATAC', 'tsne_y_ATAC']\n",
    "df_ens.index = dataset + '_' + df_ens.index.values\n",
    "df_ens['cluster_ATAC'] = [int(cluster.strip('C')) for cluster in df_ens['cluster_ATAC']]\n",
    "\n",
    "# get cells\n",
    "sql = \"\"\"SELECT * FROM cells\"\"\"\n",
    "df_cells = pd.read_sql(sql, engine)[['cell_id', 'cell_name']]\n",
    "\n",
    "# get cell_id\n",
    "df_ens = pd.merge(df_cells, df_ens, left_on='cell_name', right_index=True, how='right')\n",
    "df_ens = df_ens.drop('cell_name', axis=1)\n",
    "\n",
    "print(ens)\n",
    "print(df_ens.head())\n",
    "print(database)\n",
    "create_and_upload_to_ens(ens, df_ens, database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # update a table\n",
    "# sql = \"\"\"SELECT * FROM cells\"\"\"\n",
    "# df_cells = pd.read_sql(sql, engine)\n",
    "# df_cells['cell_name'] = df_cells['dataset'].values[0] + '_' + df_cells['cell_name']\n",
    "\n",
    "# metadata = sa.MetaData(engine)\n",
    "# table = sa.Table('cells', metadata, autoload=True)\n",
    "# for i, row in df_cells.iterrows():\n",
    "#     updater = update(table).where(table.c.cell_id==row.cell_id).values(cell_name=row.cell_name)\n",
    "#     engine.execute(updater)\n",
    "\n",
    "\n",
    "\n",
    "# delete a table\n",
    "#     gene_table = sa.Table(gene_table_name, metadata, autoload=True)\n",
    "#     d = gene_table.delete()\n",
    "#     d.execute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
